{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "214a36bc-3bf0-415a-a8c9-718fb6ed5d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle, os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "155b62bb-5767-40cd-b4d6-0c3cf69c7663",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"residues_physchem\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7a35505-4e1f-46fa-9d83-f001819bc43a",
   "metadata": {},
   "source": [
    "# Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "426d9154-c7e7-4e4c-be28-6d28dd82a371",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"7\" halign=\"left\">Residues</th>\n",
       "      <th>Label</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Amino acids</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">HHBlits</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>label_entity_id</th>\n",
       "      <th>label_asym_id</th>\n",
       "      <th>label_seq_id</th>\n",
       "      <th>auth_asym_id</th>\n",
       "      <th>auth_seq_id</th>\n",
       "      <th>pdbx_PDB_ins_code</th>\n",
       "      <th>label</th>\n",
       "      <th>label_comp_id_A</th>\n",
       "      <th>label_comp_id_C</th>\n",
       "      <th>...</th>\n",
       "      <th>M-&gt;M</th>\n",
       "      <th>M-&gt;I</th>\n",
       "      <th>M-&gt;D</th>\n",
       "      <th>I-&gt;M</th>\n",
       "      <th>I-&gt;I</th>\n",
       "      <th>D-&gt;M</th>\n",
       "      <th>D-&gt;D</th>\n",
       "      <th>Neff</th>\n",
       "      <th>Neff_I</th>\n",
       "      <th>Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6ta3</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.998615</td>\n",
       "      <td>0.001223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.088</td>\n",
       "      <td>1.005</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6ta3</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>2</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.860</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6ta3</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997923</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.656106</td>\n",
       "      <td>0.343885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.988</td>\n",
       "      <td>1.007</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6ta3</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999307</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333325</td>\n",
       "      <td>0.666649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.047</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6ta3</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.026</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>3d2p</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>452</td>\n",
       "      <td>A</td>\n",
       "      <td>432</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.906</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>3d2p</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>453</td>\n",
       "      <td>A</td>\n",
       "      <td>433</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.787</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>3d2p</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>454</td>\n",
       "      <td>A</td>\n",
       "      <td>434</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994470</td>\n",
       "      <td>0.005799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.713</td>\n",
       "      <td>1.002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>3d2p</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>455</td>\n",
       "      <td>A</td>\n",
       "      <td>435</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.639</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>423</th>\n",
       "      <td>3d2p</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>456</td>\n",
       "      <td>A</td>\n",
       "      <td>436</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.958</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108406 rows × 174 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Residues                                                          \\\n",
       "         pdb label_entity_id label_asym_id label_seq_id auth_asym_id   \n",
       "0       6ta3               2             B            1            A   \n",
       "1       6ta3               2             B            2            A   \n",
       "2       6ta3               2             B            3            A   \n",
       "3       6ta3               2             B            4            A   \n",
       "4       6ta3               2             B            5            A   \n",
       "..       ...             ...           ...          ...          ...   \n",
       "419     3d2p               1             A          452            A   \n",
       "420     3d2p               1             A          453            A   \n",
       "421     3d2p               1             A          454            A   \n",
       "422     3d2p               1             A          455            A   \n",
       "423     3d2p               1             A          456            A   \n",
       "\n",
       "                                  Label     Amino acids                  ...  \\\n",
       "    auth_seq_id pdbx_PDB_ins_code label label_comp_id_A label_comp_id_C  ...   \n",
       "0             1                 ?     0               0               0  ...   \n",
       "1             2                 ?     0               0               0  ...   \n",
       "2             3                 ?     0               0               0  ...   \n",
       "3             4                 ?     0               0               1  ...   \n",
       "4             5                 ?     0               0               0  ...   \n",
       "..          ...               ...   ...             ...             ...  ...   \n",
       "419         432                 ?     0               0               0  ...   \n",
       "420         433                 ?     0               0               0  ...   \n",
       "421         434                 ?     0               0               0  ...   \n",
       "422         435                 ?     0               0               0  ...   \n",
       "423         436                 ?     0               0               0  ...   \n",
       "\n",
       "      HHBlits                                                              \\\n",
       "         M->M      M->I M->D      I->M      I->I D->M D->D    Neff Neff_I   \n",
       "0    0.998615  0.001223  0.0  1.000000  0.000000  0.0  0.0   9.088  1.005   \n",
       "1    1.000000  0.000000  0.0  0.000000  0.000000  0.0  0.0   9.860  0.000   \n",
       "2    0.997923  0.001963  0.0  0.656106  0.343885  0.0  0.0   9.988  1.007   \n",
       "3    0.999307  0.000605  0.0  0.333325  0.666649  0.0  0.0  10.047  1.000   \n",
       "4    1.000000  0.000000  0.0  0.000000  0.000000  0.0  0.0  10.026  0.000   \n",
       "..        ...       ...  ...       ...       ...  ...  ...     ...    ...   \n",
       "419  1.000000  0.000000  0.0  0.000000  0.000000  0.0  0.0  10.906  0.000   \n",
       "420  1.000000  0.000000  0.0  0.000000  0.000000  0.0  0.0  10.787  0.000   \n",
       "421  0.994470  0.005799  0.0  1.000000  0.000000  0.0  0.0  10.713  1.002   \n",
       "422  1.000000  0.000000  0.0  0.000000  0.000000  0.0  0.0  10.639  0.000   \n",
       "423  1.000000  0.000000  0.0  1.000000  0.000000  0.0  0.0  10.958  0.000   \n",
       "\n",
       "            \n",
       "    Neff_D  \n",
       "0      0.0  \n",
       "1      0.0  \n",
       "2      0.0  \n",
       "3      0.0  \n",
       "4      0.0  \n",
       "..     ...  \n",
       "419    0.0  \n",
       "420    0.0  \n",
       "421    0.0  \n",
       "422    0.0  \n",
       "423    0.0  \n",
       "\n",
       "[108406 rows x 174 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"../training_data/6.Training_sets/residues_train.pkl\", \"rb\") as f:\n",
    "    train_data = pickle.load(f)\n",
    "\n",
    "train_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dd30ddbb-5492-4633-812e-27a17344ff54",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"7\" halign=\"left\">Residues</th>\n",
       "      <th>Label</th>\n",
       "      <th colspan=\"2\" halign=\"left\">Amino acids</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">HHBlits</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>pdb</th>\n",
       "      <th>label_entity_id</th>\n",
       "      <th>label_asym_id</th>\n",
       "      <th>label_seq_id</th>\n",
       "      <th>auth_asym_id</th>\n",
       "      <th>auth_seq_id</th>\n",
       "      <th>pdbx_PDB_ins_code</th>\n",
       "      <th>label</th>\n",
       "      <th>label_comp_id_A</th>\n",
       "      <th>label_comp_id_C</th>\n",
       "      <th>...</th>\n",
       "      <th>M-&gt;M</th>\n",
       "      <th>M-&gt;I</th>\n",
       "      <th>M-&gt;D</th>\n",
       "      <th>I-&gt;M</th>\n",
       "      <th>I-&gt;I</th>\n",
       "      <th>D-&gt;M</th>\n",
       "      <th>D-&gt;D</th>\n",
       "      <th>Neff</th>\n",
       "      <th>Neff_I</th>\n",
       "      <th>Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5kwj</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>0</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.985549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014548</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.217</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5kwj</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "      <td>1</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.268</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5kwj</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>2</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988286</td>\n",
       "      <td>0.011817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.773</td>\n",
       "      <td>1.031</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5kwj</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.995850</td>\n",
       "      <td>0.004370</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.926</td>\n",
       "      <td>1.004</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5kwj</td>\n",
       "      <td>1</td>\n",
       "      <td>B</td>\n",
       "      <td>5</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.055</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>4op0</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>263</td>\n",
       "      <td>A</td>\n",
       "      <td>261</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.993781</td>\n",
       "      <td>0.004597</td>\n",
       "      <td>0.001520</td>\n",
       "      <td>0.593368</td>\n",
       "      <td>0.406408</td>\n",
       "      <td>0.754190</td>\n",
       "      <td>0.246046</td>\n",
       "      <td>8.557</td>\n",
       "      <td>1.043</td>\n",
       "      <td>1.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>4op0</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>264</td>\n",
       "      <td>A</td>\n",
       "      <td>262</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988971</td>\n",
       "      <td>0.006289</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.549427</td>\n",
       "      <td>0.450625</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.474</td>\n",
       "      <td>1.063</td>\n",
       "      <td>1.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>4op0</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>A</td>\n",
       "      <td>263</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980099</td>\n",
       "      <td>0.002661</td>\n",
       "      <td>0.017122</td>\n",
       "      <td>0.753667</td>\n",
       "      <td>0.246387</td>\n",
       "      <td>0.690637</td>\n",
       "      <td>0.309498</td>\n",
       "      <td>8.460</td>\n",
       "      <td>1.014</td>\n",
       "      <td>1.053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>4op0</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>266</td>\n",
       "      <td>A</td>\n",
       "      <td>264</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.866337</td>\n",
       "      <td>0.133508</td>\n",
       "      <td>8.301</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>4op0</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>267</td>\n",
       "      <td>A</td>\n",
       "      <td>265</td>\n",
       "      <td>?</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.215</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31725 rows × 174 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Residues                                                          \\\n",
       "         pdb label_entity_id label_asym_id label_seq_id auth_asym_id   \n",
       "0       5kwj               1             B            1            B   \n",
       "1       5kwj               1             B            2            B   \n",
       "2       5kwj               1             B            3            B   \n",
       "3       5kwj               1             B            4            B   \n",
       "4       5kwj               1             B            5            B   \n",
       "..       ...             ...           ...          ...          ...   \n",
       "254     4op0               1             A          263            A   \n",
       "255     4op0               1             A          264            A   \n",
       "256     4op0               1             A          265            A   \n",
       "257     4op0               1             A          266            A   \n",
       "258     4op0               1             A          267            A   \n",
       "\n",
       "                                  Label     Amino acids                  ...  \\\n",
       "    auth_seq_id pdbx_PDB_ins_code label label_comp_id_A label_comp_id_C  ...   \n",
       "0             0                 ?     0               0               0  ...   \n",
       "1             1                 ?     0               0               0  ...   \n",
       "2             2                 ?     0               0               0  ...   \n",
       "3             3                 ?     0               0               0  ...   \n",
       "4             4                 ?     0               0               0  ...   \n",
       "..          ...               ...   ...             ...             ...  ...   \n",
       "254         261                 ?     0               0               0  ...   \n",
       "255         262                 ?     0               0               0  ...   \n",
       "256         263                 ?     0               0               0  ...   \n",
       "257         264                 ?     0               0               0  ...   \n",
       "258         265                 ?     0               0               0  ...   \n",
       "\n",
       "      HHBlits                                                              \\\n",
       "         M->M      M->I      M->D      I->M      I->I      D->M      D->D   \n",
       "0    0.985549  0.000000  0.014548  0.000000  0.000000  0.000000  0.000000   \n",
       "1    1.000000  0.000000  0.000000  0.000000  0.000000  1.000000  0.000000   \n",
       "2    0.988286  0.011817  0.000000  1.000000  0.000000  0.000000  0.000000   \n",
       "3    0.995850  0.004370  0.000000  0.500000  0.500000  0.000000  0.000000   \n",
       "4    1.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "254  0.993781  0.004597  0.001520  0.593368  0.406408  0.754190  0.246046   \n",
       "255  0.988971  0.006289  0.004566  0.549427  0.450625  1.000000  0.000000   \n",
       "256  0.980099  0.002661  0.017122  0.753667  0.246387  0.690637  0.309498   \n",
       "257  1.000000  0.000000  0.000000  0.000000  0.000000  0.866337  0.133508   \n",
       "258  1.000000  0.000000  0.000000  0.000000  0.000000  1.000000  0.000000   \n",
       "\n",
       "                           \n",
       "       Neff Neff_I Neff_D  \n",
       "0    10.217  0.000  0.000  \n",
       "1    11.268  0.000  1.005  \n",
       "2    11.773  1.031  0.000  \n",
       "3    11.926  1.004  0.000  \n",
       "4    12.055  0.000  0.000  \n",
       "..      ...    ...    ...  \n",
       "254   8.557  1.043  1.019  \n",
       "255   8.474  1.063  1.013  \n",
       "256   8.460  1.014  1.053  \n",
       "257   8.301  0.000  1.189  \n",
       "258   8.215  0.000  1.016  \n",
       "\n",
       "[31725 rows x 174 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"../training_data/6.Training_sets/residues_test.pkl\", \"rb\") as f:\n",
    "    test_data = pickle.load(f)\n",
    "\n",
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "041a80cc-ad8f-4284-9baf-a510d2983ca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dataframe(df):\n",
    "    df.index = df[\"Residues\"][[\"pdb\",\"label_asym_id\", \"label_seq_id\"]].apply(lambda x: \"_\".join(x), axis=1)\n",
    "    df = df.drop(columns=[\"Residues\"], level=0)\n",
    "    df.columns = map(lambda x: \"_\".join(map(str, x)), df.columns.values)\n",
    "    df.loc[:,'Label_label'] = df['Label_label'].astype(\"category\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "50371457-11a5-436e-b211-ec871e2196f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label_label</th>\n",
       "      <th>Amino acids_label_comp_id_A</th>\n",
       "      <th>Amino acids_label_comp_id_C</th>\n",
       "      <th>Amino acids_label_comp_id_D</th>\n",
       "      <th>Amino acids_label_comp_id_E</th>\n",
       "      <th>Amino acids_label_comp_id_F</th>\n",
       "      <th>Amino acids_label_comp_id_G</th>\n",
       "      <th>Amino acids_label_comp_id_H</th>\n",
       "      <th>Amino acids_label_comp_id_I</th>\n",
       "      <th>Amino acids_label_comp_id_K</th>\n",
       "      <th>...</th>\n",
       "      <th>HHBlits_M-&gt;M</th>\n",
       "      <th>HHBlits_M-&gt;I</th>\n",
       "      <th>HHBlits_M-&gt;D</th>\n",
       "      <th>HHBlits_I-&gt;M</th>\n",
       "      <th>HHBlits_I-&gt;I</th>\n",
       "      <th>HHBlits_D-&gt;M</th>\n",
       "      <th>HHBlits_D-&gt;D</th>\n",
       "      <th>HHBlits_Neff</th>\n",
       "      <th>HHBlits_Neff_I</th>\n",
       "      <th>HHBlits_Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6ta3_B_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.998615</td>\n",
       "      <td>0.001223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.088</td>\n",
       "      <td>1.005</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.860</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997923</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.656106</td>\n",
       "      <td>0.343885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.988</td>\n",
       "      <td>1.007</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999307</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333325</td>\n",
       "      <td>0.666649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.047</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.026</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_452</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.906</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_453</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.787</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_454</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994470</td>\n",
       "      <td>0.005799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.713</td>\n",
       "      <td>1.002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_455</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.639</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_456</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.958</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108406 rows × 167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Label_label  Amino acids_label_comp_id_A  \\\n",
       "6ta3_B_1              0                            0   \n",
       "6ta3_B_2              0                            0   \n",
       "6ta3_B_3              0                            0   \n",
       "6ta3_B_4              0                            0   \n",
       "6ta3_B_5              0                            0   \n",
       "...                 ...                          ...   \n",
       "3d2p_A_452            0                            0   \n",
       "3d2p_A_453            0                            0   \n",
       "3d2p_A_454            0                            0   \n",
       "3d2p_A_455            0                            0   \n",
       "3d2p_A_456            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_C  Amino acids_label_comp_id_D  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              0                            0   \n",
       "6ta3_B_4                              1                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            0   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_E  Amino acids_label_comp_id_F  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              1                            0   \n",
       "6ta3_B_4                              0                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            0   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_G  Amino acids_label_comp_id_H  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              0                            0   \n",
       "6ta3_B_4                              0                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            1   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_I  Amino acids_label_comp_id_K  ...  \\\n",
       "6ta3_B_1                              0                            0  ...   \n",
       "6ta3_B_2                              0                            0  ...   \n",
       "6ta3_B_3                              0                            0  ...   \n",
       "6ta3_B_4                              0                            0  ...   \n",
       "6ta3_B_5                              1                            0  ...   \n",
       "...                                 ...                          ...  ...   \n",
       "3d2p_A_452                            0                            0  ...   \n",
       "3d2p_A_453                            0                            0  ...   \n",
       "3d2p_A_454                            0                            0  ...   \n",
       "3d2p_A_455                            0                            0  ...   \n",
       "3d2p_A_456                            0                            0  ...   \n",
       "\n",
       "            HHBlits_M->M  HHBlits_M->I  HHBlits_M->D  HHBlits_I->M  \\\n",
       "6ta3_B_1        0.998615      0.001223           0.0      1.000000   \n",
       "6ta3_B_2        1.000000      0.000000           0.0      0.000000   \n",
       "6ta3_B_3        0.997923      0.001963           0.0      0.656106   \n",
       "6ta3_B_4        0.999307      0.000605           0.0      0.333325   \n",
       "6ta3_B_5        1.000000      0.000000           0.0      0.000000   \n",
       "...                  ...           ...           ...           ...   \n",
       "3d2p_A_452      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_453      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_454      0.994470      0.005799           0.0      1.000000   \n",
       "3d2p_A_455      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_456      1.000000      0.000000           0.0      1.000000   \n",
       "\n",
       "            HHBlits_I->I  HHBlits_D->M  HHBlits_D->D  HHBlits_Neff  \\\n",
       "6ta3_B_1        0.000000           0.0           0.0         9.088   \n",
       "6ta3_B_2        0.000000           0.0           0.0         9.860   \n",
       "6ta3_B_3        0.343885           0.0           0.0         9.988   \n",
       "6ta3_B_4        0.666649           0.0           0.0        10.047   \n",
       "6ta3_B_5        0.000000           0.0           0.0        10.026   \n",
       "...                  ...           ...           ...           ...   \n",
       "3d2p_A_452      0.000000           0.0           0.0        10.906   \n",
       "3d2p_A_453      0.000000           0.0           0.0        10.787   \n",
       "3d2p_A_454      0.000000           0.0           0.0        10.713   \n",
       "3d2p_A_455      0.000000           0.0           0.0        10.639   \n",
       "3d2p_A_456      0.000000           0.0           0.0        10.958   \n",
       "\n",
       "            HHBlits_Neff_I  HHBlits_Neff_D  \n",
       "6ta3_B_1             1.005             0.0  \n",
       "6ta3_B_2             0.000             0.0  \n",
       "6ta3_B_3             1.007             0.0  \n",
       "6ta3_B_4             1.000             0.0  \n",
       "6ta3_B_5             0.000             0.0  \n",
       "...                    ...             ...  \n",
       "3d2p_A_452           0.000             0.0  \n",
       "3d2p_A_453           0.000             0.0  \n",
       "3d2p_A_454           1.002             0.0  \n",
       "3d2p_A_455           0.000             0.0  \n",
       "3d2p_A_456           0.000             0.0  \n",
       "\n",
       "[108406 rows x 167 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train = process_dataframe(train_data)\n",
    "\n",
    "train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6efed703-ac2b-40aa-8be9-cab00d22673f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label_label</th>\n",
       "      <th>Amino acids_label_comp_id_A</th>\n",
       "      <th>Amino acids_label_comp_id_C</th>\n",
       "      <th>Amino acids_label_comp_id_D</th>\n",
       "      <th>Amino acids_label_comp_id_E</th>\n",
       "      <th>Amino acids_label_comp_id_F</th>\n",
       "      <th>Amino acids_label_comp_id_G</th>\n",
       "      <th>Amino acids_label_comp_id_H</th>\n",
       "      <th>Amino acids_label_comp_id_I</th>\n",
       "      <th>Amino acids_label_comp_id_K</th>\n",
       "      <th>...</th>\n",
       "      <th>HHBlits_M-&gt;M</th>\n",
       "      <th>HHBlits_M-&gt;I</th>\n",
       "      <th>HHBlits_M-&gt;D</th>\n",
       "      <th>HHBlits_I-&gt;M</th>\n",
       "      <th>HHBlits_I-&gt;I</th>\n",
       "      <th>HHBlits_D-&gt;M</th>\n",
       "      <th>HHBlits_D-&gt;D</th>\n",
       "      <th>HHBlits_Neff</th>\n",
       "      <th>HHBlits_Neff_I</th>\n",
       "      <th>HHBlits_Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5kwj_B_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.985549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014548</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.217</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5kwj_B_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.268</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5kwj_B_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988286</td>\n",
       "      <td>0.011817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.773</td>\n",
       "      <td>1.031</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5kwj_B_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.995850</td>\n",
       "      <td>0.004370</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.926</td>\n",
       "      <td>1.004</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5kwj_B_5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.055</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4op0_A_263</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.993781</td>\n",
       "      <td>0.004597</td>\n",
       "      <td>0.001520</td>\n",
       "      <td>0.593368</td>\n",
       "      <td>0.406408</td>\n",
       "      <td>0.754190</td>\n",
       "      <td>0.246046</td>\n",
       "      <td>8.557</td>\n",
       "      <td>1.043</td>\n",
       "      <td>1.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4op0_A_264</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.988971</td>\n",
       "      <td>0.006289</td>\n",
       "      <td>0.004566</td>\n",
       "      <td>0.549427</td>\n",
       "      <td>0.450625</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.474</td>\n",
       "      <td>1.063</td>\n",
       "      <td>1.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4op0_A_265</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.980099</td>\n",
       "      <td>0.002661</td>\n",
       "      <td>0.017122</td>\n",
       "      <td>0.753667</td>\n",
       "      <td>0.246387</td>\n",
       "      <td>0.690637</td>\n",
       "      <td>0.309498</td>\n",
       "      <td>8.460</td>\n",
       "      <td>1.014</td>\n",
       "      <td>1.053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4op0_A_266</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.866337</td>\n",
       "      <td>0.133508</td>\n",
       "      <td>8.301</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4op0_A_267</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.215</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31725 rows × 167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Label_label  Amino acids_label_comp_id_A  \\\n",
       "5kwj_B_1              0                            0   \n",
       "5kwj_B_2              0                            0   \n",
       "5kwj_B_3              0                            0   \n",
       "5kwj_B_4              0                            0   \n",
       "5kwj_B_5              0                            0   \n",
       "...                 ...                          ...   \n",
       "4op0_A_263            0                            0   \n",
       "4op0_A_264            0                            0   \n",
       "4op0_A_265            0                            0   \n",
       "4op0_A_266            0                            0   \n",
       "4op0_A_267            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_C  Amino acids_label_comp_id_D  \\\n",
       "5kwj_B_1                              0                            0   \n",
       "5kwj_B_2                              0                            0   \n",
       "5kwj_B_3                              0                            0   \n",
       "5kwj_B_4                              0                            0   \n",
       "5kwj_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "4op0_A_263                            0                            1   \n",
       "4op0_A_264                            0                            0   \n",
       "4op0_A_265                            0                            0   \n",
       "4op0_A_266                            0                            0   \n",
       "4op0_A_267                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_E  Amino acids_label_comp_id_F  \\\n",
       "5kwj_B_1                              0                            0   \n",
       "5kwj_B_2                              0                            1   \n",
       "5kwj_B_3                              0                            0   \n",
       "5kwj_B_4                              0                            0   \n",
       "5kwj_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "4op0_A_263                            0                            0   \n",
       "4op0_A_264                            0                            0   \n",
       "4op0_A_265                            0                            0   \n",
       "4op0_A_266                            0                            0   \n",
       "4op0_A_267                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_G  Amino acids_label_comp_id_H  \\\n",
       "5kwj_B_1                              0                            0   \n",
       "5kwj_B_2                              0                            0   \n",
       "5kwj_B_3                              0                            0   \n",
       "5kwj_B_4                              0                            0   \n",
       "5kwj_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "4op0_A_263                            0                            0   \n",
       "4op0_A_264                            0                            0   \n",
       "4op0_A_265                            0                            0   \n",
       "4op0_A_266                            0                            1   \n",
       "4op0_A_267                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_I  Amino acids_label_comp_id_K  ...  \\\n",
       "5kwj_B_1                              0                            0  ...   \n",
       "5kwj_B_2                              0                            0  ...   \n",
       "5kwj_B_3                              0                            0  ...   \n",
       "5kwj_B_4                              0                            0  ...   \n",
       "5kwj_B_5                              0                            0  ...   \n",
       "...                                 ...                          ...  ...   \n",
       "4op0_A_263                            0                            0  ...   \n",
       "4op0_A_264                            0                            0  ...   \n",
       "4op0_A_265                            0                            0  ...   \n",
       "4op0_A_266                            0                            0  ...   \n",
       "4op0_A_267                            0                            0  ...   \n",
       "\n",
       "            HHBlits_M->M  HHBlits_M->I  HHBlits_M->D  HHBlits_I->M  \\\n",
       "5kwj_B_1        0.985549      0.000000      0.014548      0.000000   \n",
       "5kwj_B_2        1.000000      0.000000      0.000000      0.000000   \n",
       "5kwj_B_3        0.988286      0.011817      0.000000      1.000000   \n",
       "5kwj_B_4        0.995850      0.004370      0.000000      0.500000   \n",
       "5kwj_B_5        1.000000      0.000000      0.000000      0.000000   \n",
       "...                  ...           ...           ...           ...   \n",
       "4op0_A_263      0.993781      0.004597      0.001520      0.593368   \n",
       "4op0_A_264      0.988971      0.006289      0.004566      0.549427   \n",
       "4op0_A_265      0.980099      0.002661      0.017122      0.753667   \n",
       "4op0_A_266      1.000000      0.000000      0.000000      0.000000   \n",
       "4op0_A_267      1.000000      0.000000      0.000000      0.000000   \n",
       "\n",
       "            HHBlits_I->I  HHBlits_D->M  HHBlits_D->D  HHBlits_Neff  \\\n",
       "5kwj_B_1        0.000000      0.000000      0.000000        10.217   \n",
       "5kwj_B_2        0.000000      1.000000      0.000000        11.268   \n",
       "5kwj_B_3        0.000000      0.000000      0.000000        11.773   \n",
       "5kwj_B_4        0.500000      0.000000      0.000000        11.926   \n",
       "5kwj_B_5        0.000000      0.000000      0.000000        12.055   \n",
       "...                  ...           ...           ...           ...   \n",
       "4op0_A_263      0.406408      0.754190      0.246046         8.557   \n",
       "4op0_A_264      0.450625      1.000000      0.000000         8.474   \n",
       "4op0_A_265      0.246387      0.690637      0.309498         8.460   \n",
       "4op0_A_266      0.000000      0.866337      0.133508         8.301   \n",
       "4op0_A_267      0.000000      1.000000      0.000000         8.215   \n",
       "\n",
       "            HHBlits_Neff_I  HHBlits_Neff_D  \n",
       "5kwj_B_1             0.000           0.000  \n",
       "5kwj_B_2             0.000           1.005  \n",
       "5kwj_B_3             1.031           0.000  \n",
       "5kwj_B_4             1.004           0.000  \n",
       "5kwj_B_5             0.000           0.000  \n",
       "...                    ...             ...  \n",
       "4op0_A_263           1.043           1.019  \n",
       "4op0_A_264           1.063           1.013  \n",
       "4op0_A_265           1.014           1.053  \n",
       "4op0_A_266           0.000           1.189  \n",
       "4op0_A_267           0.000           1.016  \n",
       "\n",
       "[31725 rows x 167 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = process_dataframe(test_data)\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e145ba01-749d-4a97-9abe-37b0a553ea91",
   "metadata": {},
   "source": [
    "# Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a55639c0-f74a-4ac4-91b9-1bec52ff3912",
   "metadata": {},
   "outputs": [],
   "source": [
    "from autogluon.tabular import TabularDataset, TabularPredictor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8307aca4-a31c-423c-b48d-453a1df8cc31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label_label</th>\n",
       "      <th>Amino acids_label_comp_id_A</th>\n",
       "      <th>Amino acids_label_comp_id_C</th>\n",
       "      <th>Amino acids_label_comp_id_D</th>\n",
       "      <th>Amino acids_label_comp_id_E</th>\n",
       "      <th>Amino acids_label_comp_id_F</th>\n",
       "      <th>Amino acids_label_comp_id_G</th>\n",
       "      <th>Amino acids_label_comp_id_H</th>\n",
       "      <th>Amino acids_label_comp_id_I</th>\n",
       "      <th>Amino acids_label_comp_id_K</th>\n",
       "      <th>...</th>\n",
       "      <th>HHBlits_M-&gt;M</th>\n",
       "      <th>HHBlits_M-&gt;I</th>\n",
       "      <th>HHBlits_M-&gt;D</th>\n",
       "      <th>HHBlits_I-&gt;M</th>\n",
       "      <th>HHBlits_I-&gt;I</th>\n",
       "      <th>HHBlits_D-&gt;M</th>\n",
       "      <th>HHBlits_D-&gt;D</th>\n",
       "      <th>HHBlits_Neff</th>\n",
       "      <th>HHBlits_Neff_I</th>\n",
       "      <th>HHBlits_Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6ta3_B_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.998615</td>\n",
       "      <td>0.001223</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.088</td>\n",
       "      <td>1.005</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.860</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997923</td>\n",
       "      <td>0.001963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.656106</td>\n",
       "      <td>0.343885</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.988</td>\n",
       "      <td>1.007</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999307</td>\n",
       "      <td>0.000605</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333325</td>\n",
       "      <td>0.666649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.047</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6ta3_B_5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.026</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_452</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.906</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_453</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.787</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_454</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.994470</td>\n",
       "      <td>0.005799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.713</td>\n",
       "      <td>1.002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_455</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.639</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3d2p_A_456</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.958</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>108406 rows × 167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Label_label  Amino acids_label_comp_id_A  \\\n",
       "6ta3_B_1              0                            0   \n",
       "6ta3_B_2              0                            0   \n",
       "6ta3_B_3              0                            0   \n",
       "6ta3_B_4              0                            0   \n",
       "6ta3_B_5              0                            0   \n",
       "...                 ...                          ...   \n",
       "3d2p_A_452            0                            0   \n",
       "3d2p_A_453            0                            0   \n",
       "3d2p_A_454            0                            0   \n",
       "3d2p_A_455            0                            0   \n",
       "3d2p_A_456            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_C  Amino acids_label_comp_id_D  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              0                            0   \n",
       "6ta3_B_4                              1                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            0   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_E  Amino acids_label_comp_id_F  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              1                            0   \n",
       "6ta3_B_4                              0                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            0   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_G  Amino acids_label_comp_id_H  \\\n",
       "6ta3_B_1                              0                            0   \n",
       "6ta3_B_2                              0                            0   \n",
       "6ta3_B_3                              0                            0   \n",
       "6ta3_B_4                              0                            0   \n",
       "6ta3_B_5                              0                            0   \n",
       "...                                 ...                          ...   \n",
       "3d2p_A_452                            0                            0   \n",
       "3d2p_A_453                            0                            0   \n",
       "3d2p_A_454                            0                            0   \n",
       "3d2p_A_455                            0                            1   \n",
       "3d2p_A_456                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_I  Amino acids_label_comp_id_K  ...  \\\n",
       "6ta3_B_1                              0                            0  ...   \n",
       "6ta3_B_2                              0                            0  ...   \n",
       "6ta3_B_3                              0                            0  ...   \n",
       "6ta3_B_4                              0                            0  ...   \n",
       "6ta3_B_5                              1                            0  ...   \n",
       "...                                 ...                          ...  ...   \n",
       "3d2p_A_452                            0                            0  ...   \n",
       "3d2p_A_453                            0                            0  ...   \n",
       "3d2p_A_454                            0                            0  ...   \n",
       "3d2p_A_455                            0                            0  ...   \n",
       "3d2p_A_456                            0                            0  ...   \n",
       "\n",
       "            HHBlits_M->M  HHBlits_M->I  HHBlits_M->D  HHBlits_I->M  \\\n",
       "6ta3_B_1        0.998615      0.001223           0.0      1.000000   \n",
       "6ta3_B_2        1.000000      0.000000           0.0      0.000000   \n",
       "6ta3_B_3        0.997923      0.001963           0.0      0.656106   \n",
       "6ta3_B_4        0.999307      0.000605           0.0      0.333325   \n",
       "6ta3_B_5        1.000000      0.000000           0.0      0.000000   \n",
       "...                  ...           ...           ...           ...   \n",
       "3d2p_A_452      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_453      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_454      0.994470      0.005799           0.0      1.000000   \n",
       "3d2p_A_455      1.000000      0.000000           0.0      0.000000   \n",
       "3d2p_A_456      1.000000      0.000000           0.0      1.000000   \n",
       "\n",
       "            HHBlits_I->I  HHBlits_D->M  HHBlits_D->D  HHBlits_Neff  \\\n",
       "6ta3_B_1        0.000000           0.0           0.0         9.088   \n",
       "6ta3_B_2        0.000000           0.0           0.0         9.860   \n",
       "6ta3_B_3        0.343885           0.0           0.0         9.988   \n",
       "6ta3_B_4        0.666649           0.0           0.0        10.047   \n",
       "6ta3_B_5        0.000000           0.0           0.0        10.026   \n",
       "...                  ...           ...           ...           ...   \n",
       "3d2p_A_452      0.000000           0.0           0.0        10.906   \n",
       "3d2p_A_453      0.000000           0.0           0.0        10.787   \n",
       "3d2p_A_454      0.000000           0.0           0.0        10.713   \n",
       "3d2p_A_455      0.000000           0.0           0.0        10.639   \n",
       "3d2p_A_456      0.000000           0.0           0.0        10.958   \n",
       "\n",
       "            HHBlits_Neff_I  HHBlits_Neff_D  \n",
       "6ta3_B_1             1.005             0.0  \n",
       "6ta3_B_2             0.000             0.0  \n",
       "6ta3_B_3             1.007             0.0  \n",
       "6ta3_B_4             1.000             0.0  \n",
       "6ta3_B_5             0.000             0.0  \n",
       "...                    ...             ...  \n",
       "3d2p_A_452           0.000             0.0  \n",
       "3d2p_A_453           0.000             0.0  \n",
       "3d2p_A_454           1.002             0.0  \n",
       "3d2p_A_455           0.000             0.0  \n",
       "3d2p_A_456           0.000             0.0  \n",
       "\n",
       "[108406 rows x 167 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traindataset = TabularDataset(train)\n",
    "\n",
    "traindataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6875f7db-ab9d-4319-8ef8-bc5f1eba6e88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index([], dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "traindataset.columns[ traindataset.isna().any() ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f07f1fd2-a622-44b6-b9e3-a408eb1a319c",
   "metadata": {},
   "outputs": [],
   "source": [
    "testdataset = TabularDataset(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "08ca8160-fb27-44bf-8011-52bb81b2c8ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import matthews_corrcoef"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6bcc575e-0cdf-454b-ac00-8a3ff0aafbc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from autogluon.core.metrics import make_scorer\n",
    "\n",
    "mcc = make_scorer(\n",
    "    name='mcc',\n",
    "    score_func=matthews_corrcoef,\n",
    "    optimum=1,\n",
    "    greater_is_better=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c251f2d6-42d9-4c93-a1ef-2feb1e58e18e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class TabularPredictor in module autogluon.tabular.predictor.predictor:\n",
      "\n",
      "class TabularPredictor(autogluon.tabular.predictor._deprecated_methods.TabularPredictorDeprecatedMixin)\n",
      " |  TabularPredictor(label: 'str', problem_type: 'str' = None, eval_metric: 'str | Scorer' = None, path: 'str' = None, verbosity: 'int' = 2, log_to_file: 'bool' = False, log_file_path: 'str' = 'auto', sample_weight: 'str' = None, weight_evaluation: 'bool' = False, groups: 'str' = None, positive_class: 'int | str | None' = None, **kwargs)\n",
      " |  \n",
      " |  AutoGluon TabularPredictor predicts values in a column of a tabular dataset (classification or regression).\n",
      " |  \n",
      " |  Parameters\n",
      " |  ----------\n",
      " |  label : str\n",
      " |      Name of the column that contains the target variable to predict.\n",
      " |  problem_type : str, default = None\n",
      " |      Type of prediction problem, i.e. is this a binary/multiclass classification or regression problem (options: 'binary', 'multiclass', 'regression', 'quantile').\n",
      " |      If `problem_type = None`, the prediction problem type is inferred based on the label-values in provided dataset.\n",
      " |  eval_metric : str or Scorer, default = None\n",
      " |      Metric by which predictions will be ultimately evaluated on test data.\n",
      " |      AutoGluon tunes factors such as hyperparameters, early-stopping, ensemble-weights, etc. in order to improve this metric on validation data.\n",
      " |  \n",
      " |      If `eval_metric = None`, it is automatically chosen based on `problem_type`.\n",
      " |      Defaults to 'accuracy' for binary and multiclass classification, 'root_mean_squared_error' for regression, and 'pinball_loss' for quantile.\n",
      " |  \n",
      " |      Otherwise, options for classification:\n",
      " |          ['accuracy', 'balanced_accuracy', 'log_loss', 'f1', 'f1_macro', 'f1_micro', 'f1_weighted',\n",
      " |          'roc_auc', 'roc_auc_ovo', 'roc_auc_ovo_macro', 'roc_auc_ovo_weighted', 'roc_auc_ovr', 'roc_auc_ovr_macro', 'roc_auc_ovr_micro',\n",
      " |          'roc_auc_ovr_weighted', 'average_precision', 'precision', 'precision_macro', 'precision_micro', 'precision_weighted',\n",
      " |          'recall', 'recall_macro', 'recall_micro', 'recall_weighted', 'mcc', 'pac_score']\n",
      " |      Options for regression:\n",
      " |          ['root_mean_squared_error', 'mean_squared_error', 'mean_absolute_error', 'median_absolute_error', 'mean_absolute_percentage_error', 'r2', 'symmetric_mean_absolute_percentage_error']\n",
      " |      For more information on these options, see `sklearn.metrics`: https://scikit-learn.org/stable/modules/classes.html#sklearn-metrics-metrics\n",
      " |      For metric source code, see `autogluon.core.metrics`.\n",
      " |  \n",
      " |      You can also pass your own evaluation function here as long as it follows formatting of the functions defined in folder `autogluon.core.metrics`.\n",
      " |      For detailed instructions on creating and using a custom metric, refer to https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-metric.html\n",
      " |  path : Union[str, pathlib.Path], default = None\n",
      " |      Path to directory where models and intermediate outputs should be saved.\n",
      " |      If unspecified, a time-stamped folder called \"AutogluonModels/ag-[TIMESTAMP]\" will be created in the working directory to store all models.\n",
      " |      Note: To call `fit()` twice and save all results of each fit, you must specify different `path` locations or don't specify `path` at all.\n",
      " |      Otherwise files from first `fit()` will be overwritten by second `fit()`.\n",
      " |  verbosity : int, default = 2\n",
      " |      Verbosity levels range from 0 to 4 and control how much information is printed.\n",
      " |      Higher levels correspond to more detailed print statements (you can set verbosity = 0 to suppress warnings).\n",
      " |      If using logging, you can alternatively control amount of information printed via `logger.setLevel(L)`,\n",
      " |      where `L` ranges from 0 to 50 (Note: higher values of `L` correspond to fewer print statements, opposite of verbosity levels).\n",
      " |      Verbosity levels:\n",
      " |          0: Only log exceptions\n",
      " |          1: Only log warnings + exceptions\n",
      " |          2: Standard logging\n",
      " |          3: Verbose logging (ex: log validation score every 50 iterations)\n",
      " |          4: Maximally verbose logging (ex: log validation score every iteration)\n",
      " |  log_to_file: bool, default = False\n",
      " |      Whether to save the logs into a file for later reference\n",
      " |  log_file_path: str, default = \"auto\"\n",
      " |      File path to save the logs.\n",
      " |      If auto, logs will be saved under `predictor_path/logs/predictor_log.txt`.\n",
      " |      Will be ignored if `log_to_file` is set to False\n",
      " |  sample_weight : str, default = None\n",
      " |      If specified, this column-name indicates which column of the data should be treated as sample weights. This column will NOT be considered as a predictive feature.\n",
      " |      Sample weights should be non-negative (and cannot be nan), with larger values indicating which rows are more important than others.\n",
      " |      If you want your usage of sample weights to match results obtained outside of this Predictor, then ensure sample weights for your training (or tuning) data sum to the number of rows in the training (or tuning) data.\n",
      " |      You may also specify two special strings: 'auto_weight' (automatically choose a weighting strategy based on the data) or 'balance_weight' (equally weight classes in classification, no effect in regression). If specifying your own sample_weight column, make sure its name does not match these special strings.\n",
      " |  weight_evaluation : bool, default = False\n",
      " |      Only considered when `sample_weight` column is not None. Determines whether sample weights should be taken into account when computing evaluation metrics on validation/test data.\n",
      " |      If True, then weighted metrics will be reported based on the sample weights provided in the specified `sample_weight` (in which case `sample_weight` column must also be present in test data).\n",
      " |      In this case, the 'best' model used by default for prediction will also be decided based on a weighted version of evaluation metric.\n",
      " |      Note: we do not recommend specifying `weight_evaluation` when `sample_weight` is 'auto_weight' or 'balance_weight', instead specify appropriate `eval_metric`.\n",
      " |  groups : str, default = None\n",
      " |      [Experimental] If specified, AutoGluon will use the column named the value of groups in `train_data` during `.fit` as the data splitting indices for the purposes of bagging.\n",
      " |      This column will not be used as a feature during model training.\n",
      " |      This parameter is ignored if bagging is not enabled. To instead specify a custom validation set with bagging disabled, specify `tuning_data` in `.fit`.\n",
      " |      The data will be split via `sklearn.model_selection.LeaveOneGroupOut`.\n",
      " |      Use this option to control the exact split indices AutoGluon uses.\n",
      " |      It is not recommended to use this option unless it is required for very specific situations.\n",
      " |      Bugs may arise from edge cases if the provided groups are not valid to properly train models, such as if not all classes are present during training in multiclass classification. It is up to the user to sanitize their groups.\n",
      " |  \n",
      " |      As an example, if you want your data folds to preserve adjacent rows in the table without shuffling, then for 3 fold bagging with 6 rows of data, the groups column values should be [0, 0, 1, 1, 2, 2].\n",
      " |  positive_class : str or int, default = None\n",
      " |      Used to determine the positive class in binary classification.\n",
      " |      This is used for certain metrics such as 'f1' which produce different scores depending on which class is considered the positive class.\n",
      " |      If not set, will be inferred as the second element of the existing unique classes after sorting them.\n",
      " |          If classes are [0, 1], then 1 will be selected as the positive class.\n",
      " |          If classes are ['def', 'abc'], then 'def' will be selected as the positive class.\n",
      " |          If classes are [True, False], then True will be selected as the positive class.\n",
      " |  **kwargs :\n",
      " |      learner_type : AbstractLearner, default = DefaultLearner\n",
      " |          A class which inherits from `AbstractLearner`. This dictates the inner logic of predictor.\n",
      " |          If you don't know what this is, keep it as the default.\n",
      " |      learner_kwargs : dict, default = None\n",
      " |          Kwargs to send to the learner. Options include:\n",
      " |  \n",
      " |          ignored_columns : list, default = None\n",
      " |              Banned subset of column names that predictor may not use as predictive features (e.g. unique identifier to a row or user-ID).\n",
      " |              These columns are ignored during `fit()`.\n",
      " |          label_count_threshold : int, default = 10\n",
      " |              For multi-class classification problems, this is the minimum number of times a label must appear in dataset in order to be considered an output class.\n",
      " |              AutoGluon will ignore any classes whose labels do not appear at least this many times in the dataset (i.e. will never predict them).\n",
      " |          cache_data : bool, default = True\n",
      " |              When enabled, the training and validation data are saved to disk for future reuse.\n",
      " |              Enables advanced functionality in predictor such as `fit_extra()` and feature importance calculation on the original data.\n",
      " |          trainer_type : AbstractTrainer, default = AutoTrainer\n",
      " |              A class inheriting from `AbstractTrainer` that controls training/ensembling of many models.\n",
      " |              If you don't know what this is, keep it as the default.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      TabularPredictor\n",
      " |      autogluon.tabular.predictor._deprecated_methods.TabularPredictorDeprecatedMixin\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, label: 'str', problem_type: 'str' = None, eval_metric: 'str | Scorer' = None, path: 'str' = None, verbosity: 'int' = 2, log_to_file: 'bool' = False, log_file_path: 'str' = 'auto', sample_weight: 'str' = None, weight_evaluation: 'bool' = False, groups: 'str' = None, positive_class: 'int | str | None' = None, **kwargs)\n",
      " |      Initialize self.  See help(type(self)) for accurate signature.\n",
      " |  \n",
      " |  calibrate_decision_threshold(self, data: 'pd.DataFrame | str | None' = None, metric: 'str | Scorer | None' = None, model: 'str' = 'best', decision_thresholds: 'int | List[float]' = 25, secondary_decision_thresholds: 'int | None' = 19, subsample_size: 'int | None' = 1000000, verbose: 'bool' = True) -> 'float'\n",
      " |      Calibrate the decision threshold in binary classification to optimize a given metric.\n",
      " |      You can pass the output of this method as input to `predictor.set_decision_threshold` to update the predictor.\n",
      " |      Will raise an AssertionError if `predictor.problem_type != 'binary'`.\n",
      " |      \n",
      " |      Note that while calibrating the decision threshold can help to improve a given metric,\n",
      " |      other metrics may end up having worse scores.\n",
      " |      For example, calibrating on `balanced_accuracy` will often harm `accuracy`.\n",
      " |      Users should keep this in mind while leveraging decision threshold calibration.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : pd.DataFrame or str, optional\n",
      " |          The data to use for calibration. Must contain the label column.\n",
      " |          We recommend to keep this value as None unless you are an advanced user and understand the implications.\n",
      " |          If None, will use internal data such as the holdout validation data or out-of-fold predictions.\n",
      " |      metric : autogluon.core.metrics.Scorer or str, default = None\n",
      " |          The metric to optimize during calibration.\n",
      " |          If None, uses `predictor.eval_metric`.\n",
      " |      model : str, default = 'best'\n",
      " |          The model to use prediction probabilities of when calibrating the threshold.\n",
      " |          If 'best', will use `predictor.model_best`.\n",
      " |      decision_thresholds : int | List[float], default = 25\n",
      " |          The number of decision thresholds on either side of `0.5` to search.\n",
      " |          The default of 25 will result in 51 searched thresholds: [0.00, 0.02, 0.04, ..., 0.48, 0.50, 0.52, ..., 0.96, 0.98, 1.00]\n",
      " |          Alternatively, a list of decision thresholds can be passed and only the thresholds in the list will be searched.\n",
      " |      secondary_decision_thresholds : int | None, default = 19\n",
      " |          The number of secondary decision thresholds to check on either side of the threshold identified in the first phase.\n",
      " |          Skipped if None.\n",
      " |          For example, if decision_thresholds=50 and 0.14 was identified as the optimal threshold, while secondary_decision_threshold=9,\n",
      " |              Then the following additional thresholds are checked:\n",
      " |                  [0.131, 0.132, 0.133, 0.134, 0.135, 0.136, 0.137, 0.138, 0.139, 0.141, 0.142, 0.143, 0.144, 0.145, 0.146, 0.147, 0.148, 0.149]\n",
      " |      subsample_size : int | None, default = 1000000\n",
      " |          When `subsample_size` is not None and `data` contains more rows than `subsample_size`, samples to `subsample_size` rows to speed up calibration.\n",
      " |          Usually it is not necessary to use more than 1 million rows for calibration.\n",
      " |      verbose : bool, default = True\n",
      " |          If True, will log information about the calibration process.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Decision Threshold: A float between 0 and 1 defining the decision boundary for predictions that\n",
      " |      maximizes the `metric` score on the `data` for the `model`.\n",
      " |  \n",
      " |  clone(self, path: 'str', *, return_clone: 'bool' = False, dirs_exist_ok: 'bool' = False) -> \"str | 'TabularPredictor'\"\n",
      " |      Clone the predictor and all of its artifacts to a new location on local disk.\n",
      " |      This is ideal for use-cases where saving a snapshot of the predictor is desired before performing\n",
      " |      more advanced operations (such as fit_extra and refit_full).\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      path : str\n",
      " |          Directory path the cloned predictor will be saved to.\n",
      " |      return_clone : bool, default = False\n",
      " |          If True, returns the loaded cloned TabularPredictor object.\n",
      " |          If False, returns the local path to the cloned TabularPredictor object.\n",
      " |      dirs_exist_ok : bool, default = False\n",
      " |          If True, will clone the predictor even if the path directory already exists, potentially overwriting unrelated files.\n",
      " |          If False, will raise an exception if the path directory already exists and avoid performing the copy.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      If return_clone == True, returns the loaded cloned TabularPredictor object.\n",
      " |      If return_clone == False, returns the local path to the cloned TabularPredictor object.\n",
      " |  \n",
      " |  clone_for_deployment(self, path: 'str', *, model: 'str' = 'best', return_clone: 'bool' = False, dirs_exist_ok: 'bool' = False) -> \"str | 'TabularPredictor'\"\n",
      " |      Clone the predictor and all of its artifacts to a new location on local disk,\n",
      " |      then delete the clones artifacts unnecessary during prediction.\n",
      " |      This is ideal for use-cases where saving a snapshot of the predictor is desired before performing\n",
      " |      more advanced operations (such as fit_extra and refit_full).\n",
      " |      \n",
      " |      Note that the clone can no longer fit new models,\n",
      " |      and most functionality except for predict and predict_proba will no longer work.\n",
      " |      \n",
      " |      Identical to performing the following operations in order:\n",
      " |      \n",
      " |      predictor_clone = predictor.clone(path=path, return_clone=True, dirs_exist_ok=dirs_exist_ok)\n",
      " |      predictor_clone.delete_models(models_to_keep=model, dry_run=False)\n",
      " |      predictor_clone.set_model_best(model=model, save_trainer=True)\n",
      " |      predictor_clone.save_space()\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      path : str\n",
      " |          Directory path the cloned predictor will be saved to.\n",
      " |      model : str, default = 'best'\n",
      " |          The model to use in the optimized predictor clone.\n",
      " |          All other unrelated models will be deleted to save disk space.\n",
      " |          Refer to the `models_to_keep` argument of `predictor.delete_models` for available options.\n",
      " |          Internally calls `predictor_clone.delete_models(models_to_keep=model, dry_run=False)`\n",
      " |      return_clone : bool, default = False\n",
      " |          If True, returns the loaded cloned TabularPredictor object.\n",
      " |          If False, returns the local path to the cloned TabularPredictor object.\n",
      " |      dirs_exist_ok : bool, default = False\n",
      " |          If True, will clone the predictor even if the path directory already exists, potentially overwriting unrelated files.\n",
      " |          If False, will raise an exception if the path directory already exists and avoids performing the copy.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      If return_clone == True, returns the loaded cloned TabularPredictor object.\n",
      " |      If return_clone == False, returns the local path to the cloned TabularPredictor object.\n",
      " |  \n",
      " |  compile(self, models='best', with_ancestors=True, compiler_configs='auto')\n",
      " |      Compile models for accelerated prediction.\n",
      " |      This can be helpful to reduce prediction latency and improve throughput.\n",
      " |      \n",
      " |      Note that this is currently an experimental feature, the supported compilers can be ['native', 'onnx'].\n",
      " |      \n",
      " |      In order to compile with a specific compiler, that compiler must be installed in the Python environment.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      models : list of str or str, default = 'best'\n",
      " |          Model names of models to compile.\n",
      " |          If 'best' then the model with the highest validation score is compiled (this is the model used for prediction by default).\n",
      " |          If 'all' then all models are compiled.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |      with_ancestors : bool, default = True\n",
      " |          If True, all ancestor models of the provided models will also be compiled.\n",
      " |      compiler_configs : dict or str, default = \"auto\"\n",
      " |          If \"auto\", defaults to the following:\n",
      " |              compiler_configs = {\n",
      " |                  \"RF\": {\"compiler\": \"onnx\"},\n",
      " |                  \"XT\": {\"compiler\": \"onnx\"},\n",
      " |                  \"NN_TORCH\": {\"compiler\": \"onnx\"},\n",
      " |              }\n",
      " |          Otherwise, specify a compiler_configs dictionary manually. Keys can be exact model names or model types.\n",
      " |          Exact model names take priority over types if both are valid for a model.\n",
      " |          Types can be either the true type such as RandomForestModel or the shorthand \"RF\".\n",
      " |          The dictionary key logic for types is identical to the logic in the hyperparameters argument of `predictor.fit`\n",
      " |      \n",
      " |          Example values within the configs:\n",
      " |              compiler : str, default = None\n",
      " |                  The compiler that is used for model compilation.\n",
      " |              batch_size : int, default = None\n",
      " |                  The batch size that is optimized for model prediction.\n",
      " |                  By default, the batch size is None. This means the compiler would try to leverage dynamic shape for prediction.\n",
      " |                  Using batch_size=1 would be more suitable for online prediction, which expects a result from one data point.\n",
      " |                  However, it can be slow for batch processing, because of the overhead of multiple kernel execution.\n",
      " |                  Increasing batch size to a number that is larger than 1 would help increase the prediction throughput.\n",
      " |                  This comes with an expense of utilizing larger memory for prediction.\n",
      " |  \n",
      " |  delete_models(self, models_to_keep=None, models_to_delete=None, allow_delete_cascade=False, delete_from_disk=True, dry_run=True)\n",
      " |      Deletes models from `predictor`.\n",
      " |      This can be helpful to minimize memory usage and disk usage, particularly for model deployment.\n",
      " |      This will remove all references to the models in `predictor`.\n",
      " |          For example, removed models will not appear in `predictor.leaderboard()`.\n",
      " |      WARNING: If `delete_from_disk=True`, this will DELETE ALL FILES in the deleted model directories, regardless if they were created by AutoGluon or not.\n",
      " |          DO NOT STORE FILES INSIDE OF THE MODEL DIRECTORY THAT ARE UNRELATED TO AUTOGLUON.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      models_to_keep : str or list, default = None\n",
      " |          Name of model or models to not delete.\n",
      " |          All models that are not specified and are also not required as a dependency of any model in `models_to_keep` will be deleted.\n",
      " |          Specify `models_to_keep='best'` to keep only the best model and its model dependencies.\n",
      " |          `models_to_delete` must be None if `models_to_keep` is set.\n",
      " |          To see the list of possible model names, use: `predictor.model_names()` or `predictor.leaderboard()`.\n",
      " |      models_to_delete : str or list, default = None\n",
      " |          Name of model or models to delete.\n",
      " |          All models that are not specified but depend on a model in `models_to_delete` will also be deleted.\n",
      " |          `models_to_keep` must be None if `models_to_delete` is set.\n",
      " |      allow_delete_cascade : bool, default = False\n",
      " |          If `False`, if unspecified dependent models of models in `models_to_delete` exist an exception will be raised instead of deletion occurring.\n",
      " |              An example of a dependent model is m1 if m2 is a stacker model and takes predictions from m1 as inputs. In this case, m1 would be a dependent model of m2.\n",
      " |          If `True`, all dependent models of models in `models_to_delete` will be deleted.\n",
      " |          Has no effect if `models_to_delete=None`.\n",
      " |      delete_from_disk : bool, default = True\n",
      " |          If `True`, deletes the models from disk if they were persisted.\n",
      " |          WARNING: This deletes the entire directory for the deleted models, and ALL FILES located there.\n",
      " |              It is highly recommended to first run with `dry_run=True` to understand which directories will be deleted.\n",
      " |      dry_run : bool, default = True\n",
      " |          If `True`, then deletions don't occur, and logging statements are printed describing what would have occurred.\n",
      " |          Set `dry_run=False` to perform the deletions.\n",
      " |  \n",
      " |  disk_usage(self) -> 'int'\n",
      " |      Returns the combined size of all files under the `predictor.path` directory in bytes.\n",
      " |  \n",
      " |  disk_usage_per_file(self, *, sort_by: 'str' = 'size', include_path_in_name: 'bool' = False) -> 'pd.Series'\n",
      " |      Returns the size of each file under the `predictor.path` directory in bytes.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sort_by : str, default = \"size\"\n",
      " |          If None, output files will be ordered based on order of search in os.walk(path).\n",
      " |          If \"size\", output files will be ordered in descending order of file size.\n",
      " |          If \"name\", output files will be ordered by name in ascending alphabetical order.\n",
      " |      include_path_in_name : bool, default = False\n",
      " |          If True, includes the full path of the file including the input `path` as part of the index in the output pd.Series.\n",
      " |          If False, removes the `path` prefix of the file path in the index of the output pd.Series.\n",
      " |      \n",
      " |          For example, for a file located at `foo/bar/model.pkl`, with path='foo/'\n",
      " |              If True, index will be `foo/bar/model.pkl`\n",
      " |              If False, index will be `bar/model.pkl`\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      pd.Series with index file path and value file size in bytes.\n",
      " |  \n",
      " |  distill(self, train_data: 'pd.DataFrame | str' = None, tuning_data: 'pd.DataFrame | str' = None, augmentation_data: 'pd.DataFrame' = None, time_limit: 'float' = None, hyperparameters: 'dict | str' = None, holdout_frac: 'float' = None, teacher_preds: 'str' = 'soft', augment_method: 'str' = 'spunge', augment_args: 'dict' = {'size_factor': 5, 'max_size': 100000}, models_name_suffix: 'str' = None, verbosity: 'int' = None)\n",
      " |      [EXPERIMENTAL]\n",
      " |      Distill AutoGluon's most accurate ensemble-predictor into single models which are simpler/faster and require less memory/compute.\n",
      " |      Distillation can produce a model that is more accurate than the same model fit directly on the original training data.\n",
      " |      After calling `distill()`, there will be more models available in this Predictor, which can be evaluated using `predictor.leaderboard(test_data)` and deployed with: `predictor.predict(test_data, model=MODEL_NAME)`.\n",
      " |      This will raise an exception if `cache_data=False` was previously set in `fit()`.\n",
      " |      \n",
      " |      NOTE: Until catboost v0.24 is released, `distill()` with CatBoost students in multiclass classification requires you to first install catboost-dev: `pip install catboost-dev`\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      train_data : str or :class:`pd.DataFrame`, default = None\n",
      " |          Same as `train_data` argument of `fit()`.\n",
      " |          If None, the same training data will be loaded from `fit()` call used to produce this Predictor.\n",
      " |      tuning_data : str or :class:`pd.DataFrame`, default = None\n",
      " |          Same as `tuning_data` argument of `fit()`.\n",
      " |          If `tuning_data = None` and `train_data = None`: the same training/validation splits will be loaded from `fit()` call used to produce this Predictor,\n",
      " |          unless bagging/stacking was previously used in which case a new training/validation split is performed.\n",
      " |      augmentation_data : :class:`pd.DataFrame`, default = None\n",
      " |          An optional extra dataset of unlabeled rows that can be used for augmenting the dataset used to fit student models during distillation (ignored if None).\n",
      " |      time_limit : int, default = None\n",
      " |          Approximately how long (in seconds) the distillation process should run for.\n",
      " |          If None, no time-constraint will be enforced allowing the distilled models to fully train.\n",
      " |      hyperparameters : dict or str, default = None\n",
      " |          Specifies which models to use as students and what hyperparameter-values to use for them.\n",
      " |          Same as `hyperparameters` argument of `fit()`.\n",
      " |          If = None, then student models will use the same hyperparameters from `fit()` used to produce this Predictor.\n",
      " |          Note: distillation is currently only supported for ['GBM','NN_TORCH','RF','CAT'] student models, other models and their hyperparameters are ignored here.\n",
      " |      holdout_frac : float\n",
      " |          Same as `holdout_frac` argument of :meth:`TabularPredictor.fit`.\n",
      " |      teacher_preds : str, default = 'soft'\n",
      " |          What form of teacher predictions to distill from (teacher refers to the most accurate AutoGluon ensemble-predictor).\n",
      " |          If None, we only train with original labels (no data augmentation).\n",
      " |          If 'hard', labels are hard teacher predictions given by: `teacher.predict()`\n",
      " |          If 'soft', labels are soft teacher predictions given by: `teacher.predict_proba()`\n",
      " |          Note: 'hard' and 'soft' are equivalent for regression problems.\n",
      " |          If `augment_method` is not None, teacher predictions are only used to label augmented data (training data keeps original labels).\n",
      " |          To apply label-smoothing: `teacher_preds='onehot'` will use original training data labels converted to one-hot vectors for multiclass problems (no data augmentation).\n",
      " |      augment_method : str, default='spunge'\n",
      " |          Specifies method to use for generating augmented data for distilling student models.\n",
      " |          Options include:\n",
      " |              None : no data augmentation performed.\n",
      " |              'munge' : The MUNGE algorithm (https://www.cs.cornell.edu/~caruana/compression.kdd06.pdf).\n",
      " |              'spunge' : A simpler, more efficient variant of the MUNGE algorithm.\n",
      " |      augment_args : dict, default = {'size_factor':5, 'max_size': int(1e5)}\n",
      " |          Contains the following kwargs that control the chosen `augment_method` (these are ignored if `augment_method=None`):\n",
      " |              'num_augmented_samples': int, number of augmented datapoints used during distillation. Overrides 'size_factor', 'max_size' if specified.\n",
      " |              'max_size': float, the maximum number of augmented datapoints to add (ignored if 'num_augmented_samples' specified).\n",
      " |              'size_factor': float, if n = training data sample-size, we add int(n * size_factor) augmented datapoints, up to 'max_size'.\n",
      " |              Larger values in `augment_args` will slow down the runtime of distill(), and may produce worse results if provided time_limit are too small.\n",
      " |              You can also pass in kwargs for the `spunge_augment`, `munge_augment` functions in `autogluon.tabular.augmentation.distill_utils`.\n",
      " |      models_name_suffix : str, default = None\n",
      " |          Optional suffix that can be appended at the end of all distilled student models' names.\n",
      " |          Note: all distilled models will contain '_DSTL' substring in their name by default.\n",
      " |      verbosity : int, default = None\n",
      " |          Controls amount of printed output during distillation (4 = highest, 0 = lowest).\n",
      " |          Same as `verbosity` parameter of :class:`TabularPredictor`.\n",
      " |          If None, the same `verbosity` used in previous fit is employed again.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List of names (str) corresponding to the distilled models.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from autogluon.tabular import TabularDataset, TabularPredictor\n",
      " |      >>> train_data = TabularDataset('train.csv')\n",
      " |      >>> predictor = TabularPredictor(label='class').fit(train_data, auto_stack=True)\n",
      " |      >>> distilled_model_names = predictor.distill()\n",
      " |      >>> test_data = TabularDataset('test.csv')\n",
      " |      >>> ldr = predictor.leaderboard(test_data)\n",
      " |      >>> model_to_deploy = distilled_model_names[0]\n",
      " |      >>> predictor.predict(test_data, model=model_to_deploy)\n",
      " |  \n",
      " |  evaluate(self, data: 'pd.DataFrame | str', model: 'str' = None, decision_threshold: 'float' = None, display: 'bool' = False, auxiliary_metrics: 'bool' = True, detailed_report: 'bool' = False, **kwargs) -> 'dict'\n",
      " |      Report the predictive performance evaluated over a given dataset.\n",
      " |      This is basically a shortcut for: `pred_proba = predict_proba(data); evaluate_predictions(data[label], pred_proba)`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str or :class:`pd.DataFrame`\n",
      " |          This dataset must also contain the `label` with the same column-name as previously specified.\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |          If `self.sample_weight` is set and `self.weight_evaluation==True`, then a column with the sample weight name is checked and used for weighted metric evaluation if it exists.\n",
      " |      model : str (optional)\n",
      " |          The name of the model to get prediction probabilities from. Defaults to None, which uses the highest scoring model on the validation set.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold to use when converting prediction probabilities to predictions.\n",
      " |          This will impact the scores of metrics such as `f1` and `accuracy`.\n",
      " |          If None, defaults to `predictor.decision_threshold`. Ignored unless `problem_type='binary'`.\n",
      " |          Refer to the `predictor.decision_threshold` docstring for more information.\n",
      " |      display : bool, default = False\n",
      " |          If True, performance results are printed.\n",
      " |      auxiliary_metrics: bool, default = True\n",
      " |          Should we compute other (`problem_type` specific) metrics in addition to the default metric?\n",
      " |      detailed_report : bool, default = False\n",
      " |          Should we computed more detailed versions of the `auxiliary_metrics`? (requires `auxiliary_metrics = True`)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Returns dict where keys = metrics, values = performance along each metric. To get the `eval_metric` score, do `output[predictor.eval_metric.name]`\n",
      " |      NOTE: Metrics scores always show in higher is better form.\n",
      " |      This means that metrics such as log_loss and root_mean_squared_error will have their signs FLIPPED, and values will be negative.\n",
      " |  \n",
      " |  evaluate_predictions(self, y_true, y_pred, sample_weight=None, decision_threshold=None, display: 'bool' = False, auxiliary_metrics=True, detailed_report=False, **kwargs) -> 'dict'\n",
      " |      Evaluate the provided prediction probabilities against ground truth labels.\n",
      " |      Evaluation is based on the `eval_metric` previously specified in init, or default metrics if none was specified.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      y_true : :class:`np.array` or :class:`pd.Series`\n",
      " |          The ordered collection of ground-truth labels.\n",
      " |      y_pred : :class:`pd.Series` or :class:`pd.DataFrame`\n",
      " |          The ordered collection of prediction probabilities or predictions.\n",
      " |          Obtainable via the output of `predictor.predict_proba`.\n",
      " |          Caution: For certain types of `eval_metric` (such as 'roc_auc'), `y_pred` must be predicted-probabilities rather than predicted labels.\n",
      " |      sample_weight : :class:`pd.Series`, default = None\n",
      " |          Sample weight for each row of data. If None, uniform sample weights are used.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold to use when converting prediction probabilities to predictions.\n",
      " |          This will impact the scores of metrics such as `f1` and `accuracy`.\n",
      " |          If None, defaults to `predictor.decision_threshold`. Ignored unless `problem_type='binary'`.\n",
      " |          Refer to the `predictor.decision_threshold` docstring for more information.\n",
      " |      display : bool, default = False\n",
      " |          If True, performance results are printed.\n",
      " |      auxiliary_metrics: bool, default = True\n",
      " |          Should we compute other (`problem_type` specific) metrics in addition to the default metric?\n",
      " |      detailed_report : bool, default = False\n",
      " |          Should we computed more detailed versions of the `auxiliary_metrics`? (requires `auxiliary_metrics = True`)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Returns dict where keys = metrics, values = performance along each metric.\n",
      " |      NOTE: Metrics scores always show in higher is better form.\n",
      " |      This means that metrics such as log_loss and root_mean_squared_error will have their signs FLIPPED, and values will be negative.\n",
      " |  \n",
      " |  feature_importance(self, data=None, model: 'str' = None, features: 'list' = None, feature_stage: 'str' = 'original', subsample_size: 'int' = 5000, time_limit: 'float' = None, num_shuffle_sets: 'int' = None, include_confidence_band: 'bool' = True, confidence_level: 'float' = 0.99, silent: 'bool' = False)\n",
      " |      Calculates feature importance scores for the given model via permutation importance. Refer to https://explained.ai/rf-importance/ for an explanation of permutation importance.\n",
      " |      A feature's importance score represents the performance drop that results when the model makes predictions on a perturbed copy of the data where this feature's values have been randomly shuffled across rows.\n",
      " |      A feature score of 0.01 would indicate that the predictive performance dropped by 0.01 when the feature was randomly shuffled.\n",
      " |      The higher the score a feature has, the more important it is to the model's performance.\n",
      " |      If a feature has a negative score, this means that the feature is likely harmful to the final model, and a model trained with the feature removed would be expected to achieve a better predictive performance.\n",
      " |      Note that calculating feature importance can be a very computationally expensive process, particularly if the model uses hundreds or thousands of features. In many cases, this can take longer than the original model training.\n",
      " |      To estimate how long `feature_importance(model, data, features)` will take, it is roughly the time taken by `predict_proba(data, model)` multiplied by the number of features.\n",
      " |      \n",
      " |      Note: For highly accurate importance and p_value estimates, it is recommended to set `subsample_size` to at least 5000 if possible and `num_shuffle_sets` to at least 10.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str or :class:`pd.DataFrame` (optional)\n",
      " |          This data must also contain the label-column with the same column-name as specified during `fit()`.\n",
      " |          If specified, then the data is used to calculate the feature importance scores.\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |          If not specified, the original data used during `fit()` will be used if `cache_data=True`. Otherwise, an exception will be raised.\n",
      " |          Do not pass the training data through this argument, as the feature importance scores calculated will be biased due to overfitting.\n",
      " |              More accurate feature importances will be obtained from new data that was held-out during `fit()`.\n",
      " |      model : str, default = None\n",
      " |          Model to get feature importances for, if None the best model is chosen.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`\n",
      " |      features : list, default = None\n",
      " |          List of str feature names that feature importances are calculated for and returned, specify None to get all feature importances.\n",
      " |          If you only want to compute feature importances for some of the features, you can pass their names in as a list of str.\n",
      " |          Valid feature names change depending on the `feature_stage`.\n",
      " |              To get the list of feature names for `feature_stage='original'`, call `predictor.feature_metadata_in.get_features()`.\n",
      " |              To get the list of feature names for `feature_stage='transformed'`, call `list(predictor.transform_features().columns)`.\n",
      " |              To get the list of feature names for `feature_stage=`transformed_model`, call `list(predictor.transform_features(model={model_name}).columns)`.\n",
      " |          [Advanced] Can also contain tuples as elements of (feature_name, feature_list) form.\n",
      " |              feature_name can be any string so long as it is unique with all other feature names / features in the list.\n",
      " |              feature_list can be any list of valid features in the data.\n",
      " |              This will compute importance of the combination of features in feature_list, naming the set of features in the returned DataFrame feature_name.\n",
      " |              This importance will differ from adding the individual importances of each feature in feature_list, and will be more accurate to the overall group importance.\n",
      " |              Example: ['featA', 'featB', 'featC', ('featBC', ['featB', 'featC'])]\n",
      " |              In this example, the importance of 'featBC' will be calculated by jointly permuting 'featB' and 'featC' together as if they were a single two-dimensional feature.\n",
      " |      feature_stage : str, default = 'original'\n",
      " |          What stage of feature-processing should importances be computed for.\n",
      " |          Options:\n",
      " |              'original':\n",
      " |                  Compute importances of the original features.\n",
      " |                  Warning: `data` must be specified with this option, otherwise an exception will be raised.\n",
      " |              'transformed':\n",
      " |                  Compute importances of the post-internal-transformation features (after automated feature engineering). These features may be missing some original features, or add new features entirely.\n",
      " |                  An example of new features would be ngram features generated from a text column.\n",
      " |                  Warning: For bagged models, feature importance calculation is not yet supported with this option when `data=None`. Doing so will raise an exception.\n",
      " |              'transformed_model':\n",
      " |                  Compute importances of the post-model-transformation features. These features are the internal features used by the requested model. They may differ greatly from the original features.\n",
      " |                  If the model is a stack ensemble, this will include stack ensemble features such as the prediction probability features of the stack ensemble's base (ancestor) models.\n",
      " |      subsample_size : int, default = 5000\n",
      " |          The number of rows to sample from `data` when computing feature importance.\n",
      " |          If `subsample_size=None` or `data` contains fewer than `subsample_size` rows, all rows will be used during computation.\n",
      " |          Larger values increase the accuracy of the feature importance scores.\n",
      " |          Runtime linearly scales with `subsample_size`.\n",
      " |      time_limit : float, default = None\n",
      " |          Time in seconds to limit the calculation of feature importance.\n",
      " |          If None, feature importance will calculate without early stopping.\n",
      " |          A minimum of 1 full shuffle set will always be evaluated. If a shuffle set evaluation takes longer than `time_limit`, the method will take the length of a shuffle set evaluation to return regardless of the `time_limit`.\n",
      " |      num_shuffle_sets : int, default = None\n",
      " |          The number of different permutation shuffles of the data that are evaluated.\n",
      " |          Larger values will increase the quality of the importance evaluation.\n",
      " |          It is generally recommended to increase `subsample_size` before increasing `num_shuffle_sets`.\n",
      " |          Defaults to 5 if `time_limit` is None or 10 if `time_limit` is specified.\n",
      " |          Runtime linearly scales with `num_shuffle_sets`.\n",
      " |      include_confidence_band: bool, default = True\n",
      " |          If True, returned DataFrame will include two additional columns specifying confidence interval for the true underlying importance value of each feature.\n",
      " |          Increasing `subsample_size` and `num_shuffle_sets` will tighten the confidence interval.\n",
      " |      confidence_level: float, default = 0.99\n",
      " |          This argument is only considered when `include_confidence_band` is True, and can be used to specify the confidence level used for constructing confidence intervals.\n",
      " |          For example, if `confidence_level` is set to 0.99, then the returned DataFrame will include columns 'p99_high' and 'p99_low' which indicates that the true feature importance will be between 'p99_high' and 'p99_low' 99% of the time (99% confidence interval).\n",
      " |          More generally, if `confidence_level` = 0.XX, then the columns containing the XX% confidence interval will be named 'pXX_high' and 'pXX_low'.\n",
      " |      silent : bool, default = False\n",
      " |          Whether to suppress logging output.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.DataFrame` of feature importance scores with 6 columns:\n",
      " |          index: The feature name.\n",
      " |          'importance': The estimated feature importance score.\n",
      " |          'stddev': The standard deviation of the feature importance score. If NaN, then not enough num_shuffle_sets were used to calculate a variance.\n",
      " |          'p_value': P-value for a statistical t-test of the null hypothesis: importance = 0, vs the (one-sided) alternative: importance > 0.\n",
      " |              Features with low p-value appear confidently useful to the predictor, while the other features may be useless to the predictor (or even harmful to include in its training data).\n",
      " |              A p-value of 0.01 indicates that there is a 1% chance that the feature is useless or harmful, and a 99% chance that the feature is useful.\n",
      " |              A p-value of 0.99 indicates that there is a 99% chance that the feature is useless or harmful, and a 1% chance that the feature is useful.\n",
      " |          'n': The number of shuffles performed to estimate importance score (corresponds to sample-size used to determine confidence interval for true score).\n",
      " |          'pXX_high': Upper end of XX% confidence interval for true feature importance score (where XX=99 by default).\n",
      " |          'pXX_low': Lower end of XX% confidence interval for true feature importance score.\n",
      " |  \n",
      " |  features(self, feature_stage: 'str' = 'original') -> 'list'\n",
      " |      Returns a list of feature names dependent on the value of feature_stage.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      feature_stage : str, default = 'original'\n",
      " |          If 'original', returns the list of features specified in the original training data. This feature set is required in input data when making predictions.\n",
      " |          If 'transformed', returns the list of features after pre-processing by the feature generator.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Returns a list of feature names\n",
      " |  \n",
      " |  fit(self, train_data: 'pd.DataFrame | str', tuning_data: 'pd.DataFrame | str' = None, time_limit: 'float' = None, presets: 'List[str] | str' = None, hyperparameters: 'dict | str' = None, feature_metadata='infer', infer_limit: 'float' = None, infer_limit_batch_size: 'int' = None, fit_weighted_ensemble: 'bool' = True, fit_full_last_level_weighted_ensemble: 'bool' = True, full_weighted_ensemble_additionally: 'bool' = False, dynamic_stacking: 'bool | str' = False, calibrate_decision_threshold: 'bool | str' = 'auto', num_cpus: 'int | str' = 'auto', num_gpus: 'int | str' = 'auto', fit_strategy: \"Literal['sequential', 'parallel']\" = 'sequential', memory_limit: 'float | str' = 'auto', callbacks: 'List[AbstractCallback]' = None, **kwargs) -> \"'TabularPredictor'\"\n",
      " |      Fit models to predict a column of a data table (label) based on the other columns (features).\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      train_data : :class:`pd.DataFrame` or str\n",
      " |          Table of the training data as a pandas DataFrame.\n",
      " |          If str is passed, `train_data` will be loaded using the str value as the file path.\n",
      " |      tuning_data : :class:`pd.DataFrame` or str, optional\n",
      " |          Another dataset containing validation data reserved for tuning processes such as early stopping and hyperparameter tuning.\n",
      " |          This dataset should be in the same format as `train_data`.\n",
      " |          If str is passed, `tuning_data` will be loaded using the str value as the file path.\n",
      " |          Note: final model returned may be fit on `tuning_data` as well as `train_data`. Do not provide your evaluation test data here!\n",
      " |          In particular, when `num_bag_folds` > 0 or `num_stack_levels` > 0, models will be trained on both `tuning_data` and `train_data`.\n",
      " |          If `tuning_data = None`, `fit()` will automatically hold out some random validation examples from `train_data`.\n",
      " |      time_limit : int, default = None\n",
      " |          Approximately how long `fit()` should run for (wallclock time in seconds).\n",
      " |          If not specified, `fit()` will run until all models have completed training, but will not repeatedly bag models unless `num_bag_sets` is specified.\n",
      " |      presets : list or str or dict, default = ['medium_quality']\n",
      " |          List of preset configurations for various arguments in `fit()`. Can significantly impact predictive accuracy, memory-footprint, and inference latency of trained models, and various other properties of the returned `predictor`.\n",
      " |          It is recommended to specify presets and avoid specifying most other `fit()` arguments or model hyperparameters prior to becoming familiar with AutoGluon.\n",
      " |          As an example, to get the most accurate overall predictor (regardless of its efficiency), set `presets='best_quality'`.\n",
      " |          To get good quality with minimal disk usage, set `presets=['good_quality', 'optimize_for_deployment']`\n",
      " |          Any user-specified arguments in `fit()` will override the values used by presets.\n",
      " |          If specifying a list of presets, later presets will override earlier presets if they alter the same argument.\n",
      " |          For precise definitions of the provided presets, see file: `autogluon/tabular/configs/presets_configs.py`.\n",
      " |          Users can specify custom presets by passing in a dictionary of argument values as an element to the list.\n",
      " |      \n",
      " |          Available Presets: ['best_quality', 'high_quality', 'good_quality', 'medium_quality', 'experimental_quality', 'optimize_for_deployment', 'interpretable', 'ignore_text']\n",
      " |      \n",
      " |          It is recommended to only use one `quality` based preset in a given call to `fit()` as they alter many of the same arguments and are not compatible with each-other.\n",
      " |      \n",
      " |          In-depth Preset Info:\n",
      " |              best_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'zeroshot'}\n",
      " |                  Best predictive accuracy with little consideration to inference time or disk usage. Achieve even better results by specifying a large time_limit value.\n",
      " |                  Recommended for applications that benefit from the best possible model accuracy.\n",
      " |      \n",
      " |              high_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'zeroshot', 'refit_full': True, 'set_best_to_refit_full': True, 'save_bag_folds': False}\n",
      " |                  High predictive accuracy with fast inference. ~8x faster inference and ~8x lower disk usage than `best_quality`.\n",
      " |                  Recommended for applications that require reasonable inference speed and/or model size.\n",
      " |      \n",
      " |              good_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'light', 'refit_full': True, 'set_best_to_refit_full': True, 'save_bag_folds': False}\n",
      " |                  Good predictive accuracy with very fast inference. ~4x faster inference and ~4x lower disk usage than `high_quality`.\n",
      " |                  Recommended for applications that require fast inference speed.\n",
      " |      \n",
      " |              medium_quality={'auto_stack': False}\n",
      " |                  Medium predictive accuracy with very fast inference and very fast training time. ~20x faster training than `good_quality`.\n",
      " |                  This is the default preset in AutoGluon, but should generally only be used for quick prototyping, as `good_quality` results in significantly better predictive accuracy and faster inference time.\n",
      " |      \n",
      " |              experimental_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'experimental', 'fit_strategy': 'parallel', 'num_gpus': 0}\n",
      " |                  This preset acts as a testing ground for cutting edge features and models which could later be added to the `best_quality` preset in future releases.\n",
      " |                  Recommended when `best_quality` was already being used and the user wants to push performance even further.\n",
      " |      \n",
      " |              optimize_for_deployment={'keep_only_best': True, 'save_space': True}\n",
      " |                  Optimizes result immediately for deployment by deleting unused models and removing training artifacts.\n",
      " |                  Often can reduce disk usage by ~2-4x with no negatives to model accuracy or inference speed.\n",
      " |                  This will disable numerous advanced functionality, but has no impact on inference.\n",
      " |                  This will make certain functionality less informative, such as `predictor.leaderboard()` and `predictor.fit_summary()`.\n",
      " |                      Because unused models will be deleted under this preset, methods like `predictor.leaderboard()` and `predictor.fit_summary()` will no longer show the full set of models that were trained during `fit()`.\n",
      " |                  Recommended for applications where the inner details of AutoGluon's training is not important and there is no intention of manually choosing between the final models.\n",
      " |                  This preset pairs well with the other presets such as `good_quality` to make a very compact final model.\n",
      " |                  Identical to calling `predictor.delete_models(models_to_keep='best', dry_run=False)` and `predictor.save_space()` directly after `fit()`.\n",
      " |      \n",
      " |              interpretable={'auto_stack': False, 'hyperparameters': 'interpretable'}\n",
      " |                  Fits only interpretable rule-based models from the imodels package.\n",
      " |                  Trades off predictive accuracy for conciseness.\n",
      " |      \n",
      " |              ignore_text={'_feature_generator_kwargs': {'enable_text_ngram_features': False, 'enable_text_special_features': False, 'enable_raw_text_features': False}}\n",
      " |                  Disables automated feature generation when text features are detected.\n",
      " |                  This is useful to determine how beneficial text features are to the end result, as well as to ensure features are not mistaken for text when they are not.\n",
      " |                  Ignored if `feature_generator` was also specified.\n",
      " |      \n",
      " |      hyperparameters : str or dict, default = 'default'\n",
      " |          Determines the hyperparameters used by the models.\n",
      " |          If `str` is passed, will use a preset hyperparameter configuration.\n",
      " |              Valid `str` options: ['default', 'zeroshot', 'light', 'very_light', 'toy', 'multimodal']\n",
      " |                  'default': Default AutoGluon hyperparameters intended to get strong accuracy with reasonable disk usage and inference time. Used in the 'medium_quality' preset.\n",
      " |                  'zeroshot': A powerful model portfolio learned from TabRepo's ensemble simulation on 200 datasets. Contains ~100 models and is used in 'best_quality' and 'high_quality' presets.\n",
      " |                  'light': Results in smaller models. Generally will make inference speed much faster and disk usage much lower, but with worse accuracy. Used in the 'good_quality' preset.\n",
      " |                  'very_light': Results in much smaller models. Behaves similarly to 'light', but in many cases with over 10x less disk usage and a further reduction in accuracy.\n",
      " |                  'toy': Results in extremely small models. Only use this when prototyping, as the model quality will be severely reduced.\n",
      " |                  'multimodal': [EXPERIMENTAL] Trains a multimodal transformer model alongside tabular models. Requires that some text columns appear in the data and GPU.\n",
      " |                      When combined with 'best_quality' `presets` option, this can achieve extremely strong results in multimodal data tables that contain columns with text in addition to numeric/categorical columns.\n",
      " |              Reference `autogluon/tabular/configs/hyperparameter_configs.py` for information on the hyperparameters associated with each preset.\n",
      " |          Keys are strings that indicate which model types to train.\n",
      " |              Stable model options include:\n",
      " |                  'GBM' (LightGBM)\n",
      " |                  'CAT' (CatBoost)\n",
      " |                  'XGB' (XGBoost)\n",
      " |                  'RF' (random forest)\n",
      " |                  'XT' (extremely randomized trees)\n",
      " |                  'KNN' (k-nearest neighbors)\n",
      " |                  'LR' (linear regression)\n",
      " |                  'NN_TORCH' (neural network implemented in Pytorch)\n",
      " |                  'FASTAI' (neural network with FastAI backend)\n",
      " |                  'AG_AUTOMM' (`MultimodalPredictor` from `autogluon.multimodal`. Supports Tabular, Text, and Image modalities. GPU is required.)\n",
      " |              Experimental model options include:\n",
      " |                  'FT_TRANSFORMER' (Tabular Transformer, GPU is recommended. Does not scale well to >100 features.)\n",
      " |                  'FASTTEXT' (FastText. Note: Has not been tested for a long time.)\n",
      " |                  'TABPFN' (TabPFN. Does not scale well to >100 features or >1000 rows, and does not support regression. Extremely slow inference speed.)\n",
      " |                  'VW' (VowpalWabbit. Note: Has not been tested for a long time.)\n",
      " |                  'AG_TEXT_NN' (Multimodal Text+Tabular model, GPU is required. Recommended to instead use its successor, 'AG_AUTOMM'.)\n",
      " |                  'AG_IMAGE_NN' (Image model, GPU is required. Recommended to instead use its successor, 'AG_AUTOMM'.)\n",
      " |              If a certain key is missing from hyperparameters, then `fit()` will not train any models of that type. Omitting a model key from hyperparameters is equivalent to including this model key in `excluded_model_types`.\n",
      " |              For example, set `hyperparameters = { 'NN_TORCH':{...} }` if say you only want to train (PyTorch) neural networks and no other types of models.\n",
      " |          Values = dict of hyperparameter settings for each model type, or list of dicts.\n",
      " |              Each hyperparameter can either be a single fixed value or a search space containing many possible values.\n",
      " |              Unspecified hyperparameters will be set to default values (or default search spaces if `hyperparameter_tune_kwargs='auto'`).\n",
      " |              Caution: Any provided search spaces will error if `hyperparameter_tune_kwargs=None` (Default).\n",
      " |              To train multiple models of a given type, set the value to a list of hyperparameter dictionaries.\n",
      " |                  For example, `hyperparameters = {'RF': [{'criterion': 'gini'}, {'criterion': 'entropy'}]}` will result in 2 random forest models being trained with separate hyperparameters.\n",
      " |          Advanced functionality: Bring your own model / Custom model support\n",
      " |              AutoGluon fully supports custom models. For a detailed tutorial on creating and using custom models with AutoGluon, refer to https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-model.html\n",
      " |          Advanced functionality: Custom stack levels\n",
      " |              By default, AutoGluon re-uses the same models and model hyperparameters at each level during stack ensembling.\n",
      " |              To customize this behaviour, create a hyperparameters dictionary separately for each stack level, and then add them as values to a new dictionary, with keys equal to the stack level.\n",
      " |                  Example: `hyperparameters = {1: {'RF': rf_params1}, 2: {'CAT': [cat_params1, cat_params2], 'NN_TORCH': {}}}`\n",
      " |                  This will result in a stack ensemble that has one custom random forest in level 1 followed by two CatBoost models with custom hyperparameters and a default neural network in level 2, for a total of 4 models.\n",
      " |              If a level is not specified in `hyperparameters`, it will default to using the highest specified level to train models. This can also be explicitly controlled by adding a 'default' key.\n",
      " |      \n",
      " |          Default:\n",
      " |              hyperparameters = {\n",
      " |                  'NN_TORCH': {},\n",
      " |                  'GBM': [\n",
      " |                      {'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}},\n",
      " |                      {},\n",
      " |                      {\n",
      " |                          \"learning_rate\": 0.03,\n",
      " |                          \"num_leaves\": 128,\n",
      " |                          \"feature_fraction\": 0.9,\n",
      " |                          \"min_data_in_leaf\": 3,\n",
      " |                          \"ag_args\": {\"name_suffix\": \"Large\", \"priority\": 0, \"hyperparameter_tune_kwargs\": None},\n",
      " |                      },\n",
      " |                  ],\n",
      " |                  'CAT': {},\n",
      " |                  'XGB': {},\n",
      " |                  'FASTAI': {},\n",
      " |                  'RF': [\n",
      " |                      {'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}},\n",
      " |                      {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}},\n",
      " |                      {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression']}},\n",
      " |                  ],\n",
      " |                  'XT': [\n",
      " |                      {'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}},\n",
      " |                      {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}},\n",
      " |                      {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression']}},\n",
      " |                  ],\n",
      " |                  'KNN': [\n",
      " |                      {'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}},\n",
      " |                      {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}},\n",
      " |                  ],\n",
      " |              }\n",
      " |      \n",
      " |          Details regarding the hyperparameters you can specify for each model are provided in the following files:\n",
      " |              NN: `autogluon.tabular.models.tabular_nn.hyperparameters.parameters`\n",
      " |                  Note: certain hyperparameter settings may cause these neural networks to train much slower.\n",
      " |              GBM: `autogluon.tabular.models.lgb.hyperparameters.parameters`\n",
      " |                   See also the lightGBM docs: https://lightgbm.readthedocs.io/en/latest/Parameters.html\n",
      " |              CAT: `autogluon.tabular.models.catboost.hyperparameters.parameters`\n",
      " |                   See also the CatBoost docs: https://catboost.ai/docs/concepts/parameter-tuning.html\n",
      " |              XGB: `autogluon.tabular.models.xgboost.hyperparameters.parameters`\n",
      " |                   See also the XGBoost docs: https://xgboost.readthedocs.io/en/latest/parameter.html\n",
      " |              FASTAI: `autogluon.tabular.models.fastainn.hyperparameters.parameters`\n",
      " |                   See also the FastAI docs: https://docs.fast.ai/tabular.learner.html\n",
      " |              RF: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html\n",
      " |                  Note: Hyperparameter tuning is disabled for this model.\n",
      " |              XT: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html\n",
      " |                  Note: Hyperparameter tuning is disabled for this model.\n",
      " |              KNN: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html\n",
      " |                  Note: Hyperparameter tuning is disabled for this model.\n",
      " |              LR: `autogluon.tabular.models.lr.hyperparameters.parameters`\n",
      " |                  Note: Hyperparameter tuning is disabled for this model.\n",
      " |                  Note: 'penalty' parameter can be used for regression to specify regularization method: 'L1' and 'L2' values are supported.\n",
      " |              Advanced functionality: Custom AutoGluon model arguments\n",
      " |                  These arguments are optional and can be specified in any model's hyperparameters.\n",
      " |                      Example: `hyperparameters = {'RF': {..., 'ag_args': {'name_suffix': 'CustomModelSuffix', 'disable_in_hpo': True}}`\n",
      " |                  ag_args: Dictionary of customization options related to meta properties of the model such as its name, the order it is trained, the problem types it is valid for, and the type of HPO it utilizes.\n",
      " |                      Valid keys:\n",
      " |                          name: (str) The name of the model. This overrides AutoGluon's naming logic and all other name arguments if present.\n",
      " |                          name_main: (str) The main name of the model. Example: 'RandomForest'.\n",
      " |                          name_prefix: (str) Add a custom prefix to the model name. Unused by default.\n",
      " |                          name_suffix: (str) Add a custom suffix to the model name. Unused by default.\n",
      " |                          priority: (int) Determines the order in which the model is trained. Larger values result in the model being trained earlier. Default values range from 100 (KNN) to 0 (custom), dictated by model type. If you want this model to be trained first, set priority = 999.\n",
      " |                          problem_types: (list) List of valid problem types for the model. `problem_types=['binary']` will result in the model only being trained if `problem_type` is 'binary'.\n",
      " |                          disable_in_hpo: (bool) If True, the model will only be trained if `hyperparameter_tune_kwargs=None`.\n",
      " |                          valid_stacker: (bool) If False, the model will not be trained as a level 2 or higher stacker model.\n",
      " |                          valid_base: (bool) If False, the model will not be trained as a level 1 (base) model.\n",
      " |                          hyperparameter_tune_kwargs: (dict) Refer to :meth:`TabularPredictor.fit` hyperparameter_tune_kwargs argument. If specified here, will override global HPO settings for this model.\n",
      " |                      Reference the default hyperparameters for example usage of these options.\n",
      " |                  ag_args_fit: Dictionary of model fit customization options related to how and with what constraints the model is trained. These parameters affect stacker fold models, but not stacker models themselves.\n",
      " |                      Clarification: `time_limit` is the internal time in seconds given to a particular model to train, which is dictated in part by the `time_limit` argument given during `predictor.fit()` but is not the same.\n",
      " |                      Valid keys:\n",
      " |                          stopping_metric: (str or :class:`autogluon.core.metrics.Scorer`, default=None) The metric to use for early stopping of the model. If None, model will decide.\n",
      " |                          max_memory_usage_ratio: (float, default=1.0) The ratio of memory usage relative to the default to allow before early stopping or killing the model. Values greater than 1.0 will be increasingly prone to out-of-memory errors.\n",
      " |                          max_time_limit_ratio: (float, default=1.0) The ratio of the provided time_limit to use during model `fit()`. If `time_limit=10` and `max_time_limit_ratio=0.3`, time_limit would be changed to 3. Does not alter max_time_limit or min_time_limit values.\n",
      " |                          max_time_limit: (float, default=None) Maximum amount of time to allow this model to train for (in sec). If the provided time_limit is greater than this value, it will be replaced by max_time_limit.\n",
      " |                          min_time_limit: (float, default=0) Allow this model to train for at least this long (in sec), regardless of the time limit it would otherwise be granted.\n",
      " |                              If `min_time_limit >= max_time_limit`, time_limit will be set to min_time_limit.\n",
      " |                              If `min_time_limit=None`, time_limit will be set to None and the model will have no training time restriction.\n",
      " |                          num_cpus : (int or str, default='auto')\n",
      " |                              How many CPUs to use during model fit.\n",
      " |                              If 'auto', model will decide.\n",
      " |                          num_gpus : (int or str, default='auto')\n",
      " |                              How many GPUs to use during model fit.\n",
      " |                              If 'auto', model will decide. Some models can use GPUs but don't by default due to differences in model quality.\n",
      " |                              Set to 0 to disable usage of GPUs.\n",
      " |                  ag_args_ensemble: Dictionary of hyperparameters shared by all models that control how they are ensembled, if bag mode is enabled.\n",
      " |                      Valid keys:\n",
      " |                          use_orig_features: [True, False, \"never\"], default True\n",
      " |                              Whether a stack model will use the original features along with the stack features to train (akin to skip-connections).\n",
      " |                              If True, will use the original data features.\n",
      " |                              If False, will discard the original data features and only use stack features, except when no stack features exist (such as in layer 1).\n",
      " |                              If \"never\", will always discard the original data features. Will be skipped in layer 1.\n",
      " |                          valid_stacker : bool, default True\n",
      " |                              If True, will be marked as valid to include as a stacker model.\n",
      " |                              If False, will only be fit as a base model (layer 1) and will not be fit in stack layers (layer 2+).\n",
      " |                          max_base_models : int, default 0\n",
      " |                              Maximum number of base models whose predictions form the features input to this stacker model.\n",
      " |                              If more than `max_base_models` base models are available, only the top `max_base_models` models with highest validation score are used.\n",
      " |                              If 0, the logic is skipped.\n",
      " |                          max_base_models_per_type : int | str, default \"auto\"\n",
      " |                              Similar to `max_base_models`. If more than `max_base_models_per_type` of any particular model type are available,\n",
      " |                              only the top `max_base_models_per_type` of that type are used. This occurs before the `max_base_models` filter.\n",
      " |                              If \"auto\", the value will be adaptively set based on the number of training samples.\n",
      " |                                  More samples will lead to larger values, starting at 1 with <1000 samples, increasing up to 12 at >=50000 samples.\n",
      " |                              If 0, the logic is skipped.\n",
      " |                          num_folds: (int, default=None) If specified, the number of folds to fit in the bagged model.\n",
      " |                              If specified, overrides any other value used to determine the number of folds such as predictor.fit `num_bag_folds` argument.\n",
      " |                          max_sets: (int, default=None) If specified, the maximum sets to fit in the bagged model.\n",
      " |                              The lesser of `max_sets` and the predictor.fit `num_bag_sets` argument will be used for the given model.\n",
      " |                              Useful if a particular model is expensive relative to others and you want to avoid repeated bagging of the expensive model while still repeated bagging the cheaper models.\n",
      " |                          save_bag_folds: (bool, default=True)\n",
      " |                              If True, bagged models will save their fold models (the models from each individual fold of bagging). This is required to use bagged models for prediction.\n",
      " |                              If False, bagged models will not save their fold models. This means that bagged models will not be valid models during inference.\n",
      " |                                  This should only be set to False when planning to call `predictor.refit_full()` or when `refit_full` is set and `set_best_to_refit_full=True`.\n",
      " |                                  Particularly useful if disk usage is a concern. By not saving the fold models, bagged models will use only very small amounts of disk space during training.\n",
      " |                                  In many training runs, this will reduce peak disk usage by >10x.\n",
      " |                          fold_fitting_strategy: (AbstractFoldFittingStrategy default=auto) Whether to fit folds in parallel or in sequential order.\n",
      " |                              If parallel_local, folds will be trained in parallel with evenly distributed computing resources. This could bring 2-4x speedup compared to SequentialLocalFoldFittingStrategy, but could consume much more memory.\n",
      " |                              If sequential_local, folds will be trained in sequential.\n",
      " |                              If auto, strategy will be determined by OS and whether ray is installed or not. MacOS support for parallel_local is unstable, and may crash if enabled.\n",
      " |                          num_folds_parallel: (int or str, default='auto') Number of folds to be trained in parallel if using ParallelLocalFoldFittingStrategy. Consider lowering this value if you encounter either out of memory issue or CUDA out of memory issue(when trained on gpu).\n",
      " |                              if 'auto', will try to train all folds in parallel.\n",
      " |      \n",
      " |      feature_metadata : :class:`autogluon.tabular.FeatureMetadata` or str, default = 'infer'\n",
      " |          The feature metadata used in various inner logic in feature preprocessing.\n",
      " |          If 'infer', will automatically construct a FeatureMetadata object based on the properties of `train_data`.\n",
      " |          In this case, `train_data` is input into :meth:`autogluon.tabular.FeatureMetadata.from_df` to infer `feature_metadata`.\n",
      " |          If 'infer' incorrectly assumes the dtypes of features, consider explicitly specifying `feature_metadata`.\n",
      " |      infer_limit : float, default = None\n",
      " |          The inference time limit in seconds per row to adhere to during fit.\n",
      " |          If infer_limit=0.05 and infer_limit_batch_size=1000, AutoGluon will avoid training models that take longer than 50 ms/row to predict when given a batch of 1000 rows to predict (must predict 1000 rows in no more than 50 seconds).\n",
      " |          If bagging is enabled, the inference time limit will be respected based on estimated inference speed of `_FULL` models after refit_full is called, NOT on the inference speed of the bagged ensembles.\n",
      " |          The inference times calculated for models are assuming `predictor.persist('all')` is called after fit.\n",
      " |          If None, no limit is enforced.\n",
      " |          If it is impossible to satisfy the constraint, an exception will be raised.\n",
      " |      infer_limit_batch_size : int, default = None\n",
      " |          The batch size to use when predicting in bulk to estimate per-row inference time.\n",
      " |          Must be an integer greater than 0.\n",
      " |          If None and `infer_limit` is specified, will default to 10000.\n",
      " |          It is recommended to set to 10000 unless you must satisfy an online-inference scenario.\n",
      " |          Small values, especially `infer_limit_batch_size=1`, will result in much larger per-row inference times and should be avoided if possible.\n",
      " |          Refer to `infer_limit` for more details on how this is used.\n",
      " |          If specified when `infer_limit=None`, the inference time will be logged during training but will not be limited.\n",
      " |      fit_weighted_ensemble : bool, default = True\n",
      " |          If True, a WeightedEnsembleModel will be fit in each stack layer.\n",
      " |          A weighted ensemble will often be stronger than an individual model while being very fast to train.\n",
      " |          It is recommended to keep this value set to True to maximize predictive quality.\n",
      " |      fit_full_last_level_weighted_ensemble : bool, default = True\n",
      " |          If True, the WeightedEnsembleModel of the last stacking level will be fit with all (successful) models from all previous layers as base models.\n",
      " |          If stacking is disabled, settings this to True or False makes no difference because the WeightedEnsembleModel L2 always uses all models from L1.\n",
      " |          It is recommended to keep this value set to True to maximize predictive quality.\n",
      " |      full_weighted_ensemble_additionally : bool, default = False\n",
      " |          If True, AutoGluon will fit two WeightedEnsembleModels after training all stacking levels. Setting this to True, simulates calling\n",
      " |          `fit_weighted_ensemble()` after calling `fit()`. Has no affect if `fit_full_last_level_weighted_ensemble` is False and does not fit an additional\n",
      " |          WeightedEnsembleModel if stacking is disabled.\n",
      " |      dynamic_stacking: bool | str, default = False\n",
      " |          If True and `num_stack_levels` > 0, AutoGluon will dynamically determine whether to use stacking or not by first validating AutoGluon's stacking\n",
      " |          behavior. This is done to avoid so-called stacked overfitting that can make traditional multi-layer stacking, as used in AutoGluon, fail drastically\n",
      " |          and produce unreliable validation scores.\n",
      " |          It is recommended to keep this value set to True or \"auto\" when using stacking,\n",
      " |          as long as it is unknown whether the data is affected by stacked overfitting.\n",
      " |          If it is known that the data is unaffected by stacked overfitting, then setting this value to False is expected to maximize predictive quality.\n",
      " |          If enabled, by default, AutoGluon performs dynamic stacking by spending 25% of the provided time limit for detection and all remaining\n",
      " |          time for fitting AutoGluon. This can be adjusted by specifying `ds_args` with different parameters to `fit()`.\n",
      " |          If \"auto\", will be set to `not use_bag_holdout`.\n",
      " |          See the documentation of `ds_args` for more information.\n",
      " |      calibrate_decision_threshold : bool | str, default = \"auto\"\n",
      " |          If True, will automatically calibrate the decision threshold at the end of fit for calls to `.predict` based on the evaluation metric.\n",
      " |          If \"auto\", will be set to True if `eval_metric.needs_class=True` and `problem_type=\"binary\"`.\n",
      " |          By default, the decision threshold is `0.5`, however for some metrics such as `f1` and `balanced_accuracy`,\n",
      " |          scores can be significantly improved by choosing a threshold other than `0.5`.\n",
      " |          Only valid for `problem_type='binary'`. Ignored for all other problem types.\n",
      " |      num_cpus: int | str, default = \"auto\"\n",
      " |          The total amount of cpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of cpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      num_gpus: int | str, default = \"auto\"\n",
      " |          The total amount of gpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of gpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      fit_strategy: Literal[\"sequential\", \"parallel\"], default = \"sequential\"\n",
      " |          The strategy used to fit models.\n",
      " |          If \"sequential\", models will be fit sequentially. This is the most stable option with the most readable logging.\n",
      " |          If \"parallel\", models will be fit in parallel with ray, splitting available compute between them.\n",
      " |              Note: \"parallel\" is experimental and may run into issues. It was first added in version 1.2.0.\n",
      " |          For machines with 16 or more CPU cores, it is likely that \"parallel\" will be faster than \"sequential\".\n",
      " |      \n",
      " |          .. versionadded:: 1.2.0\n",
      " |      \n",
      " |      memory_limit: float | str, default = \"auto\"\n",
      " |          The total amount of memory in GB you want AutoGluon predictor to use. \"auto\" means AutoGluon will use all available memory on the system\n",
      " |          (that is detectable by psutil).\n",
      " |          Note that this is only a soft limit! AutoGluon uses this limit to skip training models that are expected to require too much memory or stop\n",
      " |          training a model that would exceed the memory limit. AutoGluon does not guarantee the enforcement of this limit (yet). Nevertheless, we expect\n",
      " |          AutoGluon to abide by the limit in most cases or, at most, go over the limit by a small margin.\n",
      " |          For most virtualized systems (e.g., in the cloud) and local usage on a server or laptop, \"auto\" is ideal for this parameter. We recommend manually\n",
      " |          setting the memory limit (and any other resources) on systems with shared resources that are controlled by the operating system (e.g., SLURM and\n",
      " |          cgroups). Otherwise, AutoGluon might wrongly assume more resources are available for fitting a model than the operating system allows,\n",
      " |          which can result in model training failing or being very inefficient.\n",
      " |      callbacks : List[AbstractCallback], default = None\n",
      " |          :::{warning}\n",
      " |          Callbacks are an experimental feature and may change in future releases without warning.\n",
      " |          Callback support is preliminary and targeted towards developers.\n",
      " |          :::\n",
      " |          A list of callback objects inheriting from `autogluon.core.callbacks.AbstractCallback`.\n",
      " |          These objects will be called before and after each model fit within trainer.\n",
      " |          They have the ability to skip models or early stop the training process.\n",
      " |          They can also theoretically change the entire logical flow of the trainer code by interacting with the passed `trainer` object.\n",
      " |          For more details, refer to `AbstractCallback` source code.\n",
      " |          If None, no callback objects will be used.\n",
      " |      \n",
      " |          [Note] Callback objects can be mutated in-place by the fit call if they are stateful.\n",
      " |          Ensure that you avoid re-using a mutated callback object between multiple fit calls.\n",
      " |      \n",
      " |          [Note] Callback objects are deleted from trainer at the end of the fit call. They will not impact operations such as `refit_full` or `fit_extra`.\n",
      " |      **kwargs :\n",
      " |          auto_stack : bool, default = False\n",
      " |              Whether AutoGluon should automatically utilize bagging and multi-layer stack ensembling to boost predictive accuracy.\n",
      " |              Set this = True if you are willing to tolerate longer training times in order to maximize predictive accuracy!\n",
      " |              Automatically sets `num_bag_folds` and `num_stack_levels` arguments based on dataset properties.\n",
      " |              Note: Setting `num_bag_folds` and `num_stack_levels` arguments will override `auto_stack`.\n",
      " |              Note: This can increase training time (and inference time) by up to 20x, but can greatly improve predictive performance.\n",
      " |          num_bag_folds : int, default = None\n",
      " |              Number of folds used for bagging of models. When `num_bag_folds = k`, training time is roughly increased by a factor of `k` (set = 0 to disable bagging).\n",
      " |              Disabled by default (0), but we recommend values between 5-10 to maximize predictive performance.\n",
      " |              Increasing num_bag_folds will result in models with lower bias but that are more prone to overfitting.\n",
      " |              `num_bag_folds = 1` is an invalid value, and will raise a ValueError.\n",
      " |              Values > 10 may produce diminishing returns, and can even harm overall results due to overfitting.\n",
      " |              To further improve predictions, avoid increasing `num_bag_folds` much beyond 10 and instead increase `num_bag_sets`.\n",
      " |          num_bag_sets : int, default = None\n",
      " |              Number of repeats of kfold bagging to perform (values must be >= 1). Total number of models trained during bagging = `num_bag_folds * num_bag_sets`.\n",
      " |              Defaults to 1 when unspecified. Value is ignored if `num_bag_folds<=2`.\n",
      " |              Values greater than 1 will result in superior predictive performance, especially on smaller problems and with stacking enabled (reduces overall variance).\n",
      " |              Be warned: This will drastically increase overall runtime, and if using a time limit, can very commonly lead to worse performance.\n",
      " |              It is recommended to increase this value only as a last resort, as it is the least computationally efficient method to improve performance.\n",
      " |          num_stack_levels : int, default = None\n",
      " |              Number of stacking levels to use in stack ensemble. Roughly increases model training time by factor of `num_stack_levels+1` (set = 0 to disable stack ensembling).\n",
      " |              Disabled by default (0), but we recommend `num_stack_levels=1` to maximize predictive performance.\n",
      " |              To prevent overfitting, `num_bag_folds >= 2` must also be set or else a ValueError will be raised.\n",
      " |          delay_bag_sets : bool, default = False\n",
      " |              Controls when repeats of kfold bagging are executed in AutoGluon when under a time limit.\n",
      " |              We suggest sticking to `False` to avoid overfitting.\n",
      " |                  If True, AutoGluon delays repeating kfold bagging until after evaluating all models\n",
      " |                      from `hyperparameters`, if there is enough time. This allows AutoGluon to explore\n",
      " |                      more hyperparameters to obtain a better final performance but it may lead to\n",
      " |                      more overfitting.\n",
      " |                  If False, AutoGluon repeats kfold bagging immediately after evaluating each model.\n",
      " |                      Thus, AutoGluon might evaluate fewer models with less overfitting.\n",
      " |          holdout_frac : float, default = None\n",
      " |              Fraction of train_data to holdout as tuning data for optimizing hyperparameters (ignored unless `tuning_data = None`, ignored if `num_bag_folds != 0` unless `use_bag_holdout == True`).\n",
      " |              Default value (if None) is selected based on the number of rows in the training data. Default values range from 0.2 at 2,500 rows to 0.01 at 250,000 rows.\n",
      " |              Default value is doubled if `hyperparameter_tune_kwargs` is set, up to a maximum of 0.2.\n",
      " |              Disabled if `num_bag_folds >= 2` unless `use_bag_holdout == True`.\n",
      " |          use_bag_holdout : bool | str, default = False\n",
      " |              If True, a `holdout_frac` portion of the data is held-out from model bagging.\n",
      " |              This held-out data is only used to score models and determine weighted ensemble weights.\n",
      " |              Enable this if there is a large gap between score_val and score_test in stack models.\n",
      " |              Note: If `tuning_data` was specified, `tuning_data` is used as the holdout data.\n",
      " |              Disabled if not bagging.\n",
      " |              If \"auto\", will be set to True if the training data has >= 1000000 rows, else it will be set to False.\n",
      " |          hyperparameter_tune_kwargs : str or dict, default = None\n",
      " |              Hyperparameter tuning strategy and kwargs (for example, how many HPO trials to run).\n",
      " |              If None, then hyperparameter tuning will not be performed.\n",
      " |              You can either choose to provide a preset\n",
      " |                  Valid preset values:\n",
      " |                      'auto': Performs HPO via bayesian optimization search on NN_TORCH and FASTAI models, and random search on other models using local scheduler.\n",
      " |                      'random': Performs HPO via random search using local scheduler.\n",
      " |              Or provide a dict to specify searchers and schedulers\n",
      " |                  Valid keys:\n",
      " |                      'num_trials': How many HPO trials to run\n",
      " |                      'scheduler': Which scheduler to use\n",
      " |                          Valid values:\n",
      " |                              'local': Local scheduler that schedule trials FIFO\n",
      " |                      'searcher': Which searching algorithm to use\n",
      " |                          'local_random': Uses the 'random' searcher\n",
      " |                          'random': Perform random search\n",
      " |                          'auto': Perform bayesian optimization search on NN_TORCH and FASTAI models. Perform random search on other models.\n",
      " |                  The 'scheduler' and 'searcher' key are required when providing a dict.\n",
      " |                  An example of a valid dict:\n",
      " |                      hyperparameter_tune_kwargs = {\n",
      " |                          'num_trials': 5,\n",
      " |                          'scheduler' : 'local',\n",
      " |                          'searcher': 'auto',\n",
      " |                      }\n",
      " |          feature_prune_kwargs: dict, default = None\n",
      " |              Performs layer-wise feature pruning via recursive feature elimination with permutation feature importance.\n",
      " |              This fits all models in a stack layer once, discovers a pruned set of features, fits all models in the stack layer\n",
      " |              again with the pruned set of features, and updates input feature lists for models whose validation score improved.\n",
      " |              If None, do not perform feature pruning. If empty dictionary, perform feature pruning with default configurations.\n",
      " |              For valid dictionary keys, refer to :class:`autogluon.core.utils.feature_selection.FeatureSelector` and\n",
      " |              `autogluon.core.trainer.abstract_trainer.AbstractTrainer._proxy_model_feature_prune` documentation.\n",
      " |              To force all models to work with the pruned set of features, set force_prune=True in the dictionary.\n",
      " |          ag_args : dict, default = None\n",
      " |              Keyword arguments to pass to all models (i.e. common hyperparameters shared by all AutoGluon models).\n",
      " |              See the `ag_args` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      " |              Identical to specifying `ag_args` parameter for all models in `hyperparameters`.\n",
      " |              If a key in `ag_args` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      " |          ag_args_fit : dict, default = None\n",
      " |              Keyword arguments to pass to all models.\n",
      " |              See the `ag_args_fit` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      " |              Identical to specifying `ag_args_fit` parameter for all models in `hyperparameters`.\n",
      " |              If a key in `ag_args_fit` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      " |          ag_args_ensemble : dict, default = None\n",
      " |              Keyword arguments to pass to all models.\n",
      " |              See the `ag_args_ensemble` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      " |              Identical to specifying `ag_args_ensemble` parameter for all models in `hyperparameters`.\n",
      " |              If a key in `ag_args_ensemble` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      " |          ds_args : dict, see below for default\n",
      " |              Keyword arguments for dynamic stacking, only used if `dynamic_stacking=True`. These keyword arguments control the behavior of dynamic stacking\n",
      " |              and determine how AutoGluon tries to detect stacked overfitting. To detect stacked overfitting, AutoGluon will fit itself (so called sub-fits)\n",
      " |              on a subset (for holdout) or multiple subsets (for repeated cross-validation) and use the predictions of AutoGluon on the validation data to\n",
      " |              detect stacked overfitting. The sub-fits stop and stacking will be disabled if any sub-fit shows stacked overfitting.\n",
      " |              Allowed keys and values are:\n",
      " |                  `detection_time_frac` : float in (0,1), default = 1/4\n",
      " |                      Determines how much of the original training time is used for detecting stacked overfitting.\n",
      " |                      When using (repeated) cross-validation, each sub-fit will be fit for `1/n_splits * detection_time_frac * time_limit`.\n",
      " |                      If no time limit is given to AutoGluon, this parameter is ignored and AutoGluon is fit without a time limit in the sub-fit.\n",
      " |                  `validation_procedure`: str, default = 'holdout'\n",
      " |                      Determines the validation procedure used to detect stacked overfitting. Can be either `cv` or `holdout`.\n",
      " |                          If `validation_procedure='holdout'` and `holdout_data` is not specified (default), then `holdout_frac` determines the holdout data.\n",
      " |                          If `validation_procedure='holdout'` and `holdout_data` is specified, then the provided `holdout_data` is used for validation.\n",
      " |                          If `validation_procedure='cv'`, `n_folds` and `n_repeats` determine the kind cross-validation procedure.\n",
      " |                  `holdout_frac` : float in (0,1), default = 1/9\n",
      " |                      Determines how much of the original training data is used for the holdout data during holdout validation.\n",
      " |                      Ignored if `holdout_data` is not None.\n",
      " |                  `n_folds` : int in [2, +inf), default = 2\n",
      " |                      Number of folds to use for cross-validation.\n",
      " |                  `n_repeats` : int [1, +inf), default = 1\n",
      " |                      Number of repeats to use for repeated cross-validation. If set to 1, performs 1-repeated cross-validation which is equivalent to\n",
      " |                      cross-validation without repeats.\n",
      " |                  `memory_safe_fits` : bool, default = True\n",
      " |                      If True, AutoGluon runs each sub-fit in a ray-based subprocess to avoid memory leakage that exist due to Python's lackluster\n",
      " |                      garbage collector.\n",
      " |                  `clean_up_fits` : bool, default = True\n",
      " |                      If True, AutoGluon will remove all saved information from sub-fits from disk.\n",
      " |                      If False, the sub-fits are kept on disk and `self._sub_fits` will store paths to the sub-fits, which can be loaded just like any other\n",
      " |                      predictor from disk using `TabularPredictor.load()`.\n",
      " |                  `enable_ray_logging` : bool, default = True\n",
      " |                      If True, will log the dynamic stacking sub-fit when ray is used (`memory_safe_fits=True`).\n",
      " |                      Note that because of how ray works, this may cause extra unwanted logging in the main fit process after dynamic stacking completes.\n",
      " |                  `enable_callbacks` : bool, default = False\n",
      " |                      If True, will perform a deepcopy on the specified user callbacks and enable them during the DyStack call.\n",
      " |                      If False, will not include callbacks in the DyStack call.\n",
      " |                  `holdout_data`: str or :class:`pd.DataFrame`, default = None\n",
      " |                      Another dataset containing validation data reserved for detecting stacked overfitting. This dataset should be in the same format as\n",
      " |                      `train_data`. If str is passed, `holdout_data` will be loaded using the str value as the file path.\n",
      " |                      If `holdout_data` is not None, the sub-fit is fit on all of `train_data` and the full fit is fit on all of `train_data` and\n",
      " |                      `holdout_data` combined.\n",
      " |          included_model_types : list, default = None\n",
      " |              To only include listed model types for training during `fit()`.\n",
      " |              Models that are listed in `included_model_types` but not in `hyperparameters` will be ignored.\n",
      " |              Reference `hyperparameters` documentation for what models correspond to each value.\n",
      " |              Useful when only a subset of model needs to be trained and the `hyperparameters` dictionary is difficult or time-consuming.\n",
      " |                  Example: To include both 'GBM' and 'FASTAI' models, specify `included_model_types=['GBM', 'FASTAI']`.\n",
      " |          excluded_model_types : list, default = None\n",
      " |              Banned subset of model types to avoid training during `fit()`, even if present in `hyperparameters`.\n",
      " |              Reference `hyperparameters` documentation for what models correspond to each value.\n",
      " |              Useful when a particular model type such as 'KNN' or 'custom' is not desired but altering the `hyperparameters` dictionary is difficult or time-consuming.\n",
      " |                  Example: To exclude both 'KNN' and 'custom' models, specify `excluded_model_types=['KNN', 'custom']`.\n",
      " |          refit_full : bool or str, default = False\n",
      " |              Whether to retrain all models on all of the data (training + validation) after the normal training procedure.\n",
      " |              This is equivalent to calling `predictor.refit_full(model=refit_full)` after fit.\n",
      " |              If `refit_full=True`, it will be treated as `refit_full='all'`.\n",
      " |              If `refit_full=False`, refitting will not occur.\n",
      " |              Valid str values:\n",
      " |                  `all`: refits all models.\n",
      " |                  `best`: refits only the best model (and its ancestors if it is a stacker model).\n",
      " |                  `{model_name}`: refits only the specified model (and its ancestors if it is a stacker model).\n",
      " |              For bagged models:\n",
      " |                  Reduces a model's inference time by collapsing bagged ensembles into a single model fit on all of the training data.\n",
      " |                  This process will typically result in a slight accuracy reduction and a large inference speedup.\n",
      " |                  The inference speedup will generally be between 10-200x faster than the original bagged ensemble model.\n",
      " |                      The inference speedup factor is equivalent to (k * n), where k is the number of folds (`num_bag_folds`) and n is the number of finished repeats (`num_bag_sets`) in the bagged ensemble.\n",
      " |                  The runtime is generally 10% or less of the original fit runtime.\n",
      " |                      The runtime can be roughly estimated as 1 / (k * n) of the original fit runtime, with k and n defined above.\n",
      " |              For non-bagged models:\n",
      " |                  Optimizes a model's accuracy by retraining on 100% of the data without using a validation set.\n",
      " |                  Will typically result in a slight accuracy increase and no change to inference time.\n",
      " |                  The runtime will be approximately equal to the original fit runtime.\n",
      " |              This process does not alter the original models, but instead adds additional models.\n",
      " |              If stacker models are refit by this process, they will use the refit_full versions of the ancestor models during inference.\n",
      " |              Models produced by this process will not have validation scores, as they use all of the data for training.\n",
      " |                  Therefore, it is up to the user to determine if the models are of sufficient quality by including test data in `predictor.leaderboard(test_data)`.\n",
      " |                  If the user does not have additional test data, they should reference the original model's score for an estimate of the performance of the refit_full model.\n",
      " |                      Warning: Be aware that utilizing refit_full models without separately verifying on test data means that the model is untested, and has no guarantee of being consistent with the original model.\n",
      " |              The time taken by this process is not enforced by `time_limit`.\n",
      " |          save_bag_folds : bool, default = True\n",
      " |              If True, will save the bagged fold models to disk.\n",
      " |              If False, will not save the bagged fold models, only keeping their metadata and out-of-fold predictions.\n",
      " |                  Note: The bagged models will not be available for prediction, only use this if you intend to call `refit_full`.\n",
      " |                  The purpose of setting it to False is that it greatly decreases the peak disk usage of the predictor during the fit call when bagging.\n",
      " |                  Note that this makes refit_full slightly more likely to crash in scenarios where the dataset is large relative to available system memory.\n",
      " |                  This is because by default, refit_full will fall back to cloning the first fold of the bagged model in case it lacks memory to refit.\n",
      " |                  However, if `save_bag_folds=False`, this fallback isn't possible, as there is not fold model to clone because it wasn't saved.\n",
      " |                  In this scenario, refit will raise an exception for `save_bag_folds=False`, but will succeed if `save_bag_folds=True`.\n",
      " |              Final disk usage of predictor will be identical regardless of the setting after `predictor.delete_models(models_to_keep=\"best\", dry_run=False)` is called post-fit.\n",
      " |          set_best_to_refit_full : bool, default = False\n",
      " |              If True, will change the default model that Predictor uses for prediction when model is not specified to the refit_full version of the model that exhibited the highest validation score.\n",
      " |              Only valid if `refit_full` is set.\n",
      " |          keep_only_best : bool, default = False\n",
      " |              If True, only the best model and its ancestor models are saved in the outputted `predictor`. All other models are deleted.\n",
      " |                  If you only care about deploying the most accurate predictor with the smallest file-size and no longer need any of the other trained models or functionality beyond prediction on new data, then set: `keep_only_best=True`, `save_space=True`.\n",
      " |                  This is equivalent to calling `predictor.delete_models(models_to_keep='best', dry_run=False)` directly after `fit()`.\n",
      " |              If used with `refit_full` and `set_best_to_refit_full`, the best model will be the refit_full model, and the original bagged best model will be deleted.\n",
      " |                  `refit_full` will be automatically set to 'best' in this case to avoid training models which will be later deleted.\n",
      " |          save_space : bool, default = False\n",
      " |              If True, reduces the memory and disk size of predictor by deleting auxiliary model files that aren't needed for prediction on new data.\n",
      " |                  This is equivalent to calling `predictor.save_space()` directly after `fit()`.\n",
      " |              This has NO impact on inference accuracy.\n",
      " |              It is recommended if the only goal is to use the trained model for prediction.\n",
      " |              Certain advanced functionality may no longer be available if `save_space=True`. Refer to `predictor.save_space()` documentation for more details.\n",
      " |          feature_generator : :class:`autogluon.features.generators.AbstractFeatureGenerator`, default = :class:`autogluon.features.generators.AutoMLPipelineFeatureGenerator`\n",
      " |              The feature generator used by AutoGluon to process the input data to the form sent to the models. This often includes automated feature generation and data cleaning.\n",
      " |              It is generally recommended to keep the default feature generator unless handling an advanced use-case.\n",
      " |              To control aspects of the default feature generation process, you can pass in an :class:`AutoMLPipelineFeatureGenerator` object constructed using some of these kwargs:\n",
      " |                  enable_numeric_features : bool, default True\n",
      " |                      Whether to keep features of 'int' and 'float' raw types.\n",
      " |                      These features are passed without alteration to the models.\n",
      " |                      Appends IdentityFeatureGenerator(infer_features_in_args=dict(valid_raw_types=['int', 'float']))) to the generator group.\n",
      " |                  enable_categorical_features : bool, default True\n",
      " |                      Whether to keep features of 'object' and 'category' raw types.\n",
      " |                      These features are processed into memory optimized 'category' features.\n",
      " |                      Appends CategoryFeatureGenerator() to the generator group.\n",
      " |                  enable_datetime_features : bool, default True\n",
      " |                      Whether to keep features of 'datetime' raw type and 'object' features identified as 'datetime_as_object' features.\n",
      " |                      These features will be converted to 'int' features representing milliseconds since epoch.\n",
      " |                      Appends DatetimeFeatureGenerator() to the generator group.\n",
      " |                  enable_text_special_features : bool, default True\n",
      " |                      Whether to use 'object' features identified as 'text' features to generate 'text_special' features such as word count, capital letter ratio, and symbol counts.\n",
      " |                      Appends TextSpecialFeatureGenerator() to the generator group.\n",
      " |                  enable_text_ngram_features : bool, default True\n",
      " |                      Whether to use 'object' features identified as 'text' features to generate 'text_ngram' features.\n",
      " |                      Appends TextNgramFeatureGenerator(vectorizer=vectorizer) to the generator group.\n",
      " |                  enable_raw_text_features : bool, default False\n",
      " |                      Whether to keep the raw text features.\n",
      " |                      Appends IdentityFeatureGenerator(infer_features_in_args=dict(required_special_types=['text'])) to the generator group.\n",
      " |                  vectorizer : CountVectorizer, default CountVectorizer(min_df=30, ngram_range=(1, 3), max_features=10000, dtype=np.uint8)\n",
      " |                      sklearn CountVectorizer object to use in TextNgramFeatureGenerator.\n",
      " |                      Only used if `enable_text_ngram_features=True`.\n",
      " |          unlabeled_data : pd.DataFrame, default = None\n",
      " |              [Experimental Parameter]\n",
      " |              Collection of data without labels that we can use to pretrain on. This is the same schema as train_data, except\n",
      " |              without the labels. Currently, unlabeled_data is only used for pretraining a TabTransformer model.\n",
      " |              If you do not specify 'TRANSF' with unlabeled_data, then no pretraining will occur and unlabeled_data will be ignored!\n",
      " |              After the pretraining step, we will finetune using the TabTransformer model as well. If TabTransformer is ensembled\n",
      " |              with other models, like in typical AutoGluon fashion, then the output of this \"pretrain/finetune\" will be ensembled\n",
      " |              with other models, which will not used the unlabeled_data. The \"pretrain/finetune flow\" is also known as semi-supervised learning.\n",
      " |              The typical use case for unlabeled_data is to add signal to your model where you may not have sufficient training\n",
      " |              data. e.g. 500 hand-labeled samples (perhaps a hard human task), whole data set (unlabeled) is thousands/millions.\n",
      " |              However, this isn't the only use case. Given enough unlabeled data(millions of rows), you may see improvements\n",
      " |              to any amount of labeled data.\n",
      " |          verbosity : int\n",
      " |              If specified, overrides the existing `predictor.verbosity` value.\n",
      " |          raise_on_no_models_fitted: bool, default = True\n",
      " |              If True, will raise a RuntimeError if no models were successfully fit during `fit()`.\n",
      " |          calibrate: bool or str, default = 'auto'\n",
      " |              Note: It is recommended to use ['auto', False] as the values and avoid True.\n",
      " |              If 'auto' will automatically set to True if the problem_type and eval_metric are suitable for calibration.\n",
      " |              If True and the problem_type is classification, temperature scaling will be used to calibrate the Predictor's estimated class probabilities\n",
      " |              (which may improve metrics like log_loss) and will train a scalar parameter on the validation set.\n",
      " |              If True and the problem_type is quantile regression, conformalization will be used to calibrate the Predictor's estimated quantiles\n",
      " |              (which may improve the prediction interval coverage, and bagging could further improve it) and will compute a set of scalar parameters on the validation set.\n",
      " |          test_data : str or :class:`pd.DataFrame`, default = None\n",
      " |              Table of the test data.\n",
      " |              If str is passed, `test_data` will be loaded using the str value as the file path.\n",
      " |              NOTE: This test_data is NEVER SEEN by the model during training and, if specified, is only used for logging purposes (i.e. for learning curve generation).\n",
      " |              This test_data should be treated the same way test data is used in predictor.leaderboard.\n",
      " |          learning_curves : bool or dict, default = None\n",
      " |              If bool and is True, default learning curve hyperparameter ag_args will be initialized for each of the models included in the ensemble.\n",
      " |                  By default, learning curves will include eval_metric scores specified in fit call arguments.\n",
      " |                  This can be overwritten as shown below.\n",
      " |              If dict, user can pass learning_curves parameters to be initialized as ag_args in the following format:\n",
      " |                  learning_curves = {\n",
      " |                      \"metrics\": str or list(str) or Scorer or list(Scorer):\n",
      " |                          autogluon metric scorer(s) to be calculated at each iteration, represented as Scorer object(s) or scorer name(s) (str)\n",
      " |                      \"use_error\": bool : whether to use error or score format for metrics listed above\n",
      " |                  }\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`TabularPredictor` object. Returns self.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from autogluon.tabular import TabularDataset, TabularPredictor\n",
      " |      >>> train_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')\n",
      " |      >>> label = 'class'\n",
      " |      >>> predictor = TabularPredictor(label=label).fit(train_data)\n",
      " |      >>> test_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')\n",
      " |      >>> leaderboard = predictor.leaderboard(test_data)\n",
      " |      >>> y_test = test_data[label]\n",
      " |      >>> test_data = test_data.drop(columns=[label])\n",
      " |      >>> y_pred = predictor.predict(test_data)\n",
      " |      >>> perf = predictor.evaluate_predictions(y_true=y_test, y_pred=y_pred)\n",
      " |      \n",
      " |      To maximize predictive performance, use the following:\n",
      " |      \n",
      " |      >>> eval_metric = 'roc_auc'  # set this to the metric you ultimately care about\n",
      " |      >>> time_limit = 3600  # set as long as you are willing to wait (in sec)\n",
      " |      >>> predictor = TabularPredictor(label=label, eval_metric=eval_metric).fit(train_data, presets=['best_quality'], time_limit=time_limit)\n",
      " |  \n",
      " |  fit_extra(self, hyperparameters: 'str | Dict[str, Any]', time_limit: 'float' = None, base_model_names: 'List[str]' = None, fit_weighted_ensemble: 'bool' = True, fit_full_last_level_weighted_ensemble: 'bool' = True, full_weighted_ensemble_additionally: 'bool' = False, num_cpus: 'str | int' = 'auto', num_gpus: 'str | int' = 'auto', fit_strategy: \"Literal['auto', 'sequential', 'parallel']\" = 'auto', memory_limit: 'float | str' = 'auto', **kwargs) -> \"'TabularPredictor'\"\n",
      " |      Fits additional models after the original :meth:`TabularPredictor.fit` call.\n",
      " |      The original train_data and tuning_data will be used to train the models.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      hyperparameters : str or dict\n",
      " |          Refer to argument documentation in :meth:`TabularPredictor.fit`.\n",
      " |          If `base_model_names` is specified and hyperparameters is using the level-based key notation,\n",
      " |          the key of the level which directly uses the base models should be 1. The level in the hyperparameters\n",
      " |          dictionary is relative, not absolute.\n",
      " |      time_limit : int, default = None\n",
      " |          Refer to argument documentation in :meth:`TabularPredictor.fit`.\n",
      " |      base_model_names : List[str], default = None\n",
      " |          The names of the models to use as base models for this fit call.\n",
      " |          Base models will provide their out-of-fold predictions as additional features to the models in `hyperparameters`.\n",
      " |          If specified, all models trained will be stack ensembles.\n",
      " |          If None, models will be trained as if they were specified in :meth:`TabularPredictor.fit`, without depending on existing models.\n",
      " |          Only valid if bagging is enabled.\n",
      " |      fit_weighted_ensemble : bool, default = True\n",
      " |          If True, a WeightedEnsembleModel will be fit in each stack layer.\n",
      " |          A weighted ensemble will often be stronger than an individual model while being very fast to train.\n",
      " |          It is recommended to keep this value set to True to maximize predictive quality.\n",
      " |      fit_full_last_level_weighted_ensemble : bool, default = True\n",
      " |          If True, the WeightedEnsembleModel of the last stacking level will be fit with all (successful) models from all previous layers as base models.\n",
      " |          If stacking is disabled, settings this to True or False makes no difference because the WeightedEnsembleModel L2 always uses all models from L1.\n",
      " |          It is recommended to keep this value set to True to maximize predictive quality.\n",
      " |      full_weighted_ensemble_additionally : bool, default = False\n",
      " |          If True, AutoGluon will fit two WeightedEnsembleModels after training all stacking levels. Setting this to True, simulates calling\n",
      " |          `fit_weighted_ensemble()` after calling `fit()`. Has no affect if `fit_full_last_level_weighted_ensemble` is False and does not fit an additional\n",
      " |          WeightedEnsembleModel if stacking is disabled.\n",
      " |      num_cpus: int, default = \"auto\"\n",
      " |          The total amount of cpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of cpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      num_gpus: int, default = \"auto\"\n",
      " |          The total amount of gpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of gpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      fit_strategy: Literal[\"auto\", \"sequential\", \"parallel\"], default = \"auto\"\n",
      " |          The strategy used to fit models.\n",
      " |          If \"auto\", uses the same fit_strategy as used in the original :meth:`TabularPredictor.fit` call.\n",
      " |          If \"sequential\", models will be fit sequentially. This is the most stable option with the most readable logging.\n",
      " |          If \"parallel\", models will be fit in parallel with ray, splitting available compute between them.\n",
      " |              Note: \"parallel\" is experimental and may run into issues. It was first added in version 1.2.0.\n",
      " |          For machines with 16 or more CPU cores, it is likely that \"parallel\" will be faster than \"sequential\".\n",
      " |      \n",
      " |          .. versionadded:: 1.2.0\n",
      " |      \n",
      " |      memory_limit: float | str, default = \"auto\"\n",
      " |          The total amount of memory in GB you want AutoGluon predictor to use. \"auto\" means AutoGluon will use all available memory on the system\n",
      " |          (that is detectable by psutil).\n",
      " |          Note that this is only a soft limit! AutoGluon uses this limit to skip training models that are expected to require too much memory or stop\n",
      " |          training a model that would exceed the memory limit. AutoGluon does not guarantee the enforcement of this limit (yet). Nevertheless, we expect\n",
      " |          AutoGluon to abide by the limit in most cases or, at most, go over the limit by a small margin.\n",
      " |          For most virtualized systems (e.g., in the cloud) and local usage on a server or laptop, \"auto\" is ideal for this parameter. We recommend manually\n",
      " |          setting the memory limit (and any other resources) on systems with shared resources that are controlled by the operating system (e.g., SLURM and\n",
      " |          cgroups). Otherwise, AutoGluon might wrongly assume more resources are available for fitting a model than the operating system allows,\n",
      " |          which can result in model training failing or being very inefficient.\n",
      " |      **kwargs :\n",
      " |          Refer to kwargs documentation in :meth:`TabularPredictor.fit`.\n",
      " |          Note that the following kwargs are not available in `fit_extra` as they cannot be changed from their values set in `fit()`:\n",
      " |              [`holdout_frac`, `num_bag_folds`, `auto_stack`, `feature_generator`, `unlabeled_data`]\n",
      " |          Moreover, `dynamic_stacking` is also not available in `fit_extra` as the detection of stacked overfitting is only supported at the first fit time.\n",
      " |          pseudo_data : pd.DataFrame, default = None\n",
      " |              Data that has been self labeled by Autogluon model and will be incorporated into training during 'fit_extra'\n",
      " |  \n",
      " |  fit_pseudolabel(self, pseudo_data: 'pd.DataFrame', max_iter: 'int' = 3, return_pred_prob: 'bool' = False, use_ensemble: 'bool' = False, fit_ensemble: 'bool' = False, fit_ensemble_every_iter: 'bool' = False, **kwargs)\n",
      " |      [Advanced] Uses additional data (`pseudo_data`) to try to achieve better model quality.\n",
      " |      Pseudo data can come either with or without the `label` column.\n",
      " |      \n",
      " |      If `pseudo_data` is labeled, then models will be refit using the `pseudo_data` as additional training data.\n",
      " |      If bagging, each fold of the bagged ensemble will use all the `pseudo_data` as additional training data.\n",
      " |      `pseudo_data` will never be used for validation/scoring.\n",
      " |      \n",
      " |      If the data is unlabeled, such as providing the batched test data without ground truth available, then transductive learning is leveraged.\n",
      " |      In transductive learning, the existing predictor will predict on `pseudo_data`\n",
      " |      to identify the most confident rows (For example all rows with predictive probability above 95%).\n",
      " |      These rows will then be pseudo-labelled, given the label of the most confident class.\n",
      " |      The pseudo-labelled rows will then be used as additional training data when fitting the models.\n",
      " |      Then, if `max_iter > 1`, this process can repeat itself, using the new models to predict on the unused `pseudo_data` rows\n",
      " |      to see if any new rows should be used in the next iteration as training data.\n",
      " |      We recommend specifying `return_pred_prob=True` if the data is unlabeled to get the correct prediction probabilities on the `pseudo_data`,\n",
      " |      rather than calling `predictor.predict_proba(pseudo_data)`.\n",
      " |      \n",
      " |      For example:\n",
      " |          Original fit: 10000 `train_data` rows with 10-fold bagging\n",
      " |              Bagged fold models will use 9000 `train_data` rows for training, and 1000 for validation.\n",
      " |          `fit_pseudolabel` is called with 5000 row labelled `pseudo_data`.\n",
      " |              Bagged fold models are then fit again with `_PSEUDO` suffix.\n",
      " |              10000 train_data rows with 10-fold bagging + 5000 `pseudo_data` rows.\n",
      " |              Bagged fold models will use 9000 `train_data` rows + 5000 `pseudo_data` rows = 14000 rows for training, and 1000 for validation.\n",
      " |                  Note: The same validation rows will be used as was done in the original fit, so that validation scores are directly comparable.\n",
      " |          Alternatively, `fit_pseduolabel` is called with 5000 rows unlabelled `pseudo_data`.\n",
      " |              Predictor predicts on the `pseudo_data`, finds 965 rows with confident predictions.\n",
      " |              Set the ground truth of those 965 rows as the most confident prediction.\n",
      " |              Bagged fold models are then fit with `_PSEUDO` suffix.\n",
      " |              10000 train_data rows with 10-fold bagging + 965 labelled `pseudo_data` rows.\n",
      " |              Bagged fold models will use 9000 `train_data` rows + 965 `pseudo_data` rows = 9965 rows for training, and 1000 for validation.\n",
      " |                  Note: The same validation rows will be used as was done in the original fit, so that validation scores are directly comparable.\n",
      " |              Repeat the process using the new pseudo-labelled predictor on the remaining `pseudo_data`.\n",
      " |              In the example, lets assume 188 new `pseudo_data` rows have confident predictions.\n",
      " |              Now the total labelled `pseudo_data` rows is 965 + 188 = 1153.\n",
      " |              Then repeat the process, up to `max_iter` times: ex 10000 train_data rows with 10-fold bagging + 1153 `pseudo_data` rows.\n",
      " |              Early stopping will trigger if validation score improvement is not observed.\n",
      " |      \n",
      " |      Note: pseudo_data is only used for L1 models. Support for L2+ models is not yet implemented. L2+ models will only use the original train_data.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      pseudo_data : :class:`pd.DataFrame`\n",
      " |          Extra data to incorporate into training. Pre-labeled test data allowed. If no labels\n",
      " |          then pseudo-labeling algorithm will predict and filter out which rows to incorporate into training\n",
      " |      max_iter: int, default = 3\n",
      " |          Maximum iterations of pseudo-labeling allowed\n",
      " |      return_pred_prob: bool, default = False\n",
      " |          Returns held-out predictive probabilities from pseudo-labeling. If test_data is labeled then\n",
      " |          returns model's predictive probabilities.\n",
      " |      use_ensemble: bool, default = False\n",
      " |          If True will use ensemble pseudo labeling algorithm. If False will just use best model\n",
      " |          for pseudo labeling algorithm.\n",
      " |      fit_ensemble: bool, default = False\n",
      " |          If True with fit weighted ensemble model using combination of best models.\n",
      " |          Fitting weighted ensemble will be done after fitting has\n",
      " |          been completed unless otherwise specified. If False will not fit weighted ensemble\n",
      " |          over models trained with pseudo labeling and models trained without it.\n",
      " |      fit_ensemble_every_iter: bool, default = False\n",
      " |          If True fits weighted ensemble model for every iteration of pseudo labeling algorithm. If False\n",
      " |          and fit_ensemble is True will fit after all pseudo labeling training is done.\n",
      " |      **kwargs:\n",
      " |          If predictor is not already fit, then kwargs are for the functions 'fit' and 'fit_extra':\n",
      " |          Refer to parameters documentation in :meth:`TabularPredictor.fit`.\n",
      " |          Refer to parameters documentation in :meth:`TabularPredictor.fit_extra`.\n",
      " |          If predictor is fit kwargs are for 'fit_extra':\n",
      " |          Refer to parameters documentation in :meth:`TabularPredictor.fit_extra`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      self : TabularPredictor\n",
      " |          Returns self, which is a Python class of TabularPredictor\n",
      " |  \n",
      " |  fit_summary(self, verbosity: 'int' = 3, show_plot: 'bool' = False) -> 'dict'\n",
      " |      Output summary of information about models produced during `fit()`.\n",
      " |      May create various generated summary plots and store them in folder: `predictor.path`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      verbosity : int, default = 3\n",
      " |          Controls how detailed of a summary to output.\n",
      " |          Set <= 0 for no output printing, 1 to print just high-level summary,\n",
      " |          2 to print summary and create plots, >= 3 to print all information produced during `fit()`.\n",
      " |      show_plot : bool, default = False\n",
      " |          If True, shows the model summary plot in browser when verbosity > 1.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Dict containing various detailed information. We do not recommend directly printing this dict as it may be very large.\n",
      " |  \n",
      " |  fit_weighted_ensemble(self, base_models: 'list' = None, name_suffix: 'str' = 'Best', expand_pareto_frontier: 'bool' = False, time_limit: 'float' = None, refit_full: 'bool' = False, num_cpus: 'int | str' = 'auto', num_gpus: 'int | str' = 'auto')\n",
      " |      Fits new weighted ensemble models to combine predictions of previously-trained models.\n",
      " |      `cache_data` must have been set to `True` during the original training to enable this functionality.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      base_models: list, default = None\n",
      " |          List of model names the weighted ensemble can consider as candidates.\n",
      " |          If None, all previously trained models are considered except for weighted ensemble models.\n",
      " |          As an example, to train a weighted ensemble that can only have weights assigned to the models 'model_a' and 'model_b', set `base_models=['model_a', 'model_b']`\n",
      " |      name_suffix: str, default = 'Best'\n",
      " |          Name suffix to add to the name of the newly fitted ensemble model.\n",
      " |      expand_pareto_frontier: bool, default = False\n",
      " |          If True, will train N-1 weighted ensemble models instead of 1, where `N=len(base_models)`.\n",
      " |          The final model trained when True is equivalent to the model trained when False.\n",
      " |          These weighted ensemble models will attempt to expand the pareto frontier.\n",
      " |          This will create many different weighted ensembles which have different accuracy/memory/inference-speed trade-offs.\n",
      " |          This is particularly useful when inference speed is an important consideration.\n",
      " |      time_limit: float, default = None\n",
      " |          Time in seconds each weighted ensemble model is allowed to train for. If `expand_pareto_frontier=True`, the `time_limit` value is applied to each model.\n",
      " |          If None, the ensemble models train without time restriction.\n",
      " |      refit_full : bool, default = False\n",
      " |          If True, will apply refit_full to all weighted ensembles created during this call.\n",
      " |          Identical to calling `predictor.refit_full(model=predictor.fit_weighted_ensemble(...))`\n",
      " |      num_cpus: int | str, default = \"auto\"\n",
      " |          The total amount of cpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of cpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      num_gpus: int | str, default = \"auto\"\n",
      " |          The total amount of gpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of gpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List of newly trained weighted ensemble model names.\n",
      " |      If an exception is encountered while training an ensemble model, that model's name will be absent from the list.\n",
      " |  \n",
      " |  info(self)\n",
      " |      [EXPERIMENTAL] Returns a dictionary of `predictor` metadata.\n",
      " |      Warning: This functionality is currently in preview mode.\n",
      " |          The metadata information returned may change in structure in future versions without warning.\n",
      " |          The definitions of various metadata values are not yet documented.\n",
      " |          The output of this function should not be used for programmatic decisions.\n",
      " |      Contains information such as row count, column count, model training time, validation scores, hyperparameters, and much more.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Dictionary of `predictor` metadata.\n",
      " |  \n",
      " |  leaderboard(self, data: 'pd.DataFrame | str | None' = None, extra_info: 'bool' = False, extra_metrics: 'list | None' = None, decision_threshold: 'float | None' = None, score_format: 'str' = 'score', only_pareto_frontier: 'bool' = False, skip_score: 'bool' = False, refit_full: 'bool | None' = None, set_refit_score_to_parent: 'bool' = False, display: 'bool' = False, **kwargs) -> 'pd.DataFrame'\n",
      " |      Output summary of information about models produced during `fit()` as a :class:`pd.DataFrame`.\n",
      " |      Includes information on test and validation scores for all models, model training times, inference times, and stack levels.\n",
      " |      Output DataFrame columns include:\n",
      " |          'model': The name of the model.\n",
      " |      \n",
      " |          'score_val': The validation score of the model on the 'eval_metric'.\n",
      " |              NOTE: Metrics scores always show in higher is better form.\n",
      " |              This means that metrics such as log_loss and root_mean_squared_error will have their signs FLIPPED, and values will be negative.\n",
      " |              This is necessary to avoid the user needing to know the metric to understand if higher is better when looking at leaderboard.\n",
      " |          'eval_metric': The evaluation metric name used to calculate the scores.\n",
      " |              This should be identical to `predictor.eval_metric.name`.\n",
      " |          'pred_time_val': The inference time required to compute predictions on the validation data end-to-end.\n",
      " |              Equivalent to the sum of all 'pred_time_val_marginal' values for the model and all of its base models.\n",
      " |          'fit_time': The fit time required to train the model end-to-end (Including base models if the model is a stack ensemble).\n",
      " |              Equivalent to the sum of all 'fit_time_marginal' values for the model and all of its base models.\n",
      " |          'pred_time_val_marginal': The inference time required to compute predictions on the validation data (Ignoring inference times for base models).\n",
      " |              Note that this ignores the time required to load the model into memory when bagging is disabled.\n",
      " |          'fit_time_marginal': The fit time required to train the model (Ignoring base models).\n",
      " |          'stack_level': The stack level of the model.\n",
      " |              A model with stack level N can take any set of models with stack level less than N as input, with stack level 1 models having no model inputs.\n",
      " |          'can_infer': If model is able to perform inference on new data. If False, then the model either was not saved, was deleted, or an ancestor of the model cannot infer.\n",
      " |              `can_infer` is often False when `save_bag_folds=False` was specified in initial `fit()`.\n",
      " |          'fit_order': The order in which models were fit. The first model fit has `fit_order=1`, and the Nth model fit has `fit_order=N`. The order corresponds to the first child model fit in the case of bagged ensembles.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str or :class:`pd.DataFrame` (optional)\n",
      " |          This dataset must also contain the label-column with the same column-name as specified during fit().\n",
      " |          If extra_metrics=None and skip_score=True, then the label column is not required.\n",
      " |          If specified, then the leaderboard returned will contain additional columns 'score_test', 'pred_time_test', and 'pred_time_test_marginal'.\n",
      " |              'score_test': The score of the model on the 'eval_metric' for the data provided.\n",
      " |                  NOTE: Metrics scores always show in higher is better form.\n",
      " |                  This means that metrics such as log_loss and root_mean_squared_error will have their signs FLIPPED, and values will be negative.\n",
      " |                  This is necessary to avoid the user needing to know the metric to understand if higher is better when looking at leaderboard.\n",
      " |              'pred_time_test': The true end-to-end wall-clock inference time of the model for the data provided.\n",
      " |                  Equivalent to the sum of all 'pred_time_test_marginal' values for the model and all of its base models.\n",
      " |              'pred_time_test_marginal': The inference time of the model for the data provided, minus the inference time for the model's base models, if it has any.\n",
      " |                  Note that this ignores the time required to load the model into memory when bagging is disabled.\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |      extra_info : bool, default = False\n",
      " |          If `True`, will return extra columns with advanced info.\n",
      " |          This requires additional computation as advanced info data is calculated on demand.\n",
      " |          Additional output columns when `extra_info=True` include:\n",
      " |              'num_features': Number of input features used by the model.\n",
      " |                  Some models may ignore certain features in the preprocessed data.\n",
      " |              'num_models': Number of models that actually make up this \"model\" object.\n",
      " |                  For non-bagged models, this is 1. For bagged models, this is equal to the number of child models (models trained on bagged folds) the bagged ensemble contains.\n",
      " |              'num_models_w_ancestors': Equivalent to the sum of 'num_models' values for the model and its' ancestors (see below).\n",
      " |              'memory_size': The amount of memory in bytes the model requires when persisted in memory. This is not equivalent to the amount of memory the model may use during inference.\n",
      " |                  For bagged models, this is the sum of the 'memory_size' of all child models.\n",
      " |              'memory_size_w_ancestors': Equivalent to the sum of 'memory_size' values for the model and its' ancestors.\n",
      " |                  This is the amount of memory required to avoid loading any models in-between inference calls to get predictions from this model.\n",
      " |                  For online-inference, this is critical. It is important that the machine performing online inference has memory more than twice this value to avoid loading models for every call to inference by persisting models in memory.\n",
      " |              'memory_size_min': The amount of memory in bytes the model minimally requires to perform inference.\n",
      " |                  For non-bagged models, this is equivalent to 'memory_size'.\n",
      " |                  For bagged models, this is equivalent to the largest child model's 'memory_size_min'.\n",
      " |                  To minimize memory usage, child models can be loaded and un-persisted one by one to infer. This is the default behavior if a bagged model was not already persisted in memory prior to inference.\n",
      " |              'memory_size_min_w_ancestors': Equivalent to the max of the 'memory_size_min' values for the model and its' ancestors.\n",
      " |                  This is the minimum required memory to infer with the model by only loading one model at a time, as each of its ancestors will also have to be loaded into memory.\n",
      " |                  For offline-inference where latency is not a concern, this should be used to determine the required memory for a machine if 'memory_size_w_ancestors' is too large.\n",
      " |              'num_ancestors': Number of ancestor models for the given model.\n",
      " |      \n",
      " |              'num_descendants': Number of descendant models for the given model.\n",
      " |      \n",
      " |              'model_type': The type of the given model.\n",
      " |                  If the model is an ensemble type, 'child_model_type' will indicate the inner model type. A stack ensemble of bagged LightGBM models would have 'StackerEnsembleModel' as its model type.\n",
      " |              'child_model_type': The child model type. None if the model is not an ensemble. A stack ensemble of bagged LightGBM models would have 'LGBModel' as its child type.\n",
      " |                  child models are models which are used as a group to generate a given bagged ensemble model's predictions. These are the models trained on each fold of a bagged ensemble.\n",
      " |                  For 10-fold bagging, the bagged ensemble model would have 10 child models.\n",
      " |                  For 10-fold bagging with 3 repeats, the bagged ensemble model would have 30 child models.\n",
      " |                  Note that child models are distinct from ancestors and descendants.\n",
      " |              'hyperparameters': The hyperparameter values specified for the model.\n",
      " |                  All hyperparameters that do not appear in this dict remained at their default values.\n",
      " |              'hyperparameters_fit': The hyperparameters set by the model during fit.\n",
      " |                  This overrides the 'hyperparameters' value for a particular key if present in 'hyperparameters_fit' to determine the fit model's final hyperparameters.\n",
      " |                  This is most commonly set for hyperparameters that indicate model training iterations or epochs, as early stopping can find a different value from what 'hyperparameters' indicated.\n",
      " |                  In these cases, the provided hyperparameter in 'hyperparameters' is used as a maximum for the model, but the model is still able to early stop at a smaller value during training to achieve a better validation score or to satisfy time constraints.\n",
      " |                  For example, if a NN model was given `epochs=500` as a hyperparameter, but found during training that `epochs=60` resulted in optimal validation score, it would use `epoch=60` and `hyperparameters_fit={'epoch': 60}` would be set.\n",
      " |              'ag_args_fit': Special AutoGluon arguments that influence model fit.\n",
      " |                  See the documentation of the `hyperparameters` argument in `TabularPredictor.fit()` for more information.\n",
      " |              'features': List of feature names used by the model.\n",
      " |      \n",
      " |              'child_hyperparameters': Equivalent to 'hyperparameters', but for the model's children.\n",
      " |      \n",
      " |              'child_hyperparameters_fit': Equivalent to 'hyperparameters_fit', but for the model's children.\n",
      " |      \n",
      " |              'child_ag_args_fit': Equivalent to 'ag_args_fit', but for the model's children.\n",
      " |      \n",
      " |              'ancestors': The model's ancestors. Ancestor models are the models which are required to make predictions during the construction of the model's input features.\n",
      " |                  If A is an ancestor of B, then B is a descendant of A.\n",
      " |                  If a model's ancestor is deleted, the model is no longer able to infer on new data, and its 'can_infer' value will be False.\n",
      " |                  A model can only have ancestor models whose 'stack_level' are lower than itself.\n",
      " |                  'stack_level'=1 models have no ancestors.\n",
      " |              'descendants': The model's descendants. Descendant models are the models which require this model to make predictions during the construction of their input features.\n",
      " |                  If A is a descendant of B, then B is an ancestor of A.\n",
      " |                  If this model is deleted, then all descendant models will no longer be able to infer on new data, and their 'can_infer' values will be False.\n",
      " |                  A model can only have descendant models whose 'stack_level' are higher than itself.\n",
      " |      extra_metrics : list, default = None\n",
      " |          A list of metrics to calculate scores for and include in the output DataFrame.\n",
      " |          Only valid when `data` is specified. The scores refer to the scores on `data` (same data as used to calculate the `score_test` column).\n",
      " |          This list can contain any values which would also be valid for `eval_metric` in predictor init.\n",
      " |          For example, `extra_metrics=['accuracy', 'roc_auc', 'log_loss']` would be valid in binary classification.\n",
      " |          This example would return 3 additional columns in the output DataFrame, whose column names match the names of the metrics.\n",
      " |          Passing `extra_metrics=[predictor.eval_metric]` would return an extra column in the name of the eval metric that has identical values to `score_test`.\n",
      " |          This also works with custom metrics. If passing an object instead of a string, the column name will be equal to the `.name` attribute of the object.\n",
      " |          NOTE: Metrics scores always show in higher is better form.\n",
      " |          This means that metrics such as log_loss and root_mean_squared_error will have their signs FLIPPED, and values will be negative.\n",
      " |          This is necessary to avoid the user needing to know the metric to understand if higher is better when looking at leaderboard.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold to use when converting prediction probabilities to predictions.\n",
      " |          This will impact the scores of metrics such as `f1` and `accuracy`.\n",
      " |          If None, defaults to `predictor.decision_threshold`. Ignored unless `problem_type='binary'`.\n",
      " |          Refer to the `predictor.decision_threshold` docstring for more information.\n",
      " |          NOTE: `score_val` will not be impacted by this value in v0.8.\n",
      " |              `score_val` will always show the validation scores achieved with a decision threshold of `0.5`.\n",
      " |              Only test scores will be properly updated.\n",
      " |      score_format : {'score', 'error'}\n",
      " |          If \"score\", leaderboard is returned as normal.\n",
      " |          If \"error\", the column \"score_val\" is converted to \"metric_error_val\", and \"score_test\" is converted to \"metric_error_test\".\n",
      " |              \"metric_error\" is calculated by taking `predictor.eval_metric.convert_score_to_error(score)`.\n",
      " |              This will result in errors where 0 is perfect and lower is better.\n",
      " |      only_pareto_frontier : bool, default = False\n",
      " |          If `True`, only return model information of models in the Pareto frontier of the accuracy/latency trade-off (models which achieve the highest score within their end-to-end inference time).\n",
      " |          At minimum this will include the model with the highest score and the model with the lowest inference time.\n",
      " |          This is useful when deciding which model to use during inference if inference time is a consideration.\n",
      " |          Models filtered out by this process would never be optimal choices for a user that only cares about model inference time and score.\n",
      " |      skip_score : bool, default = False\n",
      " |          [Advanced, primarily for developers]\n",
      " |          If `True`, will skip computing `score_test` if `data` is specified. `score_test` will be set to NaN for all models.\n",
      " |          `pred_time_test` and related columns will still be computed.\n",
      " |      refit_full : bool, default = None\n",
      " |          If True, will return only models that have been refit (ex: have `_FULL` in the name).\n",
      " |          If False, will return only models that have not been refit.\n",
      " |          If None, will return all models.\n",
      " |      set_refit_score_to_parent : bool, default = False\n",
      " |          If True, the `score_val` of refit models will be set to the `score_val` of their parent.\n",
      " |          While this does not represent the genuine validation score of the refit model, it is a reasonable proxy.\n",
      " |      display : bool, default = False\n",
      " |          If True, the output DataFrame is printed to stdout.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.DataFrame` of model performance summary information.\n",
      " |  \n",
      " |  learning_curves(self) -> 'Tuple[dict, dict]'\n",
      " |      Retrieves learning curves generated during predictor.fit().\n",
      " |      Will not work if the learning_curves flag was not set during training.\n",
      " |      Note that learning curves are only generated for iterative learners with\n",
      " |      learning curve support.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      None\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      metadata: dict\n",
      " |          A dictionary containing metadata related to the training process.\n",
      " |      \n",
      " |      model_data: dict\n",
      " |          A dictionary containing the learning curves across all models.\n",
      " |          To see curve_data format, refer to AbstractModel's save_learning_curves() method.\n",
      " |              {\n",
      " |                  \"model\": curve_data,\n",
      " |                  \"model\": curve_data,\n",
      " |                  \"model\": curve_data,\n",
      " |                  \"model\": curve_data,\n",
      " |              }\n",
      " |  \n",
      " |  load_data_internal(self, data='train', return_X=True, return_y=True)\n",
      " |      Loads the internal data representation used during model training.\n",
      " |      Individual AutoGluon models like the neural network may apply additional feature transformations that are not reflected in this method.\n",
      " |      This method only applies universal transforms employed by all AutoGluon models.\n",
      " |      Warning, the internal representation may:\n",
      " |          Have different features compared to the original data.\n",
      " |          Have different row counts compared to the original data.\n",
      " |          Have indices which do not align with the original data.\n",
      " |          Have label values which differ from those in the original data.\n",
      " |      Internal data representations should NOT be combined with the original data, in most cases this is not possible.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str, default = 'train'\n",
      " |          The data to load.\n",
      " |          Valid values are:\n",
      " |              'train':\n",
      " |                  Load the training data used during model training.\n",
      " |                  This is a transformed and augmented version of the `train_data` passed in `fit()`.\n",
      " |              'val':\n",
      " |                  Load the validation data used during model training.\n",
      " |                  This is a transformed and augmented version of the `tuning_data` passed in `fit()`.\n",
      " |                  If `tuning_data=None` was set in `fit()`, then `tuning_data` is an automatically generated validation set created by splitting `train_data`.\n",
      " |                  Warning: Will raise an exception if called by a bagged predictor, as bagged predictors have no validation data.\n",
      " |      return_X : bool, default = True\n",
      " |          Whether to return the internal data features\n",
      " |          If set to `False`, then the first element in the returned tuple will be None.\n",
      " |      return_y : bool, default = True\n",
      " |          Whether to return the internal data labels\n",
      " |          If set to `False`, then the second element in the returned tuple will be None.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Tuple of (:class:`pd.DataFrame`, :class:`pd.Series`) corresponding to the internal data features and internal data labels, respectively.\n",
      " |  \n",
      " |  model_failures(self, verbose: 'bool' = False) -> 'pd.DataFrame'\n",
      " |      [Advanced] Get the model failures that occurred during the fitting of this model, in the form of a pandas DataFrame.\n",
      " |      \n",
      " |      This is useful for in-depth debugging of model failures and identifying bugs.\n",
      " |      \n",
      " |      For more information on model failures, refer to `predictor.info()['model_info_failures']`\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      verbose: bool, default = False\n",
      " |          If True, the output DataFrame is printed to stdout.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      model_failures_df: pd.DataFrame\n",
      " |          A DataFrame of model failures. Each row corresponds to a model failure, and columns correspond to meta information about that model.\n",
      " |      \n",
      " |          Included Columns:\n",
      " |              \"model\": The name of the model that failed\n",
      " |              \"exc_type\": The class name of the exception raised\n",
      " |              \"total_time\": The total time in seconds taken by the model prior to the exception (lost time due to the failure)\n",
      " |              \"model_type\": The class name of the model\n",
      " |              \"child_model_type\": The child class name of the model\n",
      " |              \"is_initialized\"\n",
      " |              \"is_fit\"\n",
      " |              \"is_valid\"\n",
      " |              \"can_infer\"\n",
      " |              \"num_features\"\n",
      " |              \"num_models\"\n",
      " |              \"memory_size\"\n",
      " |              \"hyperparameters\"\n",
      " |              \"hyperparameters_fit\"\n",
      " |              \"child_hyperparameters\"\n",
      " |              \"child_hyperparameters_fit\"\n",
      " |              \"exc_str\": The string message contained in the raised exception\n",
      " |              \"exc_traceback\": The full traceback message of the exception as a string\n",
      " |              \"exc_order\": The order of the model failure (starting from 1)\n",
      " |  \n",
      " |  model_names(self, stack_name: 'str' = None, level: 'int' = None, can_infer: 'bool' = None, models: 'List[str]' = None, persisted: 'bool' = None) -> 'List[str]'\n",
      " |      Returns the list of model names trained in this `predictor` object.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      stack_name: str, default = None\n",
      " |          If specified, returns only models under a given stack name.\n",
      " |      level: int, default = None\n",
      " |          If specified, returns only models at the given stack level.\n",
      " |      can_infer: bool, default = None\n",
      " |          If specified, returns only models that can/cannot infer on new data.\n",
      " |      models: List[str], default = None\n",
      " |          The list of model names to consider.\n",
      " |          If None, considers all models.\n",
      " |      persisted: bool, default = None\n",
      " |          If None: no filtering will occur based on persisted status\n",
      " |          If True: will return only the models that are persisted in memory via `predictor.persist()`\n",
      " |          If False: will return only the models that are not persisted in memory via `predictor.persist()`\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List of model names\n",
      " |  \n",
      " |  model_refit_map(self, inverse=False) -> 'Dict[str, str]'\n",
      " |      Returns a dictionary of original model name -> refit full model name.\n",
      " |      Empty unless `refit_full=True` was set during fit or `predictor.refit_full()` was called.\n",
      " |      This can be useful when determining the best model based off of `predictor.leaderboard()`, then getting the _FULL version of the model by passing its name as the key to this dictionary.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      inverse : bool, default = False\n",
      " |          If True, instead returns a dictionary of refit full model name -> original model name (Swap keys with values)\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Dictionary of original model name -> refit full model name.\n",
      " |  \n",
      " |  persist(self, models='best', with_ancestors=True, max_memory=0.4) -> 'List[str]'\n",
      " |      Persist models in memory for reduced inference latency. This is particularly important if the models are being used for online-inference where low latency is critical.\n",
      " |      If models are not persisted in memory, they are loaded from disk every time they are asked to make predictions.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      models : list of str or str, default = 'best'\n",
      " |          Model names of models to persist.\n",
      " |          If 'best' then the model with the highest validation score is persisted (this is the model used for prediction by default).\n",
      " |          If 'all' then all models are persisted.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |      with_ancestors : bool, default = True\n",
      " |          If True, all ancestor models of the provided models will also be persisted.\n",
      " |          If False, stacker models will not have the models they depend on persisted unless those models were specified in `models`. This will slow down inference as the ancestor models will still need to be loaded from disk for each predict call.\n",
      " |          Only relevant for stacker models.\n",
      " |      max_memory : float, default = 0.4\n",
      " |          Proportion of total available memory to allow for the persisted models to use.\n",
      " |          If the models' summed memory usage requires a larger proportion of memory than max_memory, they are not persisted. In this case, the output will be an empty list.\n",
      " |          If None, then models are persisted regardless of estimated memory usage. This can cause out-of-memory errors.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List of persisted model names.\n",
      " |  \n",
      " |  plot_ensemble_model(self, model: 'str' = 'best', *, prune_unused_nodes: 'bool' = True, filename: 'str' = 'ensemble_model.png') -> 'str'\n",
      " |      Output the visualized stack ensemble architecture of a model trained by `fit()`.\n",
      " |      The plot is stored to a file, `ensemble_model.png` in folder `predictor.path` (or by the name specified in `filename`)\n",
      " |      \n",
      " |      This function requires `graphviz` and `pygraphviz` to be installed because this visualization depends on those package.\n",
      " |      Unless this function will raise `ImportError` without being able to generate the visual of the ensemble model.\n",
      " |      \n",
      " |      To install the required package, run the below commands (for Ubuntu linux):\n",
      " |      \n",
      " |      $ sudo apt-get install graphviz graphviz-dev\n",
      " |      $ pip install pygraphviz\n",
      " |      \n",
      " |      For other platforms, refer to https://graphviz.org/ for Graphviz install, and https://pygraphviz.github.io/ for PyGraphviz.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      model : str, default 'best'\n",
      " |          The model to highlight in golden orange, with all component models highlighted in yellow.\n",
      " |          If 'best', will default to the best model returned from `self.model_best`\n",
      " |      prune_unused_nodes : bool, default True\n",
      " |          If True, only plot the models that are components of the specified `model`.\n",
      " |          If False, will plot all models.\n",
      " |      filename : str, default 'ensemble_model.png'\n",
      " |          The filename to save the plot as. Will be located under the `self.path` folder.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      The file name with the full path to the saved graphic on disk.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from autogluon.tabular import TabularDataset, TabularPredictor\n",
      " |      >>> train_data = TabularDataset('train.csv')\n",
      " |      >>> predictor = TabularPredictor(label='class').fit(train_data)\n",
      " |      >>> path_to_png = predictor.plot_ensemble_model()\n",
      " |      >>>\n",
      " |      >>> # To view the plot inside a Jupyter Notebook, use the below code:\n",
      " |      >>> from IPython.display import Image, display\n",
      " |      >>> display(Image(filename=path_to_png))\n",
      " |  \n",
      " |  predict(self, data: 'pd.DataFrame | str', model: 'str | None' = None, as_pandas: 'bool' = True, transform_features: 'bool' = True, *, decision_threshold: 'float | None' = None) -> 'pd.Series | np.ndarray'\n",
      " |      Use trained models to produce predictions of `label` column values for new data.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : :class:`pd.DataFrame` or str\n",
      " |          The data to make predictions for. Should contain same column names as training data and follow same format\n",
      " |          (may contain extra columns that won't be used by Predictor, including the label-column itself).\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |      model : str (optional)\n",
      " |          The name of the model to get predictions from. Defaults to None, which uses the highest scoring model on the validation set.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`\n",
      " |      as_pandas : bool, default = True\n",
      " |          Whether to return the output as a :class:`pd.Series` (True) or :class:`np.ndarray` (False).\n",
      " |      transform_features : bool, default = True\n",
      " |          If True, preprocesses data before predicting with models.\n",
      " |          If False, skips global feature preprocessing.\n",
      " |              This is useful to save on inference time if you have already called `data = predictor.transform_features(data)`.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold used to convert prediction probabilities to predictions.\n",
      " |          Only relevant for binary classification, otherwise ignored.\n",
      " |          If None, defaults to `predictor.decision_threshold`.\n",
      " |          Valid values are in the range [0.0, 1.0]\n",
      " |          You can obtain an optimized `decision_threshold` by first calling `predictor.calibrate_decision_threshold()`.\n",
      " |          Useful to set for metrics such as `balanced_accuracy` and `f1` as `0.5` is often not an optimal threshold.\n",
      " |          Predictions are calculated via the following logic on the positive class: `1 if pred > decision_threshold else 0`\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Array of predictions, one corresponding to each row in given dataset. Either :class:`np.ndarray` or :class:`pd.Series` depending on `as_pandas` argument.\n",
      " |  \n",
      " |  predict_from_proba(self, y_pred_proba: 'pd.DataFrame | np.ndarray', decision_threshold: 'float | None' = None) -> 'pd.Series | np.array'\n",
      " |      Given prediction probabilities, convert to predictions.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      y_pred_proba : :class:`pd.DataFrame` or :class:`np.ndarray`\n",
      " |          The prediction probabilities to convert to predictions.\n",
      " |          Obtainable via the output of `predictor.predict_proba`.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold used to convert prediction probabilities to predictions.\n",
      " |          Only relevant for binary classification, otherwise ignored.\n",
      " |          If None, defaults to `predictor.decision_threshold`.\n",
      " |          Valid values are in the range [0.0, 1.0]\n",
      " |          You can obtain an optimized `decision_threshold` by first calling `predictor.calibrate_decision_threshold()`.\n",
      " |          Useful to set for metrics such as `balanced_accuracy` and `f1` as `0.5` is often not an optimal threshold.\n",
      " |          Predictions are calculated via the following logic on the positive class: `1 if pred > decision_threshold else 0`\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Array of predictions, one corresponding to each row in given dataset. Either :class:`np.ndarray` or :class:`pd.Series` depending on `y_pred_proba` dtype.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from autogluon.tabular import TabularPredictor\n",
      " |      >>> predictor = TabularPredictor(label='class').fit('train.csv', label='class')\n",
      " |      >>> y_pred_proba = predictor.predict_proba('test.csv')\n",
      " |      >>>\n",
      " |      >>> # y_pred and y_pred_from_proba are identical\n",
      " |      >>> y_pred = predictor.predict('test.csv')\n",
      " |      >>> y_pred_from_proba = predictor.predict_from_proba(y_pred_proba=y_pred_proba)\n",
      " |  \n",
      " |  predict_multi(self, data: 'pd.DataFrame' = None, models: 'List[str]' = None, as_pandas: 'bool' = True, transform_features: 'bool' = True, inverse_transform: 'bool' = True, *, decision_threshold: 'float' = None) -> 'dict[str, pd.Series] | dict[str, np.ndarray]'\n",
      " |      Returns a dictionary of predictions where the key is\n",
      " |      the model name and the value is the model's prediction probabilities on the data.\n",
      " |      \n",
      " |      Equivalent output to:\n",
      " |      ```\n",
      " |      predict_dict = {}\n",
      " |      for m in models:\n",
      " |          predict_dict[m] = predictor.predict(data, model=m)\n",
      " |      ```\n",
      " |      \n",
      " |      Note that this will generally be much faster than calling :meth:`TabularPredictor.predict` separately for each model\n",
      " |      because this method leverages the model dependency graph to avoid redundant computation.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : DataFrame, default = None\n",
      " |          The data to predict on.\n",
      " |          If None:\n",
      " |              If self.has_val, the validation data is used.\n",
      " |              Else, prediction is skipped and the out-of-fold (OOF) predictions are returned, equivalent to:\n",
      " |              ```\n",
      " |              predict_dict = {}\n",
      " |              for m in models:\n",
      " |                  predict_dict[m] = predictor.predict_oof(model=m)\n",
      " |              ```\n",
      " |      models : List[str], default = None\n",
      " |          The list of models to get predictions for.\n",
      " |          If None, all models that can infer are used.\n",
      " |      as_pandas : bool, default = True\n",
      " |          Whether to return the output of each model as a :class:`pd.Series` (True) or :class:`np.ndarray` (False).\n",
      " |      transform_features : bool, default = True\n",
      " |          If True, preprocesses data before predicting with models.\n",
      " |          If False, skips global feature preprocessing.\n",
      " |              This is useful to save on inference time if you have already called `data = predictor.transform_features(data)`.\n",
      " |      inverse_transform : bool, default = True\n",
      " |          If True, will return predictions in the original format.\n",
      " |          If False (advanced), will return predictions in AutoGluon's internal format.\n",
      " |      decision_threshold : float, default = None\n",
      " |          The decision threshold used to convert prediction probabilities to predictions.\n",
      " |          Only relevant for binary classification, otherwise ignored.\n",
      " |          If None, defaults to `0.5`.\n",
      " |          Valid values are in the range [0.0, 1.0]\n",
      " |          You can obtain an optimized `decision_threshold` by first calling :meth:`TabularPredictor.calibrate_decision_threshold`.\n",
      " |          Useful to set for metrics such as `balanced_accuracy` and `f1` as `0.5` is often not an optimal threshold.\n",
      " |          Predictions are calculated via the following logic on the positive class: `1 if pred > decision_threshold else 0`\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      dict[str, pd.Series] | dict[str, np.ndarray]\n",
      " |          Dictionary with model names as keys and model predictions as values.\n",
      " |  \n",
      " |  predict_oof(self, model: 'str' = None, *, transformed=False, train_data=None, internal_oof=False, decision_threshold=None, can_infer=None) -> 'pd.Series'\n",
      " |      Note: This is advanced functionality not intended for normal usage.\n",
      " |      \n",
      " |      Returns the out-of-fold (OOF) predictions for every row in the training data.\n",
      " |      \n",
      " |      For a similar method, refer to :meth:`TabularPredictor.predict_multi` with `data=None`.\n",
      " |      For more information, refer to `predict_proba_oof()` documentation.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      model : str (optional)\n",
      " |          Refer to `predict_proba_oof()` documentation.\n",
      " |      transformed : bool, default = False\n",
      " |          Refer to `predict_proba_oof()` documentation.\n",
      " |      train_data : pd.DataFrame, default = None\n",
      " |          Refer to `predict_proba_oof()` documentation.\n",
      " |      internal_oof : bool, default = False\n",
      " |          Refer to `predict_proba_oof()` documentation.\n",
      " |      decision_threshold : float, default = None\n",
      " |          Refer to `predict_multi` documentation.\n",
      " |      can_infer : bool, default = None\n",
      " |          Refer to `predict_proba_oof()` documentation.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.Series` object of the out-of-fold training predictions of the model.\n",
      " |  \n",
      " |  predict_proba(self, data: 'pd.DataFrame | str', model: 'str | None' = None, as_pandas: 'bool' = True, as_multiclass: 'bool' = True, transform_features: 'bool' = True) -> 'pd.DataFrame | pd.Series | np.ndarray'\n",
      " |      Use trained models to produce predicted class probabilities rather than class-labels (if task is classification).\n",
      " |      If `predictor.problem_type` is regression or quantile, this will raise an AssertionError.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : :class:`pd.DataFrame` or str\n",
      " |          The data to make predictions for. Should contain same column names as training dataset and follow same format\n",
      " |          (may contain extra columns that won't be used by Predictor, including the label-column itself).\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |      model : str (optional)\n",
      " |          The name of the model to get prediction probabilities from. Defaults to None, which uses the highest scoring model on the validation set.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |      as_pandas : bool, default = True\n",
      " |          Whether to return the output as a pandas object (True) or numpy array (False).\n",
      " |          Pandas object is a DataFrame if this is a multiclass problem or `as_multiclass=True`, otherwise it is a Series.\n",
      " |          If the output is a DataFrame, the column order will be equivalent to `predictor.class_labels`.\n",
      " |      as_multiclass : bool, default = True\n",
      " |          Whether to return binary classification probabilities as if they were for multiclass classification.\n",
      " |              Output will contain two columns, and if `as_pandas=True`, the column names will correspond to the binary class labels.\n",
      " |              The columns will be the same order as `predictor.class_labels`.\n",
      " |          If False, output will contain only 1 column for the positive class (get positive_class name via `predictor.positive_class`).\n",
      " |          Only impacts output for binary classification problems.\n",
      " |      transform_features : bool, default = True\n",
      " |          If True, preprocesses data before predicting with models.\n",
      " |          If False, skips global feature preprocessing.\n",
      " |              This is useful to save on inference time if you have already called `data = predictor.transform_features(data)`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Array of predicted class-probabilities, corresponding to each row in the given data.\n",
      " |      May be a :class:`np.ndarray` or :class:`pd.DataFrame` / :class:`pd.Series` depending on `as_pandas` and `as_multiclass` arguments and the type of prediction problem.\n",
      " |      For binary classification problems, the output contains for each datapoint the predicted probabilities of the negative and positive classes, unless you specify `as_multiclass=False`.\n",
      " |  \n",
      " |  predict_proba_multi(self, data: 'pd.DataFrame' = None, models: 'List[str]' = None, as_pandas: 'bool' = True, as_multiclass: 'bool' = True, transform_features: 'bool' = True, inverse_transform: 'bool' = True) -> 'dict[str, pd.DataFrame] | dict[str, pd.Series] | dict[str, np.ndarray]'\n",
      " |      Returns a dictionary of prediction probabilities where the key is\n",
      " |      the model name and the value is the model's prediction probabilities on the data.\n",
      " |      \n",
      " |      Equivalent output to:\n",
      " |      ```\n",
      " |      predict_proba_dict = {}\n",
      " |      for m in models:\n",
      " |          predict_proba_dict[m] = predictor.predict_proba(data, model=m)\n",
      " |      ```\n",
      " |      \n",
      " |      Note that this will generally be much faster than calling :meth:`TabularPredictor.predict_proba` separately for each model\n",
      " |      because this method leverages the model dependency graph to avoid redundant computation.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str or DataFrame, default = None\n",
      " |          The data to predict on.\n",
      " |          If None:\n",
      " |              If self.has_val, the validation data is used.\n",
      " |              Else, prediction is skipped and the out-of-fold (OOF) prediction probabilities are returned, equivalent to:\n",
      " |              ```\n",
      " |              predict_proba_dict = {}\n",
      " |              for m in models:\n",
      " |                  predict_proba_dict[m] = predictor.predict_proba_oof(model=m)\n",
      " |              ```\n",
      " |      models : List[str], default = None\n",
      " |          The list of models to get predictions for.\n",
      " |          If None, all models that can infer are used.\n",
      " |      as_pandas : bool, default = True\n",
      " |          Whether to return the output of each model as a pandas object (True) or numpy array (False).\n",
      " |          Pandas object is a :class:`pd.DataFrame` if this is a multiclass problem or `as_multiclass=True`, otherwise it is a :class:`pd.Series`.\n",
      " |          If the output is a :class:`pd.DataFrame`, the column order will be equivalent to `predictor.classes_`.\n",
      " |      as_multiclass : bool, default = True\n",
      " |          Whether to return binary classification probabilities as if they were for multiclass classification.\n",
      " |              Output will contain two columns, and if `as_pandas=True`, the column names will correspond to the binary class labels.\n",
      " |              The columns will be the same order as `predictor.class_labels`.\n",
      " |          If False, output will contain only 1 column for the positive class (get positive_class name via `predictor.positive_class`).\n",
      " |          Only impacts output for binary classification problems.\n",
      " |      transform_features : bool, default = True\n",
      " |          If True, preprocesses data before predicting with models.\n",
      " |          If False, skips global feature preprocessing.\n",
      " |              This is useful to save on inference time if you have already called `data = predictor.transform_features(data)`.\n",
      " |      inverse_transform : bool, default = True\n",
      " |          If True, will return prediction probabilities in the original format.\n",
      " |          If False (advanced), will return prediction probabilities in AutoGluon's internal format.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      dict\n",
      " |          Dictionary with model names as keys and model prediction probabilities as values.\n",
      " |  \n",
      " |  predict_proba_oof(self, model: 'str' = None, *, transformed=False, as_multiclass=True, train_data=None, internal_oof=False, can_infer=None) -> 'pd.DataFrame | pd.Series'\n",
      " |      Note: This is advanced functionality not intended for normal usage.\n",
      " |      \n",
      " |      Returns the out-of-fold (OOF) predicted class probabilities for every row in the training data.\n",
      " |      OOF prediction probabilities may provide unbiased estimates of generalization accuracy (reflecting how predictions will behave on new data)\n",
      " |      Predictions for each row are only made using models that were fit to a subset of data where this row was held-out.\n",
      " |      \n",
      " |      For a similar method, refer to :meth:`TabularPredictor.predict_proba_multi` with `data=None`.\n",
      " |      \n",
      " |      Warning: This method will raise an exception if called on a model that is not a bagged ensemble. Only bagged models (such a stacker models) can produce OOF predictions.\n",
      " |          This also means that refit_full models and distilled models will raise an exception.\n",
      " |      Warning: If intending to join the output of this method with the original training data, be aware that a rare edge-case issue exists:\n",
      " |          Multiclass problems with rare classes combined with the use of the 'log_loss' eval_metric may have forced AutoGluon to duplicate rows in the training data to satisfy minimum class counts in the data.\n",
      " |          If this has occurred, then the indices and row counts of the returned :class:`pd.Series` in this method may not align with the training data.\n",
      " |          In this case, consider fetching the processed training data using `predictor.load_data_internal()` instead of using the original training data.\n",
      " |          A more benign version of this issue occurs when 'log_loss' wasn't specified as the eval_metric but rare classes were dropped by AutoGluon.\n",
      " |          In this case, not all original training data rows will have an OOF prediction. It is recommended to either drop these rows during the join or to get direct predictions on the missing rows via :meth:`TabularPredictor.predict_proba`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      model : str (optional)\n",
      " |          The name of the model to get out-of-fold predictions from. Defaults to None, which uses the highest scoring model on the validation set.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`\n",
      " |      transformed : bool, default = False\n",
      " |          Whether the output values should be of the original label representation (False) or the internal label representation (True).\n",
      " |          The internal representation for binary and multiclass classification are integers numbering the k possible classes from 0 to k-1, while the original representation is identical to the label classes provided during fit.\n",
      " |          Generally, most users will want the original representation and keep `transformed=False`.\n",
      " |      as_multiclass : bool, default = True\n",
      " |          Whether to return binary classification probabilities as if they were for multiclass classification.\n",
      " |              Output will contain two columns, and if `transformed=False`, the column names will correspond to the binary class labels.\n",
      " |              The columns will be the same order as `predictor.class_labels`.\n",
      " |          If False, output will contain only 1 column for the positive class (get positive_class name via `predictor.positive_class`).\n",
      " |          Only impacts output for binary classification problems.\n",
      " |      train_data : pd.DataFrame, default = None\n",
      " |          Specify the original `train_data` to ensure that any training rows that were originally dropped internally are properly handled.\n",
      " |          If None, then output will not contain all rows if training rows were dropped internally during fit.\n",
      " |          If `train_data` is specified and `model` is unable to predict and rows were dropped internally, an exception will be raised.\n",
      " |      internal_oof : bool, default = False\n",
      " |          [Advanced Option] Return the internal OOF preds rather than the externally facing OOF preds.\n",
      " |          Internal OOF preds may have more/fewer rows than was provided in train_data, and are incompatible with external data.\n",
      " |          If you don't know what this does, keep it as False.\n",
      " |      can_infer : bool, default = None\n",
      " |          Only used if `model` is not specified.\n",
      " |          This is used to determine if the best model must be one that is able to predict on new data (True).\n",
      " |          If None, the best model does not need to be able to infer on new data.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.Series` or :class:`pd.DataFrame` object of the out-of-fold training prediction probabilities of the model.\n",
      " |  \n",
      " |  refit_full(self, model: 'str | List[str]' = 'all', set_best_to_refit_full: 'bool' = True, train_data_extra: 'pd.DataFrame' = None, num_cpus: 'int | str' = 'auto', num_gpus: 'int | str' = 'auto', fit_strategy: \"Literal['auto', 'sequential', 'parallel']\" = 'auto', **kwargs) -> 'Dict[str, str]'\n",
      " |      Retrain model on all of the data (training + validation).\n",
      " |      For bagged models:\n",
      " |          Optimizes a model's inference time by collapsing bagged ensembles into a single model fit on all of the training data.\n",
      " |          This process will typically result in a slight accuracy reduction and a large inference speedup.\n",
      " |          The inference speedup will generally be between 10-200x faster than the original bagged ensemble model.\n",
      " |              The inference speedup factor is equivalent to (k * n), where k is the number of folds (`num_bag_folds`) and n is the number of finished repeats (`num_bag_sets`) in the bagged ensemble.\n",
      " |          The runtime is generally 10% or less of the original fit runtime.\n",
      " |              The runtime can be roughly estimated as 1 / (k * n) of the original fit runtime, with k and n defined above.\n",
      " |      For non-bagged models:\n",
      " |          Optimizes a model's accuracy by retraining on 100% of the data without using a validation set.\n",
      " |          Will typically result in a slight accuracy increase and no change to inference time.\n",
      " |          The runtime will be approximately equal to the original fit runtime.\n",
      " |      This process does not alter the original models, but instead adds additional models.\n",
      " |      If stacker models are refit by this process, they will use the refit_full versions of the ancestor models during inference.\n",
      " |      Models produced by this process will not have validation scores, as they use all of the data for training.\n",
      " |          Therefore, it is up to the user to determine if the models are of sufficient quality by including test data in `predictor.leaderboard(test_data)`.\n",
      " |          If the user does not have additional test data, they should reference the original model's score for an estimate of the performance of the refit_full model.\n",
      " |              Warning: Be aware that utilizing refit_full models without separately verifying on test data means that the model is untested, and has no guarantee of being consistent with the original model.\n",
      " |      `cache_data` must have been set to `True` during the original training to enable this functionality.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      model : str | List[str], default = 'all'\n",
      " |          Model name of model(s) to refit.\n",
      " |              If 'all' then all models are refitted.\n",
      " |              If 'best' then the model with the highest validation score is refit.\n",
      " |          All ancestor models will also be refit in the case that the selected model is a weighted or stacker ensemble.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |      set_best_to_refit_full : bool | str, default = True\n",
      " |          If True, sets best model to the refit_full version of the prior best model.\n",
      " |          This means the model used when `predictor.predict(data)` is called will be the refit_full version instead of the original version of the model.\n",
      " |          Ignored if `model` is not the best model.\n",
      " |          If str, interprets as a model name and sets best model to the refit_full version of the model `set_best_to_refit_full`.\n",
      " |      train_data_extra : pd.DataFrame, default = None\n",
      " |          If specified, will be used as additional rows of training data when refitting models.\n",
      " |          Requires label column. Will only be used for L1 models.\n",
      " |      num_cpus: int | str, default = \"auto\"\n",
      " |          The total amount of cpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of cpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      num_gpus: int | str, default = \"auto\"\n",
      " |          The total amount of gpus you want AutoGluon predictor to use.\n",
      " |          Auto means AutoGluon will make the decision based on the total number of gpus available and the model requirement for best performance.\n",
      " |          Users generally don't need to set this value\n",
      " |      fit_strategy: Literal[\"auto\", \"sequential\", \"parallel\"], default = \"auto\"\n",
      " |          The strategy used to fit models.\n",
      " |          If \"auto\", uses the same fit_strategy as used in the original :meth:`TabularPredictor.fit` call.\n",
      " |          If \"sequential\", models will be fit sequentially. This is the most stable option with the most readable logging.\n",
      " |          If \"parallel\", models will be fit in parallel with ray, splitting available compute between them.\n",
      " |              Note: \"parallel\" is experimental and may run into issues. It was first added in version 1.2.0.\n",
      " |          For machines with 16 or more CPU cores, it is likely that \"parallel\" will be faster than \"sequential\".\n",
      " |      \n",
      " |          .. versionadded:: 1.2.0\n",
      " |      \n",
      " |      **kwargs\n",
      " |          [Advanced] Developer debugging arguments.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Dictionary of original model names -> refit_full model names.\n",
      " |  \n",
      " |  save(self, silent: 'bool' = False)\n",
      " |      Save this Predictor to file in directory specified by this Predictor's `path`.\n",
      " |      Note that :meth:`TabularPredictor.fit` already saves the predictor object automatically\n",
      " |      (we do not recommend modifying the Predictor object yourself as it tracks many trained models).\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      silent : bool, default = False\n",
      " |          Whether to save without logging a message.\n",
      " |  \n",
      " |  save_space(self, remove_data=True, remove_fit_stack=True, requires_save=True, reduce_children=False)\n",
      " |      Reduces the memory and disk size of predictor by deleting auxiliary model files that aren't needed for prediction on new data.\n",
      " |      This function has NO impact on inference accuracy.\n",
      " |      It is recommended to invoke this method if the only goal is to use the trained model for prediction.\n",
      " |      However, certain advanced functionality may no longer be available after `save_space()` has been called.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      remove_data : bool, default = True\n",
      " |          Whether to remove cached files of the original training and validation data.\n",
      " |          Only reduces disk usage, it has no impact on memory usage.\n",
      " |          This is especially useful when the original data was large.\n",
      " |          This is equivalent to setting `cache_data=False` during the original `fit()`.\n",
      " |              Will disable all advanced functionality that requires `cache_data=True`.\n",
      " |      remove_fit_stack : bool, default = True\n",
      " |          Whether to remove information required to fit new stacking models and continue fitting bagged models with new folds.\n",
      " |          Only reduces disk usage, it has no impact on memory usage.\n",
      " |          This includes:\n",
      " |              out-of-fold (OOF) predictions\n",
      " |          This is useful for multiclass problems with many classes, as OOF predictions can become very large on disk. (1 GB per model in extreme cases)\n",
      " |          This disables `predictor.refit_full()` for stacker models.\n",
      " |      requires_save : bool, default = True\n",
      " |          Whether to remove information that requires the model to be saved again to disk.\n",
      " |          Typically this only includes flag variables that don't have significant impact on memory or disk usage, but should technically be updated due to the removal of more important information.\n",
      " |              An example is the `is_data_saved` boolean variable in `trainer`, which should be updated to `False` if `remove_data=True` was set.\n",
      " |      reduce_children : bool, default = False\n",
      " |          Whether to apply the reduction rules to bagged ensemble children models. These are the models trained for each fold of the bagged ensemble.\n",
      " |          This should generally be kept as `False` since the most important memory and disk reduction techniques are automatically applied to these models during the original `fit()` call.\n",
      " |  \n",
      " |  set_decision_threshold(self, decision_threshold: 'float')\n",
      " |      Set `predictor.decision_threshold`. Problem type must be 'binary', and the value must be between 0 and 1.\n",
      " |  \n",
      " |  set_model_best(self, model: 'str', save_trainer: 'bool' = False)\n",
      " |      Sets the model to be used by default when calling `predictor.predict(data)`.\n",
      " |      By default, this is the model with the best validation score, but this is not always the case.\n",
      " |      If manually set, this can be overwritten internally if further training occurs, such as through fit_extra, refit_full, or distill.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      model : str\n",
      " |          Name of model to set to best. If model does not exist or cannot infer, raises an AssertionError.\n",
      " |      save_trainer : bool, default = False\n",
      " |          If True, self._trainer is saved with the new model_best value, such that it is reflected when predictor is loaded in future from disk.\n",
      " |  \n",
      " |  simulation_artifact(self, test_data: 'pd.DataFrame' = None) -> 'dict'\n",
      " |      [Advanced] Computes and returns the necessary information to perform zeroshot HPO simulation.\n",
      " |      For a usage example, refer to https://github.com/autogluon/tabrepo/blob/main/examples/run_quickstart_from_scratch.py\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      test_data: pd.DataFrame, default = None\n",
      " |          The test data to predict with.\n",
      " |          If None, the keys `pred_proba_dict_test` and `y_test` will not be present in the output.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      simulation_dict: dict\n",
      " |          The dictionary of information required for zeroshot HPO simulation.\n",
      " |          Keys are as follows:\n",
      " |              pred_proba_dict_val: Dictionary of model name to prediction probabilities (or predictions if regression) on the validation data\n",
      " |              pred_proba_dict_test: Dictionary of model name to prediction probabilities (or predictions if regression) on the test data\n",
      " |              y_val: Pandas Series of ground truth labels for the validation data (internal representation)\n",
      " |              y_test: Pandas Series of ground truth labels for the test data (internal representation)\n",
      " |              eval_metric: The string name of the evaluation metric (obtained via `predictor.eval_metric.name`)\n",
      " |              problem_type: The problem type (obtained via `predictor.problem_type`)\n",
      " |              problem_type_transform: The transformed (internal) problem type (obtained via `predictor._learner.label_cleaner.problem_type_transform,`)\n",
      " |              ordered_class_labels: The original class labels (`predictor._learner.label_cleaner.ordered_class_labels`)\n",
      " |              ordered_Class_labels_transformed: The transformed (internal) class labels (`predictor._learner.label_cleaner.ordered_class_labels_transformed`)\n",
      " |              num_classes: The number of internal classes (`self._learner.label_cleaner.num_classes`)\n",
      " |              label: The label column name (`predictor.label`)\n",
      " |  \n",
      " |  transform_features(self, data: 'pd.DataFrame | str' = None, model: 'str' = None, base_models: 'list[str]' = None, return_original_features: 'bool' = True) -> 'pd.DataFrame'\n",
      " |      Transforms data features through the AutoGluon feature generator.\n",
      " |      This is useful to gain an understanding of how AutoGluon interprets the data features.\n",
      " |      The output of this function can be used to train further models, even outside of AutoGluon.\n",
      " |      This can be useful for training your own models on the same data representation as AutoGluon.\n",
      " |      Individual AutoGluon models like the neural network may apply additional feature transformations that are not reflected in this method.\n",
      " |      This method only applies universal transforms employed by all AutoGluon models.\n",
      " |      When `data=None`, `base_models=[{best_model}], and bagging was enabled during fit():\n",
      " |          This returns the out-of-fold predictions of the best model, which can be used as training input to a custom user stacker model.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data: :class:`pd.DataFrame` or str (optional)\n",
      " |          The data to apply feature transformation to.\n",
      " |          This data does not require the label column.\n",
      " |          If str is passed, `data` will be loaded using the str value as the file path.\n",
      " |          If not specified, the original data used during fit() will be used if fit() was previously called with `cache_data=True`. Otherwise, an exception will be raised.\n",
      " |              For non-bagged mode predictors:\n",
      " |                  The data used when not specified is the validation set.\n",
      " |                  This can either be an automatically generated validation set or the user-defined `tuning_data` if passed during fit().\n",
      " |                  If all parameters are unspecified, then the output is equivalent to `predictor.load_data_internal(data='val', return_X=True, return_y=False)[0]`.\n",
      " |                  To get the label values of the output, call `predictor.load_data_internal(data='val', return_X=False, return_y=True)[1]`.\n",
      " |                  If the original training set is desired, it can be passed in through `data`.\n",
      " |                      Warning: Do not pass the original training set if `model` or `base_models` are set. This will result in overfit feature transformation.\n",
      " |              For bagged mode predictors:\n",
      " |                  The data used when not specified is the full training set.\n",
      " |                  If all parameters are unspecified, then the output is equivalent to `predictor.load_data_internal(data='train', return_X=True, return_y=False)[0]`.\n",
      " |                  To get the label values of the output, call `predictor.load_data_internal(data='train', return_X=False, return_y=True)[1]`.\n",
      " |                  `base_model` features generated in this instance will be from out-of-fold predictions.\n",
      " |                  Note that the training set may differ from the training set originally passed during fit(), as AutoGluon may choose to drop or duplicate rows during training.\n",
      " |                  Warning: Do not pass the original training set through `data` if `model` or `base_models` are set. This will result in overfit feature transformation. Instead set `data=None`.\n",
      " |      model: str, default = None\n",
      " |          Model to generate input features for.\n",
      " |          The output data will be equivalent to the input data that would be sent into `model.predict_proba(data)`.\n",
      " |              Note: This only applies to cases where `data` is not the training data.\n",
      " |          If `None`, then only return generically preprocessed features prior to any model fitting.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |          Specifying a `refit_full` model will cause an exception if `data=None`.\n",
      " |          `base_models=None` is a requirement when specifying `model`.\n",
      " |      base_models: List[str], default = None\n",
      " |          List of model names to use as base_models for a hypothetical stacker model when generating input features.\n",
      " |          If `None`, then only return generically preprocessed features prior to any model fitting.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names()`.\n",
      " |          If a stacker model S exists with `base_models=M`, then setting `base_models=M` is equivalent to setting `model=S`.\n",
      " |          `model=None` is a requirement when specifying `base_models`.\n",
      " |      return_original_features: bool, default = True\n",
      " |          Whether to return the original features.\n",
      " |          If False, only returns the additional output columns from specifying `model` or `base_models`.\n",
      " |              This is useful to set to False if the intent is to use the output as input to further stacker models without the original features.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.DataFrame` of the provided `data` after feature transformation has been applied.\n",
      " |      This output does not include the label column, and will remove it if present in the supplied `data`.\n",
      " |      If a transformed label column is desired, use `predictor.transform_labels`.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> from autogluon.tabular import TabularPredictor\n",
      " |      >>> predictor = TabularPredictor(label='class').fit('train.csv', label='class', auto_stack=True)  # predictor is in bagged mode.\n",
      " |      >>> model = 'WeightedEnsemble_L2'\n",
      " |      >>> train_data_transformed = predictor.transform_features(model=model)  # Internal training DataFrame used as input to `model.fit()` for each model trained in predictor.fit()`\n",
      " |      >>> test_data_transformed = predictor.transform_features('test.csv', model=model)  # Internal test DataFrame used as input to `model.predict_proba()` during `predictor.predict_proba(test_data, model=model)`\n",
      " |  \n",
      " |  transform_labels(self, labels: 'np.ndarray | pd.Series', inverse: 'bool' = False, proba: 'bool' = False) -> 'pd.Series | pd.DataFrame'\n",
      " |      Transforms data labels to the internal label representation.\n",
      " |      This can be useful for training your own models on the same data label representation as AutoGluon.\n",
      " |      Regression problems do not differ between original and internal representation, and thus this method will return the provided labels.\n",
      " |      Warning: When `inverse=False`, it is possible for the output to contain NaN label values in multiclass problems if the provided label was dropped during training.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      labels: :class:`np.ndarray` or :class:`pd.Series`\n",
      " |          Labels to transform.\n",
      " |          If `proba=False`, an example input would be the output of `predictor.predict(test_data)`.\n",
      " |          If `proba=True`, an example input would be the output of `predictor.predict_proba(test_data, as_multiclass=False)`.\n",
      " |      inverse: bool, default = False\n",
      " |          When `True`, the input labels are treated as being in the internal representation and the original representation is outputted.\n",
      " |      proba: bool, default = False\n",
      " |          When `True`, the input labels are treated as probabilities and the output will be the internal representation of probabilities.\n",
      " |              In this case, it is expected that `labels` be a :class:`pd.DataFrame` or :class:`np.ndarray`.\n",
      " |              If the `problem_type` is multiclass:\n",
      " |                  The input column order must be equal to `predictor.class_labels`.\n",
      " |                  The output column order will be equal to `predictor.class_labels_internal`.\n",
      " |                  if `inverse=True`, the same logic applies, but with input and output columns interchanged.\n",
      " |          When `False`, the input labels are treated as actual labels and the output will be the internal representation of the labels.\n",
      " |              In this case, it is expected that `labels` be a :class:`pd.Series` or :class:`np.ndarray`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      :class:`pd.Series` of labels if `proba=False` or :class:`pd.DataFrame` of label probabilities if `proba=True`.\n",
      " |  \n",
      " |  unpersist(self, models='all') -> 'List[str]'\n",
      " |      Unpersist models in memory for reduced memory usage.\n",
      " |      If models are not persisted in memory, they are loaded from disk every time they are asked to make predictions.\n",
      " |      Note: Another way to reset the predictor and unpersist models is to reload the predictor from disk via `predictor = TabularPredictor.load(predictor.path)`.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      models : list of str or str, default = 'all'\n",
      " |          Model names of models to unpersist.\n",
      " |          If 'all' then all models are unpersisted.\n",
      " |          Valid models are listed in this `predictor` by calling `predictor.model_names(persisted=True)`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List of unpersisted model names.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  load(path: 'str', verbosity: 'int' = None, require_version_match: 'bool' = True, require_py_version_match: 'bool' = True, check_packages: 'bool' = False) -> \"'TabularPredictor'\"\n",
      " |      Load a TabularPredictor object previously produced by `fit()` from file and returns this object. It is highly recommended the predictor be loaded with the exact AutoGluon version it was fit with.\n",
      " |      \n",
      " |      .. warning::\n",
      " |      \n",
      " |          :meth:`autogluon.tabular.TabularPredictor.load` uses `pickle` module implicitly, which is known to\n",
      " |          be insecure. It is possible to construct malicious pickle data which will execute arbitrary code during\n",
      " |          unpickling. Never load data that could have come from an untrusted source, or that could have been tampered\n",
      " |          with. **Only load data you trust.**\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      path : str\n",
      " |          The path to directory in which this Predictor was previously saved.\n",
      " |      verbosity : int, default = None\n",
      " |          Sets the verbosity level of this Predictor after it is loaded.\n",
      " |          Valid values range from 0 (least verbose) to 4 (most verbose).\n",
      " |          If None, logging verbosity is not changed from existing values.\n",
      " |          Specify larger values to see more information printed when using Predictor during inference, smaller values to see less information.\n",
      " |          Refer to TabularPredictor init for more information.\n",
      " |      require_version_match : bool, default = True\n",
      " |          If True, will raise an AssertionError if the `autogluon.tabular` version of the loaded predictor does not match the installed version of `autogluon.tabular`.\n",
      " |          If False, will allow loading of models trained on incompatible versions, but is NOT recommended. Users may run into numerous issues if attempting this.\n",
      " |      require_py_version_match : bool, default = True\n",
      " |          If True, will raise an AssertionError if the Python version of the loaded predictor does not match the installed Python version.\n",
      " |              Micro version differences such as 3.9.2 and 3.9.7 will log a warning but will not raise an exception.\n",
      " |          If False, will allow loading of models trained on incompatible python versions, but is NOT recommended. Users may run into numerous issues if attempting this.\n",
      " |      check_packages : bool, default = False\n",
      " |          If True, checks package versions of the loaded predictor against the package versions of the current environment.\n",
      " |          Warnings will be logged for each mismatch of package version.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      predictor : TabularPredictor\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> predictor = TabularPredictor.load(path_to_predictor)\n",
      " |  \n",
      " |  load_log(predictor_path: 'str' = None, log_file_path: 'Optional[str]' = None) -> 'List[str]'\n",
      " |      Load log files of a predictor\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      predictor_path: Optional[str], default = None\n",
      " |          Path to the predictor to load the log.\n",
      " |          This can be used when the predictor was initialized with `log_file_path=\"auto\"` to fetch the log file automatically\n",
      " |      log_file_path: Optional[str], default = None\n",
      " |          Path to the log file.\n",
      " |          If you specified a `log_file_path` while initializing the predictor, you should use `log_file_path` to load the log file instead.\n",
      " |          At least one of `predictor_path` or `log_file_path` must to be specified\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      List[str]\n",
      " |          A list containing lines of the log file\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Readonly properties defined here:\n",
      " |  \n",
      " |  can_predict_proba\n",
      " |      Return True if predictor can return prediction probabilities via `.predict_proba`, otherwise return False.\n",
      " |      Raises an AssertionError if called before fitting.\n",
      " |  \n",
      " |  class_labels\n",
      " |      Alias to self.classes_\n",
      " |  \n",
      " |  class_labels_internal\n",
      " |      For multiclass problems, this list contains the internal class labels in sorted order of internal `predict_proba()` output.\n",
      " |      For binary problems, this list contains the internal class labels in sorted order of internal `predict_proba(as_multiclass=True)` output.\n",
      " |          The value will always be `class_labels_internal=[0, 1]` for binary problems, with 0 as the negative class, and 1 as the positive class.\n",
      " |      For other problem types, will equal None.\n",
      " |  \n",
      " |  class_labels_internal_map\n",
      " |      For binary and multiclass classification problems, this dictionary contains the mapping of the original labels to the internal labels.\n",
      " |      For example, in binary classification, label values of 'True' and 'False' will be mapped to the internal representation `1` and `0`.\n",
      " |          Therefore, class_labels_internal_map would equal {'True': 1, 'False': 0}\n",
      " |      For other problem types, will equal None.\n",
      " |      For multiclass, it is possible for not all of the label values to have a mapping.\n",
      " |          This indicates that the internal models will never predict those missing labels, and training rows associated with the missing labels were dropped.\n",
      " |  \n",
      " |  classes_\n",
      " |      For multiclass problems, this list contains the class labels in sorted order of `predict_proba()` output.\n",
      " |      For binary problems, this list contains the class labels in sorted order of `predict_proba(as_multiclass=True)` output.\n",
      " |          `classes_[0]` corresponds to internal label = 0 (negative class), `classes_[1]` corresponds to internal label = 1 (positive class).\n",
      " |          This is relevant for certain metrics such as F1 where True and False labels impact the metric score differently.\n",
      " |      For other problem types, will equal None.\n",
      " |      For example if `pred = predict_proba(x, as_multiclass=True)`, then ith index of `pred` provides predicted probability that `x` belongs to class given by `classes_[i]`.\n",
      " |  \n",
      " |  decision_threshold\n",
      " |      The decision threshold used to convert prediction probabilities to predictions.\n",
      " |      Only relevant for binary classification, otherwise the value will be None.\n",
      " |      Valid values are in the range [0.0, 1.0]\n",
      " |      You can obtain an optimized `decision_threshold` by first calling `predictor.calibrate_decision_threshold()`.\n",
      " |      Useful to set for metrics such as `balanced_accuracy` and `f1` as `0.5` is often not an optimal threshold.\n",
      " |      Predictions are calculated via the following logic on the positive class: `1 if pred > decision_threshold else 0`\n",
      " |  \n",
      " |  eval_metric\n",
      " |      The metric used to evaluate predictive performance\n",
      " |  \n",
      " |  feature_metadata\n",
      " |      Returns the internal FeatureMetadata.\n",
      " |      \n",
      " |      Inferred data type of each predictive variable after preprocessing transformation (i.e. column of training data table used to predict `label`).\n",
      " |      Contains both raw dtype and special dtype information. Each feature has exactly 1 raw dtype (such as 'int', 'float', 'category') and zero to many special dtypes (such as 'datetime_as_int', 'text', 'text_ngram').\n",
      " |      Special dtypes are AutoGluon specific feature types that are used to identify features with meaning beyond what the raw dtype can convey.\n",
      " |          `feature_metadata.type_map_raw`: Dictionary of feature name -> raw dtype mappings.\n",
      " |          `feature_metadata.type_group_map_special`: Dictionary of lists of special feature names, grouped by special feature dtype.\n",
      " |  \n",
      " |  feature_metadata_in\n",
      " |      Returns the input FeatureMetadata.\n",
      " |      \n",
      " |      Inferred data type of each predictive variable before preprocessing transformation.\n",
      " |      Contains both raw dtype and special dtype information. Each feature has exactly 1 raw dtype (such as 'int', 'float', 'category') and zero to many special dtypes (such as 'datetime_as_int', 'text', 'text_ngram').\n",
      " |      Special dtypes are AutoGluon specific feature types that are used to identify features with meaning beyond what the raw dtype can convey.\n",
      " |          `feature_metadata.type_map_raw`: Dictionary of feature name -> raw dtype mappings.\n",
      " |          `feature_metadata.type_group_map_special`: Dictionary of lists of special feature names, grouped by special feature dtype.\n",
      " |  \n",
      " |  has_val\n",
      " |      Return True if holdout validation data was used during fit, else return False.\n",
      " |  \n",
      " |  is_fit\n",
      " |      Return True if `predictor.fit` has been called, otherwise return False.\n",
      " |  \n",
      " |  label\n",
      " |      Name of table column that contains data from the variable to predict (often referred to as: labels, response variable, target variable, dependent variable, y, etc).\n",
      " |  \n",
      " |  model_best\n",
      " |      Returns the string model name of the best model by validation score that can infer.\n",
      " |      This is the same model used during inference when `predictor.predict` is called without specifying a model.\n",
      " |      This can be updated to be a model other than the model with best validation score by methods such as refit_full and set_model_best.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      String model name of the best model\n",
      " |  \n",
      " |  original_features\n",
      " |      Original features user passed in to fit before processing\n",
      " |  \n",
      " |  path\n",
      " |      Path to directory where all models used by this Predictor are stored\n",
      " |  \n",
      " |  positive_class\n",
      " |      Returns the positive class name in binary classification. Useful for computing metrics such as F1 which require a positive and negative class.\n",
      " |      In binary classification, :class:`TabularPredictor.predict_proba(as_multiclass=False)` returns the estimated probability that each row belongs to the positive class.\n",
      " |      Will print a warning and return None if called when `predictor.problem_type != 'binary'`.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      The positive class name in binary classification or None if the problem is not binary classification.\n",
      " |  \n",
      " |  problem_type\n",
      " |      What type of prediction problem this Predictor has been trained for\n",
      " |  \n",
      " |  quantile_levels\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes defined here:\n",
      " |  \n",
      " |  Dataset = <class 'autogluon.common.dataset.TabularDataset'>\n",
      " |      A dataset in tabular format (with rows = samples, columns = features/variables).\n",
      " |      This class returns a :class:`pd.DataFrame` when initialized and all existing pandas methods can be applied to it.\n",
      " |      For full list of methods/attributes, see pandas Dataframe documentation: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html\n",
      " |      \n",
      " |      The purpose of this class is to provide an easy-to-use shorthand for loading a pandas DataFrame to use in AutoGluon.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      data : str, :class:`pd.DataFrame`, :class:`np.ndarray`, Iterable, or dict\n",
      " |          If str, path to data file (CSV or Parquet format).\n",
      " |          If you already have your data in a :class:`pd.DataFrame`, you can specify it here. In this case, the same DataFrame will be returned with no changes.\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      >>> import pandas as pd\n",
      " |      >>> from autogluon.common import TabularDataset\n",
      " |      >>> train_data = TabularDataset(\"https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv\")\n",
      " |      >>> train_data_pd = pd.read_csv(\"https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv\")\n",
      " |      >>> assert isinstance(train_data, pd.DataFrame)  # True\n",
      " |      >>> assert train_data.equals(train_data_pd)  # True\n",
      " |      >>> assert type(train_data) == type(train_data_pd)  # True\n",
      " |  \n",
      " |  \n",
      " |  predictor_file_name = 'predictor.pkl'\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from autogluon.tabular.predictor._deprecated_methods.TabularPredictorDeprecatedMixin:\n",
      " |  \n",
      " |  get_model_best(self) -> 'str'\n",
      " |      Deprecated method. Use `model_best` instead.\n",
      " |  \n",
      " |  get_model_full_dict(self, *args, **kwargs) -> 'Dict[str, str]'\n",
      " |      Deprecated method. Use `model_refit_map` instead.\n",
      " |  \n",
      " |  get_model_names(self, *args, **kwargs) -> 'List[str]'\n",
      " |      Deprecated method. Use `model_names` instead.\n",
      " |  \n",
      " |  get_model_names_persisted(self) -> 'List[str]'\n",
      " |      Deprecated method. Use `model_names(persisted=True)` instead.\n",
      " |  \n",
      " |  get_oof_pred(self, *args, **kwargs) -> 'pd.Series'\n",
      " |      Deprecated method. Use `predict_oof` instead.\n",
      " |  \n",
      " |  get_oof_pred_proba(self, *args, **kwargs) -> 'pd.DataFrame | pd.Series'\n",
      " |      Deprecated method. Use `predict_proba_oof` instead.\n",
      " |  \n",
      " |  get_pred_from_proba(self, *args, **kwargs) -> 'pd.Series | np.array'\n",
      " |      Deprecated method. Use `predict_from_proba` instead.\n",
      " |  \n",
      " |  get_size_disk(self) -> 'int'\n",
      " |      Deprecated method. Use `disk_usage` instead.\n",
      " |  \n",
      " |  get_size_disk_per_file(self, *args, **kwargs) -> 'pd.Series'\n",
      " |      Deprecated method. Use `disk_usage_per_file` instead.\n",
      " |  \n",
      " |  persist_models(self, *args, **kwargs) -> 'List[str]'\n",
      " |      Deprecated method. Use `persist` instead.\n",
      " |  \n",
      " |  unpersist_models(self, *args, **kwargs) -> 'List[str]'\n",
      " |      Deprecated method. Use `unpersist` instead.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from autogluon.tabular.predictor._deprecated_methods.TabularPredictorDeprecatedMixin:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(help(TabularPredictor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b070d740-72ec-465d-b349-a08bc1e27244",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function fit in module autogluon.tabular.predictor.predictor:\n",
      "\n",
      "fit(self, train_data: 'pd.DataFrame | str', tuning_data: 'pd.DataFrame | str' = None, time_limit: 'float' = None, presets: 'List[str] | str' = None, hyperparameters: 'dict | str' = None, feature_metadata='infer', infer_limit: 'float' = None, infer_limit_batch_size: 'int' = None, fit_weighted_ensemble: 'bool' = True, fit_full_last_level_weighted_ensemble: 'bool' = True, full_weighted_ensemble_additionally: 'bool' = False, dynamic_stacking: 'bool | str' = False, calibrate_decision_threshold: 'bool | str' = 'auto', num_cpus: 'int | str' = 'auto', num_gpus: 'int | str' = 'auto', fit_strategy: \"Literal['sequential', 'parallel']\" = 'sequential', memory_limit: 'float | str' = 'auto', callbacks: 'List[AbstractCallback]' = None, **kwargs) -> \"'TabularPredictor'\"\n",
      "    Fit models to predict a column of a data table (label) based on the other columns (features).\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    train_data : :class:`pd.DataFrame` or str\n",
      "        Table of the training data as a pandas DataFrame.\n",
      "        If str is passed, `train_data` will be loaded using the str value as the file path.\n",
      "    tuning_data : :class:`pd.DataFrame` or str, optional\n",
      "        Another dataset containing validation data reserved for tuning processes such as early stopping and hyperparameter tuning.\n",
      "        This dataset should be in the same format as `train_data`.\n",
      "        If str is passed, `tuning_data` will be loaded using the str value as the file path.\n",
      "        Note: final model returned may be fit on `tuning_data` as well as `train_data`. Do not provide your evaluation test data here!\n",
      "        In particular, when `num_bag_folds` > 0 or `num_stack_levels` > 0, models will be trained on both `tuning_data` and `train_data`.\n",
      "        If `tuning_data = None`, `fit()` will automatically hold out some random validation examples from `train_data`.\n",
      "    time_limit : int, default = None\n",
      "        Approximately how long `fit()` should run for (wallclock time in seconds).\n",
      "        If not specified, `fit()` will run until all models have completed training, but will not repeatedly bag models unless `num_bag_sets` is specified.\n",
      "    presets : list or str or dict, default = ['medium_quality']\n",
      "        List of preset configurations for various arguments in `fit()`. Can significantly impact predictive accuracy, memory-footprint, and inference latency of trained models, and various other properties of the returned `predictor`.\n",
      "        It is recommended to specify presets and avoid specifying most other `fit()` arguments or model hyperparameters prior to becoming familiar with AutoGluon.\n",
      "        As an example, to get the most accurate overall predictor (regardless of its efficiency), set `presets='best_quality'`.\n",
      "        To get good quality with minimal disk usage, set `presets=['good_quality', 'optimize_for_deployment']`\n",
      "        Any user-specified arguments in `fit()` will override the values used by presets.\n",
      "        If specifying a list of presets, later presets will override earlier presets if they alter the same argument.\n",
      "        For precise definitions of the provided presets, see file: `autogluon/tabular/configs/presets_configs.py`.\n",
      "        Users can specify custom presets by passing in a dictionary of argument values as an element to the list.\n",
      "    \n",
      "        Available Presets: ['best_quality', 'high_quality', 'good_quality', 'medium_quality', 'experimental_quality', 'optimize_for_deployment', 'interpretable', 'ignore_text']\n",
      "    \n",
      "        It is recommended to only use one `quality` based preset in a given call to `fit()` as they alter many of the same arguments and are not compatible with each-other.\n",
      "    \n",
      "        In-depth Preset Info:\n",
      "            best_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'zeroshot'}\n",
      "                Best predictive accuracy with little consideration to inference time or disk usage. Achieve even better results by specifying a large time_limit value.\n",
      "                Recommended for applications that benefit from the best possible model accuracy.\n",
      "    \n",
      "            high_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'zeroshot', 'refit_full': True, 'set_best_to_refit_full': True, 'save_bag_folds': False}\n",
      "                High predictive accuracy with fast inference. ~8x faster inference and ~8x lower disk usage than `best_quality`.\n",
      "                Recommended for applications that require reasonable inference speed and/or model size.\n",
      "    \n",
      "            good_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'light', 'refit_full': True, 'set_best_to_refit_full': True, 'save_bag_folds': False}\n",
      "                Good predictive accuracy with very fast inference. ~4x faster inference and ~4x lower disk usage than `high_quality`.\n",
      "                Recommended for applications that require fast inference speed.\n",
      "    \n",
      "            medium_quality={'auto_stack': False}\n",
      "                Medium predictive accuracy with very fast inference and very fast training time. ~20x faster training than `good_quality`.\n",
      "                This is the default preset in AutoGluon, but should generally only be used for quick prototyping, as `good_quality` results in significantly better predictive accuracy and faster inference time.\n",
      "    \n",
      "            experimental_quality={'auto_stack': True, 'dynamic_stacking': 'auto', 'hyperparameters': 'experimental', 'fit_strategy': 'parallel', 'num_gpus': 0}\n",
      "                This preset acts as a testing ground for cutting edge features and models which could later be added to the `best_quality` preset in future releases.\n",
      "                Recommended when `best_quality` was already being used and the user wants to push performance even further.\n",
      "    \n",
      "            optimize_for_deployment={'keep_only_best': True, 'save_space': True}\n",
      "                Optimizes result immediately for deployment by deleting unused models and removing training artifacts.\n",
      "                Often can reduce disk usage by ~2-4x with no negatives to model accuracy or inference speed.\n",
      "                This will disable numerous advanced functionality, but has no impact on inference.\n",
      "                This will make certain functionality less informative, such as `predictor.leaderboard()` and `predictor.fit_summary()`.\n",
      "                    Because unused models will be deleted under this preset, methods like `predictor.leaderboard()` and `predictor.fit_summary()` will no longer show the full set of models that were trained during `fit()`.\n",
      "                Recommended for applications where the inner details of AutoGluon's training is not important and there is no intention of manually choosing between the final models.\n",
      "                This preset pairs well with the other presets such as `good_quality` to make a very compact final model.\n",
      "                Identical to calling `predictor.delete_models(models_to_keep='best', dry_run=False)` and `predictor.save_space()` directly after `fit()`.\n",
      "    \n",
      "            interpretable={'auto_stack': False, 'hyperparameters': 'interpretable'}\n",
      "                Fits only interpretable rule-based models from the imodels package.\n",
      "                Trades off predictive accuracy for conciseness.\n",
      "    \n",
      "            ignore_text={'_feature_generator_kwargs': {'enable_text_ngram_features': False, 'enable_text_special_features': False, 'enable_raw_text_features': False}}\n",
      "                Disables automated feature generation when text features are detected.\n",
      "                This is useful to determine how beneficial text features are to the end result, as well as to ensure features are not mistaken for text when they are not.\n",
      "                Ignored if `feature_generator` was also specified.\n",
      "    \n",
      "    hyperparameters : str or dict, default = 'default'\n",
      "        Determines the hyperparameters used by the models.\n",
      "        If `str` is passed, will use a preset hyperparameter configuration.\n",
      "            Valid `str` options: ['default', 'zeroshot', 'light', 'very_light', 'toy', 'multimodal']\n",
      "                'default': Default AutoGluon hyperparameters intended to get strong accuracy with reasonable disk usage and inference time. Used in the 'medium_quality' preset.\n",
      "                'zeroshot': A powerful model portfolio learned from TabRepo's ensemble simulation on 200 datasets. Contains ~100 models and is used in 'best_quality' and 'high_quality' presets.\n",
      "                'light': Results in smaller models. Generally will make inference speed much faster and disk usage much lower, but with worse accuracy. Used in the 'good_quality' preset.\n",
      "                'very_light': Results in much smaller models. Behaves similarly to 'light', but in many cases with over 10x less disk usage and a further reduction in accuracy.\n",
      "                'toy': Results in extremely small models. Only use this when prototyping, as the model quality will be severely reduced.\n",
      "                'multimodal': [EXPERIMENTAL] Trains a multimodal transformer model alongside tabular models. Requires that some text columns appear in the data and GPU.\n",
      "                    When combined with 'best_quality' `presets` option, this can achieve extremely strong results in multimodal data tables that contain columns with text in addition to numeric/categorical columns.\n",
      "            Reference `autogluon/tabular/configs/hyperparameter_configs.py` for information on the hyperparameters associated with each preset.\n",
      "        Keys are strings that indicate which model types to train.\n",
      "            Stable model options include:\n",
      "                'GBM' (LightGBM)\n",
      "                'CAT' (CatBoost)\n",
      "                'XGB' (XGBoost)\n",
      "                'RF' (random forest)\n",
      "                'XT' (extremely randomized trees)\n",
      "                'KNN' (k-nearest neighbors)\n",
      "                'LR' (linear regression)\n",
      "                'NN_TORCH' (neural network implemented in Pytorch)\n",
      "                'FASTAI' (neural network with FastAI backend)\n",
      "                'AG_AUTOMM' (`MultimodalPredictor` from `autogluon.multimodal`. Supports Tabular, Text, and Image modalities. GPU is required.)\n",
      "            Experimental model options include:\n",
      "                'FT_TRANSFORMER' (Tabular Transformer, GPU is recommended. Does not scale well to >100 features.)\n",
      "                'FASTTEXT' (FastText. Note: Has not been tested for a long time.)\n",
      "                'TABPFN' (TabPFN. Does not scale well to >100 features or >1000 rows, and does not support regression. Extremely slow inference speed.)\n",
      "                'VW' (VowpalWabbit. Note: Has not been tested for a long time.)\n",
      "                'AG_TEXT_NN' (Multimodal Text+Tabular model, GPU is required. Recommended to instead use its successor, 'AG_AUTOMM'.)\n",
      "                'AG_IMAGE_NN' (Image model, GPU is required. Recommended to instead use its successor, 'AG_AUTOMM'.)\n",
      "            If a certain key is missing from hyperparameters, then `fit()` will not train any models of that type. Omitting a model key from hyperparameters is equivalent to including this model key in `excluded_model_types`.\n",
      "            For example, set `hyperparameters = { 'NN_TORCH':{...} }` if say you only want to train (PyTorch) neural networks and no other types of models.\n",
      "        Values = dict of hyperparameter settings for each model type, or list of dicts.\n",
      "            Each hyperparameter can either be a single fixed value or a search space containing many possible values.\n",
      "            Unspecified hyperparameters will be set to default values (or default search spaces if `hyperparameter_tune_kwargs='auto'`).\n",
      "            Caution: Any provided search spaces will error if `hyperparameter_tune_kwargs=None` (Default).\n",
      "            To train multiple models of a given type, set the value to a list of hyperparameter dictionaries.\n",
      "                For example, `hyperparameters = {'RF': [{'criterion': 'gini'}, {'criterion': 'entropy'}]}` will result in 2 random forest models being trained with separate hyperparameters.\n",
      "        Advanced functionality: Bring your own model / Custom model support\n",
      "            AutoGluon fully supports custom models. For a detailed tutorial on creating and using custom models with AutoGluon, refer to https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-model.html\n",
      "        Advanced functionality: Custom stack levels\n",
      "            By default, AutoGluon re-uses the same models and model hyperparameters at each level during stack ensembling.\n",
      "            To customize this behaviour, create a hyperparameters dictionary separately for each stack level, and then add them as values to a new dictionary, with keys equal to the stack level.\n",
      "                Example: `hyperparameters = {1: {'RF': rf_params1}, 2: {'CAT': [cat_params1, cat_params2], 'NN_TORCH': {}}}`\n",
      "                This will result in a stack ensemble that has one custom random forest in level 1 followed by two CatBoost models with custom hyperparameters and a default neural network in level 2, for a total of 4 models.\n",
      "            If a level is not specified in `hyperparameters`, it will default to using the highest specified level to train models. This can also be explicitly controlled by adding a 'default' key.\n",
      "    \n",
      "        Default:\n",
      "            hyperparameters = {\n",
      "                'NN_TORCH': {},\n",
      "                'GBM': [\n",
      "                    {'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}},\n",
      "                    {},\n",
      "                    {\n",
      "                        \"learning_rate\": 0.03,\n",
      "                        \"num_leaves\": 128,\n",
      "                        \"feature_fraction\": 0.9,\n",
      "                        \"min_data_in_leaf\": 3,\n",
      "                        \"ag_args\": {\"name_suffix\": \"Large\", \"priority\": 0, \"hyperparameter_tune_kwargs\": None},\n",
      "                    },\n",
      "                ],\n",
      "                'CAT': {},\n",
      "                'XGB': {},\n",
      "                'FASTAI': {},\n",
      "                'RF': [\n",
      "                    {'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}},\n",
      "                    {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}},\n",
      "                    {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression']}},\n",
      "                ],\n",
      "                'XT': [\n",
      "                    {'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}},\n",
      "                    {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}},\n",
      "                    {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression']}},\n",
      "                ],\n",
      "                'KNN': [\n",
      "                    {'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}},\n",
      "                    {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}},\n",
      "                ],\n",
      "            }\n",
      "    \n",
      "        Details regarding the hyperparameters you can specify for each model are provided in the following files:\n",
      "            NN: `autogluon.tabular.models.tabular_nn.hyperparameters.parameters`\n",
      "                Note: certain hyperparameter settings may cause these neural networks to train much slower.\n",
      "            GBM: `autogluon.tabular.models.lgb.hyperparameters.parameters`\n",
      "                 See also the lightGBM docs: https://lightgbm.readthedocs.io/en/latest/Parameters.html\n",
      "            CAT: `autogluon.tabular.models.catboost.hyperparameters.parameters`\n",
      "                 See also the CatBoost docs: https://catboost.ai/docs/concepts/parameter-tuning.html\n",
      "            XGB: `autogluon.tabular.models.xgboost.hyperparameters.parameters`\n",
      "                 See also the XGBoost docs: https://xgboost.readthedocs.io/en/latest/parameter.html\n",
      "            FASTAI: `autogluon.tabular.models.fastainn.hyperparameters.parameters`\n",
      "                 See also the FastAI docs: https://docs.fast.ai/tabular.learner.html\n",
      "            RF: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html\n",
      "                Note: Hyperparameter tuning is disabled for this model.\n",
      "            XT: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html\n",
      "                Note: Hyperparameter tuning is disabled for this model.\n",
      "            KNN: See sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html\n",
      "                Note: Hyperparameter tuning is disabled for this model.\n",
      "            LR: `autogluon.tabular.models.lr.hyperparameters.parameters`\n",
      "                Note: Hyperparameter tuning is disabled for this model.\n",
      "                Note: 'penalty' parameter can be used for regression to specify regularization method: 'L1' and 'L2' values are supported.\n",
      "            Advanced functionality: Custom AutoGluon model arguments\n",
      "                These arguments are optional and can be specified in any model's hyperparameters.\n",
      "                    Example: `hyperparameters = {'RF': {..., 'ag_args': {'name_suffix': 'CustomModelSuffix', 'disable_in_hpo': True}}`\n",
      "                ag_args: Dictionary of customization options related to meta properties of the model such as its name, the order it is trained, the problem types it is valid for, and the type of HPO it utilizes.\n",
      "                    Valid keys:\n",
      "                        name: (str) The name of the model. This overrides AutoGluon's naming logic and all other name arguments if present.\n",
      "                        name_main: (str) The main name of the model. Example: 'RandomForest'.\n",
      "                        name_prefix: (str) Add a custom prefix to the model name. Unused by default.\n",
      "                        name_suffix: (str) Add a custom suffix to the model name. Unused by default.\n",
      "                        priority: (int) Determines the order in which the model is trained. Larger values result in the model being trained earlier. Default values range from 100 (KNN) to 0 (custom), dictated by model type. If you want this model to be trained first, set priority = 999.\n",
      "                        problem_types: (list) List of valid problem types for the model. `problem_types=['binary']` will result in the model only being trained if `problem_type` is 'binary'.\n",
      "                        disable_in_hpo: (bool) If True, the model will only be trained if `hyperparameter_tune_kwargs=None`.\n",
      "                        valid_stacker: (bool) If False, the model will not be trained as a level 2 or higher stacker model.\n",
      "                        valid_base: (bool) If False, the model will not be trained as a level 1 (base) model.\n",
      "                        hyperparameter_tune_kwargs: (dict) Refer to :meth:`TabularPredictor.fit` hyperparameter_tune_kwargs argument. If specified here, will override global HPO settings for this model.\n",
      "                    Reference the default hyperparameters for example usage of these options.\n",
      "                ag_args_fit: Dictionary of model fit customization options related to how and with what constraints the model is trained. These parameters affect stacker fold models, but not stacker models themselves.\n",
      "                    Clarification: `time_limit` is the internal time in seconds given to a particular model to train, which is dictated in part by the `time_limit` argument given during `predictor.fit()` but is not the same.\n",
      "                    Valid keys:\n",
      "                        stopping_metric: (str or :class:`autogluon.core.metrics.Scorer`, default=None) The metric to use for early stopping of the model. If None, model will decide.\n",
      "                        max_memory_usage_ratio: (float, default=1.0) The ratio of memory usage relative to the default to allow before early stopping or killing the model. Values greater than 1.0 will be increasingly prone to out-of-memory errors.\n",
      "                        max_time_limit_ratio: (float, default=1.0) The ratio of the provided time_limit to use during model `fit()`. If `time_limit=10` and `max_time_limit_ratio=0.3`, time_limit would be changed to 3. Does not alter max_time_limit or min_time_limit values.\n",
      "                        max_time_limit: (float, default=None) Maximum amount of time to allow this model to train for (in sec). If the provided time_limit is greater than this value, it will be replaced by max_time_limit.\n",
      "                        min_time_limit: (float, default=0) Allow this model to train for at least this long (in sec), regardless of the time limit it would otherwise be granted.\n",
      "                            If `min_time_limit >= max_time_limit`, time_limit will be set to min_time_limit.\n",
      "                            If `min_time_limit=None`, time_limit will be set to None and the model will have no training time restriction.\n",
      "                        num_cpus : (int or str, default='auto')\n",
      "                            How many CPUs to use during model fit.\n",
      "                            If 'auto', model will decide.\n",
      "                        num_gpus : (int or str, default='auto')\n",
      "                            How many GPUs to use during model fit.\n",
      "                            If 'auto', model will decide. Some models can use GPUs but don't by default due to differences in model quality.\n",
      "                            Set to 0 to disable usage of GPUs.\n",
      "                ag_args_ensemble: Dictionary of hyperparameters shared by all models that control how they are ensembled, if bag mode is enabled.\n",
      "                    Valid keys:\n",
      "                        use_orig_features: [True, False, \"never\"], default True\n",
      "                            Whether a stack model will use the original features along with the stack features to train (akin to skip-connections).\n",
      "                            If True, will use the original data features.\n",
      "                            If False, will discard the original data features and only use stack features, except when no stack features exist (such as in layer 1).\n",
      "                            If \"never\", will always discard the original data features. Will be skipped in layer 1.\n",
      "                        valid_stacker : bool, default True\n",
      "                            If True, will be marked as valid to include as a stacker model.\n",
      "                            If False, will only be fit as a base model (layer 1) and will not be fit in stack layers (layer 2+).\n",
      "                        max_base_models : int, default 0\n",
      "                            Maximum number of base models whose predictions form the features input to this stacker model.\n",
      "                            If more than `max_base_models` base models are available, only the top `max_base_models` models with highest validation score are used.\n",
      "                            If 0, the logic is skipped.\n",
      "                        max_base_models_per_type : int | str, default \"auto\"\n",
      "                            Similar to `max_base_models`. If more than `max_base_models_per_type` of any particular model type are available,\n",
      "                            only the top `max_base_models_per_type` of that type are used. This occurs before the `max_base_models` filter.\n",
      "                            If \"auto\", the value will be adaptively set based on the number of training samples.\n",
      "                                More samples will lead to larger values, starting at 1 with <1000 samples, increasing up to 12 at >=50000 samples.\n",
      "                            If 0, the logic is skipped.\n",
      "                        num_folds: (int, default=None) If specified, the number of folds to fit in the bagged model.\n",
      "                            If specified, overrides any other value used to determine the number of folds such as predictor.fit `num_bag_folds` argument.\n",
      "                        max_sets: (int, default=None) If specified, the maximum sets to fit in the bagged model.\n",
      "                            The lesser of `max_sets` and the predictor.fit `num_bag_sets` argument will be used for the given model.\n",
      "                            Useful if a particular model is expensive relative to others and you want to avoid repeated bagging of the expensive model while still repeated bagging the cheaper models.\n",
      "                        save_bag_folds: (bool, default=True)\n",
      "                            If True, bagged models will save their fold models (the models from each individual fold of bagging). This is required to use bagged models for prediction.\n",
      "                            If False, bagged models will not save their fold models. This means that bagged models will not be valid models during inference.\n",
      "                                This should only be set to False when planning to call `predictor.refit_full()` or when `refit_full` is set and `set_best_to_refit_full=True`.\n",
      "                                Particularly useful if disk usage is a concern. By not saving the fold models, bagged models will use only very small amounts of disk space during training.\n",
      "                                In many training runs, this will reduce peak disk usage by >10x.\n",
      "                        fold_fitting_strategy: (AbstractFoldFittingStrategy default=auto) Whether to fit folds in parallel or in sequential order.\n",
      "                            If parallel_local, folds will be trained in parallel with evenly distributed computing resources. This could bring 2-4x speedup compared to SequentialLocalFoldFittingStrategy, but could consume much more memory.\n",
      "                            If sequential_local, folds will be trained in sequential.\n",
      "                            If auto, strategy will be determined by OS and whether ray is installed or not. MacOS support for parallel_local is unstable, and may crash if enabled.\n",
      "                        num_folds_parallel: (int or str, default='auto') Number of folds to be trained in parallel if using ParallelLocalFoldFittingStrategy. Consider lowering this value if you encounter either out of memory issue or CUDA out of memory issue(when trained on gpu).\n",
      "                            if 'auto', will try to train all folds in parallel.\n",
      "    \n",
      "    feature_metadata : :class:`autogluon.tabular.FeatureMetadata` or str, default = 'infer'\n",
      "        The feature metadata used in various inner logic in feature preprocessing.\n",
      "        If 'infer', will automatically construct a FeatureMetadata object based on the properties of `train_data`.\n",
      "        In this case, `train_data` is input into :meth:`autogluon.tabular.FeatureMetadata.from_df` to infer `feature_metadata`.\n",
      "        If 'infer' incorrectly assumes the dtypes of features, consider explicitly specifying `feature_metadata`.\n",
      "    infer_limit : float, default = None\n",
      "        The inference time limit in seconds per row to adhere to during fit.\n",
      "        If infer_limit=0.05 and infer_limit_batch_size=1000, AutoGluon will avoid training models that take longer than 50 ms/row to predict when given a batch of 1000 rows to predict (must predict 1000 rows in no more than 50 seconds).\n",
      "        If bagging is enabled, the inference time limit will be respected based on estimated inference speed of `_FULL` models after refit_full is called, NOT on the inference speed of the bagged ensembles.\n",
      "        The inference times calculated for models are assuming `predictor.persist('all')` is called after fit.\n",
      "        If None, no limit is enforced.\n",
      "        If it is impossible to satisfy the constraint, an exception will be raised.\n",
      "    infer_limit_batch_size : int, default = None\n",
      "        The batch size to use when predicting in bulk to estimate per-row inference time.\n",
      "        Must be an integer greater than 0.\n",
      "        If None and `infer_limit` is specified, will default to 10000.\n",
      "        It is recommended to set to 10000 unless you must satisfy an online-inference scenario.\n",
      "        Small values, especially `infer_limit_batch_size=1`, will result in much larger per-row inference times and should be avoided if possible.\n",
      "        Refer to `infer_limit` for more details on how this is used.\n",
      "        If specified when `infer_limit=None`, the inference time will be logged during training but will not be limited.\n",
      "    fit_weighted_ensemble : bool, default = True\n",
      "        If True, a WeightedEnsembleModel will be fit in each stack layer.\n",
      "        A weighted ensemble will often be stronger than an individual model while being very fast to train.\n",
      "        It is recommended to keep this value set to True to maximize predictive quality.\n",
      "    fit_full_last_level_weighted_ensemble : bool, default = True\n",
      "        If True, the WeightedEnsembleModel of the last stacking level will be fit with all (successful) models from all previous layers as base models.\n",
      "        If stacking is disabled, settings this to True or False makes no difference because the WeightedEnsembleModel L2 always uses all models from L1.\n",
      "        It is recommended to keep this value set to True to maximize predictive quality.\n",
      "    full_weighted_ensemble_additionally : bool, default = False\n",
      "        If True, AutoGluon will fit two WeightedEnsembleModels after training all stacking levels. Setting this to True, simulates calling\n",
      "        `fit_weighted_ensemble()` after calling `fit()`. Has no affect if `fit_full_last_level_weighted_ensemble` is False and does not fit an additional\n",
      "        WeightedEnsembleModel if stacking is disabled.\n",
      "    dynamic_stacking: bool | str, default = False\n",
      "        If True and `num_stack_levels` > 0, AutoGluon will dynamically determine whether to use stacking or not by first validating AutoGluon's stacking\n",
      "        behavior. This is done to avoid so-called stacked overfitting that can make traditional multi-layer stacking, as used in AutoGluon, fail drastically\n",
      "        and produce unreliable validation scores.\n",
      "        It is recommended to keep this value set to True or \"auto\" when using stacking,\n",
      "        as long as it is unknown whether the data is affected by stacked overfitting.\n",
      "        If it is known that the data is unaffected by stacked overfitting, then setting this value to False is expected to maximize predictive quality.\n",
      "        If enabled, by default, AutoGluon performs dynamic stacking by spending 25% of the provided time limit for detection and all remaining\n",
      "        time for fitting AutoGluon. This can be adjusted by specifying `ds_args` with different parameters to `fit()`.\n",
      "        If \"auto\", will be set to `not use_bag_holdout`.\n",
      "        See the documentation of `ds_args` for more information.\n",
      "    calibrate_decision_threshold : bool | str, default = \"auto\"\n",
      "        If True, will automatically calibrate the decision threshold at the end of fit for calls to `.predict` based on the evaluation metric.\n",
      "        If \"auto\", will be set to True if `eval_metric.needs_class=True` and `problem_type=\"binary\"`.\n",
      "        By default, the decision threshold is `0.5`, however for some metrics such as `f1` and `balanced_accuracy`,\n",
      "        scores can be significantly improved by choosing a threshold other than `0.5`.\n",
      "        Only valid for `problem_type='binary'`. Ignored for all other problem types.\n",
      "    num_cpus: int | str, default = \"auto\"\n",
      "        The total amount of cpus you want AutoGluon predictor to use.\n",
      "        Auto means AutoGluon will make the decision based on the total number of cpus available and the model requirement for best performance.\n",
      "        Users generally don't need to set this value\n",
      "    num_gpus: int | str, default = \"auto\"\n",
      "        The total amount of gpus you want AutoGluon predictor to use.\n",
      "        Auto means AutoGluon will make the decision based on the total number of gpus available and the model requirement for best performance.\n",
      "        Users generally don't need to set this value\n",
      "    fit_strategy: Literal[\"sequential\", \"parallel\"], default = \"sequential\"\n",
      "        The strategy used to fit models.\n",
      "        If \"sequential\", models will be fit sequentially. This is the most stable option with the most readable logging.\n",
      "        If \"parallel\", models will be fit in parallel with ray, splitting available compute between them.\n",
      "            Note: \"parallel\" is experimental and may run into issues. It was first added in version 1.2.0.\n",
      "        For machines with 16 or more CPU cores, it is likely that \"parallel\" will be faster than \"sequential\".\n",
      "    \n",
      "        .. versionadded:: 1.2.0\n",
      "    \n",
      "    memory_limit: float | str, default = \"auto\"\n",
      "        The total amount of memory in GB you want AutoGluon predictor to use. \"auto\" means AutoGluon will use all available memory on the system\n",
      "        (that is detectable by psutil).\n",
      "        Note that this is only a soft limit! AutoGluon uses this limit to skip training models that are expected to require too much memory or stop\n",
      "        training a model that would exceed the memory limit. AutoGluon does not guarantee the enforcement of this limit (yet). Nevertheless, we expect\n",
      "        AutoGluon to abide by the limit in most cases or, at most, go over the limit by a small margin.\n",
      "        For most virtualized systems (e.g., in the cloud) and local usage on a server or laptop, \"auto\" is ideal for this parameter. We recommend manually\n",
      "        setting the memory limit (and any other resources) on systems with shared resources that are controlled by the operating system (e.g., SLURM and\n",
      "        cgroups). Otherwise, AutoGluon might wrongly assume more resources are available for fitting a model than the operating system allows,\n",
      "        which can result in model training failing or being very inefficient.\n",
      "    callbacks : List[AbstractCallback], default = None\n",
      "        :::{warning}\n",
      "        Callbacks are an experimental feature and may change in future releases without warning.\n",
      "        Callback support is preliminary and targeted towards developers.\n",
      "        :::\n",
      "        A list of callback objects inheriting from `autogluon.core.callbacks.AbstractCallback`.\n",
      "        These objects will be called before and after each model fit within trainer.\n",
      "        They have the ability to skip models or early stop the training process.\n",
      "        They can also theoretically change the entire logical flow of the trainer code by interacting with the passed `trainer` object.\n",
      "        For more details, refer to `AbstractCallback` source code.\n",
      "        If None, no callback objects will be used.\n",
      "    \n",
      "        [Note] Callback objects can be mutated in-place by the fit call if they are stateful.\n",
      "        Ensure that you avoid re-using a mutated callback object between multiple fit calls.\n",
      "    \n",
      "        [Note] Callback objects are deleted from trainer at the end of the fit call. They will not impact operations such as `refit_full` or `fit_extra`.\n",
      "    **kwargs :\n",
      "        auto_stack : bool, default = False\n",
      "            Whether AutoGluon should automatically utilize bagging and multi-layer stack ensembling to boost predictive accuracy.\n",
      "            Set this = True if you are willing to tolerate longer training times in order to maximize predictive accuracy!\n",
      "            Automatically sets `num_bag_folds` and `num_stack_levels` arguments based on dataset properties.\n",
      "            Note: Setting `num_bag_folds` and `num_stack_levels` arguments will override `auto_stack`.\n",
      "            Note: This can increase training time (and inference time) by up to 20x, but can greatly improve predictive performance.\n",
      "        num_bag_folds : int, default = None\n",
      "            Number of folds used for bagging of models. When `num_bag_folds = k`, training time is roughly increased by a factor of `k` (set = 0 to disable bagging).\n",
      "            Disabled by default (0), but we recommend values between 5-10 to maximize predictive performance.\n",
      "            Increasing num_bag_folds will result in models with lower bias but that are more prone to overfitting.\n",
      "            `num_bag_folds = 1` is an invalid value, and will raise a ValueError.\n",
      "            Values > 10 may produce diminishing returns, and can even harm overall results due to overfitting.\n",
      "            To further improve predictions, avoid increasing `num_bag_folds` much beyond 10 and instead increase `num_bag_sets`.\n",
      "        num_bag_sets : int, default = None\n",
      "            Number of repeats of kfold bagging to perform (values must be >= 1). Total number of models trained during bagging = `num_bag_folds * num_bag_sets`.\n",
      "            Defaults to 1 when unspecified. Value is ignored if `num_bag_folds<=2`.\n",
      "            Values greater than 1 will result in superior predictive performance, especially on smaller problems and with stacking enabled (reduces overall variance).\n",
      "            Be warned: This will drastically increase overall runtime, and if using a time limit, can very commonly lead to worse performance.\n",
      "            It is recommended to increase this value only as a last resort, as it is the least computationally efficient method to improve performance.\n",
      "        num_stack_levels : int, default = None\n",
      "            Number of stacking levels to use in stack ensemble. Roughly increases model training time by factor of `num_stack_levels+1` (set = 0 to disable stack ensembling).\n",
      "            Disabled by default (0), but we recommend `num_stack_levels=1` to maximize predictive performance.\n",
      "            To prevent overfitting, `num_bag_folds >= 2` must also be set or else a ValueError will be raised.\n",
      "        delay_bag_sets : bool, default = False\n",
      "            Controls when repeats of kfold bagging are executed in AutoGluon when under a time limit.\n",
      "            We suggest sticking to `False` to avoid overfitting.\n",
      "                If True, AutoGluon delays repeating kfold bagging until after evaluating all models\n",
      "                    from `hyperparameters`, if there is enough time. This allows AutoGluon to explore\n",
      "                    more hyperparameters to obtain a better final performance but it may lead to\n",
      "                    more overfitting.\n",
      "                If False, AutoGluon repeats kfold bagging immediately after evaluating each model.\n",
      "                    Thus, AutoGluon might evaluate fewer models with less overfitting.\n",
      "        holdout_frac : float, default = None\n",
      "            Fraction of train_data to holdout as tuning data for optimizing hyperparameters (ignored unless `tuning_data = None`, ignored if `num_bag_folds != 0` unless `use_bag_holdout == True`).\n",
      "            Default value (if None) is selected based on the number of rows in the training data. Default values range from 0.2 at 2,500 rows to 0.01 at 250,000 rows.\n",
      "            Default value is doubled if `hyperparameter_tune_kwargs` is set, up to a maximum of 0.2.\n",
      "            Disabled if `num_bag_folds >= 2` unless `use_bag_holdout == True`.\n",
      "        use_bag_holdout : bool | str, default = False\n",
      "            If True, a `holdout_frac` portion of the data is held-out from model bagging.\n",
      "            This held-out data is only used to score models and determine weighted ensemble weights.\n",
      "            Enable this if there is a large gap between score_val and score_test in stack models.\n",
      "            Note: If `tuning_data` was specified, `tuning_data` is used as the holdout data.\n",
      "            Disabled if not bagging.\n",
      "            If \"auto\", will be set to True if the training data has >= 1000000 rows, else it will be set to False.\n",
      "        hyperparameter_tune_kwargs : str or dict, default = None\n",
      "            Hyperparameter tuning strategy and kwargs (for example, how many HPO trials to run).\n",
      "            If None, then hyperparameter tuning will not be performed.\n",
      "            You can either choose to provide a preset\n",
      "                Valid preset values:\n",
      "                    'auto': Performs HPO via bayesian optimization search on NN_TORCH and FASTAI models, and random search on other models using local scheduler.\n",
      "                    'random': Performs HPO via random search using local scheduler.\n",
      "            Or provide a dict to specify searchers and schedulers\n",
      "                Valid keys:\n",
      "                    'num_trials': How many HPO trials to run\n",
      "                    'scheduler': Which scheduler to use\n",
      "                        Valid values:\n",
      "                            'local': Local scheduler that schedule trials FIFO\n",
      "                    'searcher': Which searching algorithm to use\n",
      "                        'local_random': Uses the 'random' searcher\n",
      "                        'random': Perform random search\n",
      "                        'auto': Perform bayesian optimization search on NN_TORCH and FASTAI models. Perform random search on other models.\n",
      "                The 'scheduler' and 'searcher' key are required when providing a dict.\n",
      "                An example of a valid dict:\n",
      "                    hyperparameter_tune_kwargs = {\n",
      "                        'num_trials': 5,\n",
      "                        'scheduler' : 'local',\n",
      "                        'searcher': 'auto',\n",
      "                    }\n",
      "        feature_prune_kwargs: dict, default = None\n",
      "            Performs layer-wise feature pruning via recursive feature elimination with permutation feature importance.\n",
      "            This fits all models in a stack layer once, discovers a pruned set of features, fits all models in the stack layer\n",
      "            again with the pruned set of features, and updates input feature lists for models whose validation score improved.\n",
      "            If None, do not perform feature pruning. If empty dictionary, perform feature pruning with default configurations.\n",
      "            For valid dictionary keys, refer to :class:`autogluon.core.utils.feature_selection.FeatureSelector` and\n",
      "            `autogluon.core.trainer.abstract_trainer.AbstractTrainer._proxy_model_feature_prune` documentation.\n",
      "            To force all models to work with the pruned set of features, set force_prune=True in the dictionary.\n",
      "        ag_args : dict, default = None\n",
      "            Keyword arguments to pass to all models (i.e. common hyperparameters shared by all AutoGluon models).\n",
      "            See the `ag_args` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      "            Identical to specifying `ag_args` parameter for all models in `hyperparameters`.\n",
      "            If a key in `ag_args` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      "        ag_args_fit : dict, default = None\n",
      "            Keyword arguments to pass to all models.\n",
      "            See the `ag_args_fit` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      "            Identical to specifying `ag_args_fit` parameter for all models in `hyperparameters`.\n",
      "            If a key in `ag_args_fit` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      "        ag_args_ensemble : dict, default = None\n",
      "            Keyword arguments to pass to all models.\n",
      "            See the `ag_args_ensemble` argument from \"Advanced functionality: Custom AutoGluon model arguments\" in the `hyperparameters` argument documentation for valid values.\n",
      "            Identical to specifying `ag_args_ensemble` parameter for all models in `hyperparameters`.\n",
      "            If a key in `ag_args_ensemble` is already specified for a model in `hyperparameters`, it will not be altered through this argument.\n",
      "        ds_args : dict, see below for default\n",
      "            Keyword arguments for dynamic stacking, only used if `dynamic_stacking=True`. These keyword arguments control the behavior of dynamic stacking\n",
      "            and determine how AutoGluon tries to detect stacked overfitting. To detect stacked overfitting, AutoGluon will fit itself (so called sub-fits)\n",
      "            on a subset (for holdout) or multiple subsets (for repeated cross-validation) and use the predictions of AutoGluon on the validation data to\n",
      "            detect stacked overfitting. The sub-fits stop and stacking will be disabled if any sub-fit shows stacked overfitting.\n",
      "            Allowed keys and values are:\n",
      "                `detection_time_frac` : float in (0,1), default = 1/4\n",
      "                    Determines how much of the original training time is used for detecting stacked overfitting.\n",
      "                    When using (repeated) cross-validation, each sub-fit will be fit for `1/n_splits * detection_time_frac * time_limit`.\n",
      "                    If no time limit is given to AutoGluon, this parameter is ignored and AutoGluon is fit without a time limit in the sub-fit.\n",
      "                `validation_procedure`: str, default = 'holdout'\n",
      "                    Determines the validation procedure used to detect stacked overfitting. Can be either `cv` or `holdout`.\n",
      "                        If `validation_procedure='holdout'` and `holdout_data` is not specified (default), then `holdout_frac` determines the holdout data.\n",
      "                        If `validation_procedure='holdout'` and `holdout_data` is specified, then the provided `holdout_data` is used for validation.\n",
      "                        If `validation_procedure='cv'`, `n_folds` and `n_repeats` determine the kind cross-validation procedure.\n",
      "                `holdout_frac` : float in (0,1), default = 1/9\n",
      "                    Determines how much of the original training data is used for the holdout data during holdout validation.\n",
      "                    Ignored if `holdout_data` is not None.\n",
      "                `n_folds` : int in [2, +inf), default = 2\n",
      "                    Number of folds to use for cross-validation.\n",
      "                `n_repeats` : int [1, +inf), default = 1\n",
      "                    Number of repeats to use for repeated cross-validation. If set to 1, performs 1-repeated cross-validation which is equivalent to\n",
      "                    cross-validation without repeats.\n",
      "                `memory_safe_fits` : bool, default = True\n",
      "                    If True, AutoGluon runs each sub-fit in a ray-based subprocess to avoid memory leakage that exist due to Python's lackluster\n",
      "                    garbage collector.\n",
      "                `clean_up_fits` : bool, default = True\n",
      "                    If True, AutoGluon will remove all saved information from sub-fits from disk.\n",
      "                    If False, the sub-fits are kept on disk and `self._sub_fits` will store paths to the sub-fits, which can be loaded just like any other\n",
      "                    predictor from disk using `TabularPredictor.load()`.\n",
      "                `enable_ray_logging` : bool, default = True\n",
      "                    If True, will log the dynamic stacking sub-fit when ray is used (`memory_safe_fits=True`).\n",
      "                    Note that because of how ray works, this may cause extra unwanted logging in the main fit process after dynamic stacking completes.\n",
      "                `enable_callbacks` : bool, default = False\n",
      "                    If True, will perform a deepcopy on the specified user callbacks and enable them during the DyStack call.\n",
      "                    If False, will not include callbacks in the DyStack call.\n",
      "                `holdout_data`: str or :class:`pd.DataFrame`, default = None\n",
      "                    Another dataset containing validation data reserved for detecting stacked overfitting. This dataset should be in the same format as\n",
      "                    `train_data`. If str is passed, `holdout_data` will be loaded using the str value as the file path.\n",
      "                    If `holdout_data` is not None, the sub-fit is fit on all of `train_data` and the full fit is fit on all of `train_data` and\n",
      "                    `holdout_data` combined.\n",
      "        included_model_types : list, default = None\n",
      "            To only include listed model types for training during `fit()`.\n",
      "            Models that are listed in `included_model_types` but not in `hyperparameters` will be ignored.\n",
      "            Reference `hyperparameters` documentation for what models correspond to each value.\n",
      "            Useful when only a subset of model needs to be trained and the `hyperparameters` dictionary is difficult or time-consuming.\n",
      "                Example: To include both 'GBM' and 'FASTAI' models, specify `included_model_types=['GBM', 'FASTAI']`.\n",
      "        excluded_model_types : list, default = None\n",
      "            Banned subset of model types to avoid training during `fit()`, even if present in `hyperparameters`.\n",
      "            Reference `hyperparameters` documentation for what models correspond to each value.\n",
      "            Useful when a particular model type such as 'KNN' or 'custom' is not desired but altering the `hyperparameters` dictionary is difficult or time-consuming.\n",
      "                Example: To exclude both 'KNN' and 'custom' models, specify `excluded_model_types=['KNN', 'custom']`.\n",
      "        refit_full : bool or str, default = False\n",
      "            Whether to retrain all models on all of the data (training + validation) after the normal training procedure.\n",
      "            This is equivalent to calling `predictor.refit_full(model=refit_full)` after fit.\n",
      "            If `refit_full=True`, it will be treated as `refit_full='all'`.\n",
      "            If `refit_full=False`, refitting will not occur.\n",
      "            Valid str values:\n",
      "                `all`: refits all models.\n",
      "                `best`: refits only the best model (and its ancestors if it is a stacker model).\n",
      "                `{model_name}`: refits only the specified model (and its ancestors if it is a stacker model).\n",
      "            For bagged models:\n",
      "                Reduces a model's inference time by collapsing bagged ensembles into a single model fit on all of the training data.\n",
      "                This process will typically result in a slight accuracy reduction and a large inference speedup.\n",
      "                The inference speedup will generally be between 10-200x faster than the original bagged ensemble model.\n",
      "                    The inference speedup factor is equivalent to (k * n), where k is the number of folds (`num_bag_folds`) and n is the number of finished repeats (`num_bag_sets`) in the bagged ensemble.\n",
      "                The runtime is generally 10% or less of the original fit runtime.\n",
      "                    The runtime can be roughly estimated as 1 / (k * n) of the original fit runtime, with k and n defined above.\n",
      "            For non-bagged models:\n",
      "                Optimizes a model's accuracy by retraining on 100% of the data without using a validation set.\n",
      "                Will typically result in a slight accuracy increase and no change to inference time.\n",
      "                The runtime will be approximately equal to the original fit runtime.\n",
      "            This process does not alter the original models, but instead adds additional models.\n",
      "            If stacker models are refit by this process, they will use the refit_full versions of the ancestor models during inference.\n",
      "            Models produced by this process will not have validation scores, as they use all of the data for training.\n",
      "                Therefore, it is up to the user to determine if the models are of sufficient quality by including test data in `predictor.leaderboard(test_data)`.\n",
      "                If the user does not have additional test data, they should reference the original model's score for an estimate of the performance of the refit_full model.\n",
      "                    Warning: Be aware that utilizing refit_full models without separately verifying on test data means that the model is untested, and has no guarantee of being consistent with the original model.\n",
      "            The time taken by this process is not enforced by `time_limit`.\n",
      "        save_bag_folds : bool, default = True\n",
      "            If True, will save the bagged fold models to disk.\n",
      "            If False, will not save the bagged fold models, only keeping their metadata and out-of-fold predictions.\n",
      "                Note: The bagged models will not be available for prediction, only use this if you intend to call `refit_full`.\n",
      "                The purpose of setting it to False is that it greatly decreases the peak disk usage of the predictor during the fit call when bagging.\n",
      "                Note that this makes refit_full slightly more likely to crash in scenarios where the dataset is large relative to available system memory.\n",
      "                This is because by default, refit_full will fall back to cloning the first fold of the bagged model in case it lacks memory to refit.\n",
      "                However, if `save_bag_folds=False`, this fallback isn't possible, as there is not fold model to clone because it wasn't saved.\n",
      "                In this scenario, refit will raise an exception for `save_bag_folds=False`, but will succeed if `save_bag_folds=True`.\n",
      "            Final disk usage of predictor will be identical regardless of the setting after `predictor.delete_models(models_to_keep=\"best\", dry_run=False)` is called post-fit.\n",
      "        set_best_to_refit_full : bool, default = False\n",
      "            If True, will change the default model that Predictor uses for prediction when model is not specified to the refit_full version of the model that exhibited the highest validation score.\n",
      "            Only valid if `refit_full` is set.\n",
      "        keep_only_best : bool, default = False\n",
      "            If True, only the best model and its ancestor models are saved in the outputted `predictor`. All other models are deleted.\n",
      "                If you only care about deploying the most accurate predictor with the smallest file-size and no longer need any of the other trained models or functionality beyond prediction on new data, then set: `keep_only_best=True`, `save_space=True`.\n",
      "                This is equivalent to calling `predictor.delete_models(models_to_keep='best', dry_run=False)` directly after `fit()`.\n",
      "            If used with `refit_full` and `set_best_to_refit_full`, the best model will be the refit_full model, and the original bagged best model will be deleted.\n",
      "                `refit_full` will be automatically set to 'best' in this case to avoid training models which will be later deleted.\n",
      "        save_space : bool, default = False\n",
      "            If True, reduces the memory and disk size of predictor by deleting auxiliary model files that aren't needed for prediction on new data.\n",
      "                This is equivalent to calling `predictor.save_space()` directly after `fit()`.\n",
      "            This has NO impact on inference accuracy.\n",
      "            It is recommended if the only goal is to use the trained model for prediction.\n",
      "            Certain advanced functionality may no longer be available if `save_space=True`. Refer to `predictor.save_space()` documentation for more details.\n",
      "        feature_generator : :class:`autogluon.features.generators.AbstractFeatureGenerator`, default = :class:`autogluon.features.generators.AutoMLPipelineFeatureGenerator`\n",
      "            The feature generator used by AutoGluon to process the input data to the form sent to the models. This often includes automated feature generation and data cleaning.\n",
      "            It is generally recommended to keep the default feature generator unless handling an advanced use-case.\n",
      "            To control aspects of the default feature generation process, you can pass in an :class:`AutoMLPipelineFeatureGenerator` object constructed using some of these kwargs:\n",
      "                enable_numeric_features : bool, default True\n",
      "                    Whether to keep features of 'int' and 'float' raw types.\n",
      "                    These features are passed without alteration to the models.\n",
      "                    Appends IdentityFeatureGenerator(infer_features_in_args=dict(valid_raw_types=['int', 'float']))) to the generator group.\n",
      "                enable_categorical_features : bool, default True\n",
      "                    Whether to keep features of 'object' and 'category' raw types.\n",
      "                    These features are processed into memory optimized 'category' features.\n",
      "                    Appends CategoryFeatureGenerator() to the generator group.\n",
      "                enable_datetime_features : bool, default True\n",
      "                    Whether to keep features of 'datetime' raw type and 'object' features identified as 'datetime_as_object' features.\n",
      "                    These features will be converted to 'int' features representing milliseconds since epoch.\n",
      "                    Appends DatetimeFeatureGenerator() to the generator group.\n",
      "                enable_text_special_features : bool, default True\n",
      "                    Whether to use 'object' features identified as 'text' features to generate 'text_special' features such as word count, capital letter ratio, and symbol counts.\n",
      "                    Appends TextSpecialFeatureGenerator() to the generator group.\n",
      "                enable_text_ngram_features : bool, default True\n",
      "                    Whether to use 'object' features identified as 'text' features to generate 'text_ngram' features.\n",
      "                    Appends TextNgramFeatureGenerator(vectorizer=vectorizer) to the generator group.\n",
      "                enable_raw_text_features : bool, default False\n",
      "                    Whether to keep the raw text features.\n",
      "                    Appends IdentityFeatureGenerator(infer_features_in_args=dict(required_special_types=['text'])) to the generator group.\n",
      "                vectorizer : CountVectorizer, default CountVectorizer(min_df=30, ngram_range=(1, 3), max_features=10000, dtype=np.uint8)\n",
      "                    sklearn CountVectorizer object to use in TextNgramFeatureGenerator.\n",
      "                    Only used if `enable_text_ngram_features=True`.\n",
      "        unlabeled_data : pd.DataFrame, default = None\n",
      "            [Experimental Parameter]\n",
      "            Collection of data without labels that we can use to pretrain on. This is the same schema as train_data, except\n",
      "            without the labels. Currently, unlabeled_data is only used for pretraining a TabTransformer model.\n",
      "            If you do not specify 'TRANSF' with unlabeled_data, then no pretraining will occur and unlabeled_data will be ignored!\n",
      "            After the pretraining step, we will finetune using the TabTransformer model as well. If TabTransformer is ensembled\n",
      "            with other models, like in typical AutoGluon fashion, then the output of this \"pretrain/finetune\" will be ensembled\n",
      "            with other models, which will not used the unlabeled_data. The \"pretrain/finetune flow\" is also known as semi-supervised learning.\n",
      "            The typical use case for unlabeled_data is to add signal to your model where you may not have sufficient training\n",
      "            data. e.g. 500 hand-labeled samples (perhaps a hard human task), whole data set (unlabeled) is thousands/millions.\n",
      "            However, this isn't the only use case. Given enough unlabeled data(millions of rows), you may see improvements\n",
      "            to any amount of labeled data.\n",
      "        verbosity : int\n",
      "            If specified, overrides the existing `predictor.verbosity` value.\n",
      "        raise_on_no_models_fitted: bool, default = True\n",
      "            If True, will raise a RuntimeError if no models were successfully fit during `fit()`.\n",
      "        calibrate: bool or str, default = 'auto'\n",
      "            Note: It is recommended to use ['auto', False] as the values and avoid True.\n",
      "            If 'auto' will automatically set to True if the problem_type and eval_metric are suitable for calibration.\n",
      "            If True and the problem_type is classification, temperature scaling will be used to calibrate the Predictor's estimated class probabilities\n",
      "            (which may improve metrics like log_loss) and will train a scalar parameter on the validation set.\n",
      "            If True and the problem_type is quantile regression, conformalization will be used to calibrate the Predictor's estimated quantiles\n",
      "            (which may improve the prediction interval coverage, and bagging could further improve it) and will compute a set of scalar parameters on the validation set.\n",
      "        test_data : str or :class:`pd.DataFrame`, default = None\n",
      "            Table of the test data.\n",
      "            If str is passed, `test_data` will be loaded using the str value as the file path.\n",
      "            NOTE: This test_data is NEVER SEEN by the model during training and, if specified, is only used for logging purposes (i.e. for learning curve generation).\n",
      "            This test_data should be treated the same way test data is used in predictor.leaderboard.\n",
      "        learning_curves : bool or dict, default = None\n",
      "            If bool and is True, default learning curve hyperparameter ag_args will be initialized for each of the models included in the ensemble.\n",
      "                By default, learning curves will include eval_metric scores specified in fit call arguments.\n",
      "                This can be overwritten as shown below.\n",
      "            If dict, user can pass learning_curves parameters to be initialized as ag_args in the following format:\n",
      "                learning_curves = {\n",
      "                    \"metrics\": str or list(str) or Scorer or list(Scorer):\n",
      "                        autogluon metric scorer(s) to be calculated at each iteration, represented as Scorer object(s) or scorer name(s) (str)\n",
      "                    \"use_error\": bool : whether to use error or score format for metrics listed above\n",
      "                }\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    :class:`TabularPredictor` object. Returns self.\n",
      "    \n",
      "    Examples\n",
      "    --------\n",
      "    >>> from autogluon.tabular import TabularDataset, TabularPredictor\n",
      "    >>> train_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/train.csv')\n",
      "    >>> label = 'class'\n",
      "    >>> predictor = TabularPredictor(label=label).fit(train_data)\n",
      "    >>> test_data = TabularDataset('https://autogluon.s3.amazonaws.com/datasets/Inc/test.csv')\n",
      "    >>> leaderboard = predictor.leaderboard(test_data)\n",
      "    >>> y_test = test_data[label]\n",
      "    >>> test_data = test_data.drop(columns=[label])\n",
      "    >>> y_pred = predictor.predict(test_data)\n",
      "    >>> perf = predictor.evaluate_predictions(y_true=y_test, y_pred=y_pred)\n",
      "    \n",
      "    To maximize predictive performance, use the following:\n",
      "    \n",
      "    >>> eval_metric = 'roc_auc'  # set this to the metric you ultimately care about\n",
      "    >>> time_limit = 3600  # set as long as you are willing to wait (in sec)\n",
      "    >>> predictor = TabularPredictor(label=label, eval_metric=eval_metric).fit(train_data, presets=['best_quality'], time_limit=time_limit)\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(help(TabularPredictor.fit))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b32b63cb-af16-4169-afee-e11db28c62b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Verbosity: 2 (Standard Logging)\n",
      "=================== System Info ===================\n",
      "AutoGluon Version:  1.2\n",
      "Python Version:     3.11.9\n",
      "Operating System:   Linux\n",
      "Platform Machine:   x86_64\n",
      "Platform Version:   #53~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Jan 15 19:18:46 UTC 2\n",
      "CPU Count:          32\n",
      "Memory Avail:       47.66 GB / 61.92 GB (77.0%)\n",
      "Disk Space Avail:   49.85 GB / 906.65 GB (5.5%)\n",
      "===================================================\n",
      "Presets specified: ['best_quality']\n",
      "Beginning AutoGluon training ... Time limit = 3600s\n",
      "AutoGluon will save models to \"/home/fnerin/Desktop/allodb_new/models/residues_physchem\"\n",
      "Train Data Rows:    108406\n",
      "Train Data Columns: 166\n",
      "Label Column:       Label_label\n",
      "Problem Type:       binary\n",
      "Preprocessing data ...\n",
      "Selected class <--> label mapping:  class 1 = 1, class 0 = 0\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    48765.57 MB\n",
      "\tTrain Data (Original)  Memory Usage: 137.29 MB (0.3% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\t\t\tNote: Converting 28 features to boolean dtype as they only contain 2 unique values.\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tStage 5 Generators:\n",
      "\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\tUseless Original Features (Count: 1): ['Amino acids_label_comp_id_X']\n",
      "\t\tThese features carry no predictive signal and should be manually investigated.\n",
      "\t\tThis is typically a feature which has the same value for all rows.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\tUnused Original Features (Count: 1): ['Graphein_hphob_fauchere']\n",
      "\t\tThese features were not used to generate any of the output features. Add a feature generator compatible with these features to utilize them.\n",
      "\t\tFeatures can also be unused if they carry very little information, such as being categorical but having almost entirely unique values or being duplicates of other features.\n",
      "\t\tThese features do not need to be present at inference time.\n",
      "\t\t('float', []) : 1 | ['Graphein_hphob_fauchere']\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 131 | ['Graphein_dim_1', 'Graphein_dim_2', 'Graphein_dim_3', 'Graphein_dim_4', 'Graphein_dim_5', ...]\n",
      "\t\t('int', [])   :  33 | ['Amino acids_label_comp_id_A', 'Amino acids_label_comp_id_C', 'Amino acids_label_comp_id_D', 'Amino acids_label_comp_id_E', 'Amino acids_label_comp_id_F', ...]\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('float', [])     : 131 | ['Graphein_dim_1', 'Graphein_dim_2', 'Graphein_dim_3', 'Graphein_dim_4', 'Graphein_dim_5', ...]\n",
      "\t\t('int', [])       :   5 | ['DSSP_NH_O_1_relidx', 'DSSP_O_NH_1_relidx', 'DSSP_NH_O_2_relidx', 'DSSP_O_NH_2_relidx', 'Melodia_propensity']\n",
      "\t\t('int', ['bool']) :  28 | ['Amino acids_label_comp_id_A', 'Amino acids_label_comp_id_C', 'Amino acids_label_comp_id_D', 'Amino acids_label_comp_id_E', 'Amino acids_label_comp_id_F', ...]\n",
      "\t0.6s = Fit runtime\n",
      "\t164 features in original data used to generate 164 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 115.38 MB (0.2% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 0.62s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'mcc'\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "Automatically generating train/validation split with holdout_frac=0.023061454163053705, Train Rows: 105906, Val Rows: 2500\n",
      "Large model count detected (112 configs) ... Only displaying the first 3 models of each family. To see all, set `verbosity=3`.\n",
      "User-specified model hyperparameters to be fit:\n",
      "{\n",
      "\t'NN_TORCH': [{}, {'activation': 'elu', 'dropout_prob': 0.10077639529843717, 'hidden_size': 108, 'learning_rate': 0.002735937344002146, 'num_layers': 4, 'use_batchnorm': True, 'weight_decay': 1.356433327634438e-12, 'ag_args': {'name_suffix': '_r79', 'priority': -2}}, {'activation': 'elu', 'dropout_prob': 0.11897478034205347, 'hidden_size': 213, 'learning_rate': 0.0010474382260641949, 'num_layers': 4, 'use_batchnorm': False, 'weight_decay': 5.594471067786272e-10, 'ag_args': {'name_suffix': '_r22', 'priority': -7}}],\n",
      "\t'GBM': [{'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}}, {}, {'learning_rate': 0.03, 'num_leaves': 128, 'feature_fraction': 0.9, 'min_data_in_leaf': 3, 'ag_args': {'name_suffix': 'Large', 'priority': 0, 'hyperparameter_tune_kwargs': None}}],\n",
      "\t'CAT': [{}, {'depth': 6, 'grow_policy': 'SymmetricTree', 'l2_leaf_reg': 2.1542798306067823, 'learning_rate': 0.06864209415792857, 'max_ctr_complexity': 4, 'one_hot_max_size': 10, 'ag_args': {'name_suffix': '_r177', 'priority': -1}}, {'depth': 8, 'grow_policy': 'Depthwise', 'l2_leaf_reg': 2.7997999596449104, 'learning_rate': 0.031375015734637225, 'max_ctr_complexity': 2, 'one_hot_max_size': 3, 'ag_args': {'name_suffix': '_r9', 'priority': -5}}],\n",
      "\t'XGB': [{}, {'colsample_bytree': 0.6917311125174739, 'enable_categorical': False, 'learning_rate': 0.018063876087523967, 'max_depth': 10, 'min_child_weight': 0.6028633586934382, 'ag_args': {'name_suffix': '_r33', 'priority': -8}}, {'colsample_bytree': 0.6628423832084077, 'enable_categorical': False, 'learning_rate': 0.08775715546881824, 'max_depth': 5, 'min_child_weight': 0.6294123374222513, 'ag_args': {'name_suffix': '_r89', 'priority': -16}}],\n",
      "\t'FASTAI': [{}, {'bs': 256, 'emb_drop': 0.5411770367537934, 'epochs': 43, 'layers': [800, 400], 'lr': 0.01519848858318159, 'ps': 0.23782946566604385, 'ag_args': {'name_suffix': '_r191', 'priority': -4}}, {'bs': 2048, 'emb_drop': 0.05070411322605811, 'epochs': 29, 'layers': [200, 100], 'lr': 0.08974235041576624, 'ps': 0.10393466140748028, 'ag_args': {'name_suffix': '_r102', 'priority': -11}}],\n",
      "\t'RF': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'XT': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'KNN': [{'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}}, {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}}],\n",
      "}\n",
      "Fitting 110 L1 models, fit_strategy=\"sequential\" ...\n",
      "Fitting model: KNeighborsUnif ... Training model for up to 3599.38s of the 3599.38s of remaining time.\n",
      "\t0.1314\t = Validation score   (mcc)\n",
      "\t0.21s\t = Training   runtime\n",
      "\t0.17s\t = Validation runtime\n",
      "Fitting model: KNeighborsDist ... Training model for up to 3598.90s of the 3598.90s of remaining time.\n",
      "\t0.2963\t = Validation score   (mcc)\n",
      "\t0.21s\t = Training   runtime\n",
      "\t0.14s\t = Validation runtime\n",
      "Fitting model: LightGBMXT ... Training model for up to 3598.46s of the 3598.46s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.12s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: LightGBM ... Training model for up to 3597.33s of the 3597.33s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t0.66s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: RandomForestGini ... Training model for up to 3596.66s of the 3596.66s of remaining time.\n",
      "\t0.3043\t = Validation score   (mcc)\n",
      "\t7.54s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: RandomForestEntr ... Training model for up to 3588.90s of the 3588.90s of remaining time.\n",
      "\t0.3335\t = Validation score   (mcc)\n",
      "\t6.81s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: CatBoost ... Training model for up to 3581.90s of the 3581.90s of remaining time.\n",
      "\t0.4174\t = Validation score   (mcc)\n",
      "\t51.5s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: ExtraTreesGini ... Training model for up to 3530.38s of the 3530.38s of remaining time.\n",
      "\t0.3517\t = Validation score   (mcc)\n",
      "\t3.56s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: ExtraTreesEntr ... Training model for up to 3526.44s of the 3526.44s of remaining time.\n",
      "\t0.3235\t = Validation score   (mcc)\n",
      "\t3.52s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI ... Training model for up to 3522.55s of the 3522.54s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 9: early stopping\n",
      "\t0.0557\t = Validation score   (mcc)\n",
      "\t39.23s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: XGBoost ... Training model for up to 3483.25s of the 3483.25s of remaining time.\n",
      "\t0.0783\t = Validation score   (mcc)\n",
      "\t1.29s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch ... Training model for up to 3481.95s of the 3481.94s of remaining time.\n",
      "\t0.3337\t = Validation score   (mcc)\n",
      "\t291.13s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: LightGBMLarge ... Training model for up to 3190.78s of the 3190.78s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.76s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: CatBoost_r177 ... Training model for up to 3189.01s of the 3189.00s of remaining time.\n",
      "\t0.4257\t = Validation score   (mcc)\n",
      "\t49.46s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r79 ... Training model for up to 3139.53s of the 3139.52s of remaining time.\n",
      "\t0.3962\t = Validation score   (mcc)\n",
      "\t150.63s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: LightGBM_r131 ... Training model for up to 2988.85s of the 2988.85s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.37s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r191 ... Training model for up to 2987.48s of the 2987.48s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 4: early stopping\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t94.31s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: CatBoost_r9 ... Training model for up to 2893.11s of the 2893.10s of remaining time.\n",
      "\t0.3851\t = Validation score   (mcc)\n",
      "\t53.82s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: LightGBM_r96 ... Training model for up to 2839.26s of the 2839.26s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t0.65s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r22 ... Training model for up to 2838.60s of the 2838.60s of remaining time.\n",
      "\t0.3742\t = Validation score   (mcc)\n",
      "\t304.81s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: XGBoost_r33 ... Training model for up to 2533.75s of the 2533.74s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t15.88s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: ExtraTrees_r42 ... Training model for up to 2517.85s of the 2517.85s of remaining time.\n",
      "\t0.2939\t = Validation score   (mcc)\n",
      "\t43.43s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: CatBoost_r137 ... Training model for up to 2474.18s of the 2474.18s of remaining time.\n",
      "\t0.3894\t = Validation score   (mcc)\n",
      "\t108.38s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r102 ... Training model for up to 2365.79s of the 2365.78s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "\t0.115\t = Validation score   (mcc)\n",
      "\t20.15s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: CatBoost_r13 ... Training model for up to 2345.60s of the 2345.60s of remaining time.\n",
      "\t0.3857\t = Validation score   (mcc)\n",
      "\t210.23s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: RandomForest_r195 ... Training model for up to 2135.32s of the 2135.32s of remaining time.\n",
      "\t0.3427\t = Validation score   (mcc)\n",
      "\t147.72s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: LightGBM_r188 ... Training model for up to 1987.09s of the 1987.09s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t0.73s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r145 ... Training model for up to 1986.35s of the 1986.35s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 4: early stopping\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t122.8s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: XGBoost_r89 ... Training model for up to 1863.45s of the 1863.45s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t22.32s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r30 ... Training model for up to 1841.12s of the 1841.12s of remaining time.\n",
      "\t0.3152\t = Validation score   (mcc)\n",
      "\t303.39s\t = Training   runtime\n",
      "\t0.04s\t = Validation runtime\n",
      "Fitting model: LightGBM_r130 ... Training model for up to 1537.68s of the 1537.68s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.33s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r86 ... Training model for up to 1536.34s of the 1536.33s of remaining time.\n",
      "\t0.3562\t = Validation score   (mcc)\n",
      "\t195.35s\t = Training   runtime\n",
      "\t0.04s\t = Validation runtime\n",
      "Fitting model: CatBoost_r50 ... Training model for up to 1340.94s of the 1340.94s of remaining time.\n",
      "\t0.3988\t = Validation score   (mcc)\n",
      "\t58.21s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r11 ... Training model for up to 1282.71s of the 1282.71s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 2: early stopping\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t120.79s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: XGBoost_r194 ... Training model for up to 1161.82s of the 1161.82s of remaining time.\n",
      "\t0.4988\t = Validation score   (mcc)\n",
      "\t24.53s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: ExtraTrees_r172 ... Training model for up to 1137.25s of the 1137.25s of remaining time.\n",
      "\t0.1568\t = Validation score   (mcc)\n",
      "\t51.43s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: CatBoost_r69 ... Training model for up to 1085.67s of the 1085.67s of remaining time.\n",
      "\t0.3988\t = Validation score   (mcc)\n",
      "\t102.56s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r103 ... Training model for up to 983.09s of the 983.09s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 5: early stopping\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t65.35s\t = Training   runtime\n",
      "\t0.17s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r14 ... Training model for up to 917.45s of the 917.45s of remaining time.\n",
      "\t0.2631\t = Validation score   (mcc)\n",
      "\t80.36s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: LightGBM_r161 ... Training model for up to 837.06s of the 837.05s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.48s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r143 ... Training model for up to 835.56s of the 835.56s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t39.99s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: CatBoost_r70 ... Training model for up to 795.54s of the 795.54s of remaining time.\n",
      "\t0.4386\t = Validation score   (mcc)\n",
      "\t57.7s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r156 ... Training model for up to 737.82s of the 737.82s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 17: early stopping\n",
      "\t-0.0114\t = Validation score   (mcc)\n",
      "\t42.8s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: LightGBM_r196 ... Training model for up to 695.00s of the 695.00s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.42s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: RandomForest_r39 ... Training model for up to 693.57s of the 693.57s of remaining time.\n",
      "\t0.2939\t = Validation score   (mcc)\n",
      "\t137.63s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: CatBoost_r167 ... Training model for up to 555.81s of the 555.81s of remaining time.\n",
      "\t0.3879\t = Validation score   (mcc)\n",
      "\t42.15s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r95 ... Training model for up to 513.65s of the 513.64s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "No improvement since epoch 2: early stopping\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t114.81s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r41 ... Training model for up to 398.75s of the 398.74s of remaining time.\n",
      "\t0.2883\t = Validation score   (mcc)\n",
      "\t213.63s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: XGBoost_r98 ... Training model for up to 185.07s of the 185.07s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.34s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: LightGBM_r15 ... Training model for up to 183.71s of the 183.71s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t1.28s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch_r158 ... Training model for up to 182.42s of the 182.42s of remaining time.\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t46.45s\t = Training   runtime\n",
      "\t0.04s\t = Validation runtime\n",
      "Fitting model: CatBoost_r86 ... Training model for up to 135.93s of the 135.93s of remaining time.\n",
      "\t0.4314\t = Validation score   (mcc)\n",
      "\t114.35s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI_r37 ... Training model for up to 21.55s of the 21.55s of remaining time.\n",
      "Metric mcc is not supported by this model - using log_loss instead\n",
      "\tRan out of time, stopping training early. (Stopping on epoch 11)\n",
      "\t0.0\t = Validation score   (mcc)\n",
      "\t21.81s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "AutoGluon training complete, total runtime = 3600.56s ... Best model: XGBoost_r194 | Estimated inference throughput: 155071.2 rows/s (2500 batch size)\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"/home/fnerin/Desktop/allodb_new/models/residues_physchem\")\n"
     ]
    }
   ],
   "source": [
    "if not os.path.isdir(path):\n",
    "    predictor = TabularPredictor(\n",
    "        label='Label_label', problem_type=\"binary\", eval_metric=mcc,\n",
    "        path=path\n",
    "    )\n",
    "    \n",
    "    fit = predictor.fit(\n",
    "        traindataset,\n",
    "        presets=\"best_quality\",\n",
    "        time_limit=3600,\n",
    "        num_gpus=0, num_cpus=30,\n",
    "        verbosity=2,\n",
    "        auto_stack=False, num_stack_levels=0,\n",
    "        fit_weighted_ensemble=False\n",
    "    )\n",
    "else:\n",
    "    fit = predictor.load(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f008a54b-8e64-4723-95f9-1016937ef9ba",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n",
      "/home/fnerin/miniconda3/envs/pybiomed/lib/python3.11/site-packages/torch/storage.py:414: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  return torch.load(io.BytesIO(b))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>score_test</th>\n",
       "      <th>score_val</th>\n",
       "      <th>eval_metric</th>\n",
       "      <th>pred_time_test</th>\n",
       "      <th>pred_time_val</th>\n",
       "      <th>fit_time</th>\n",
       "      <th>pred_time_test_marginal</th>\n",
       "      <th>pred_time_val_marginal</th>\n",
       "      <th>fit_time_marginal</th>\n",
       "      <th>stack_level</th>\n",
       "      <th>can_infer</th>\n",
       "      <th>fit_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CatBoost</td>\n",
       "      <td>0.159878</td>\n",
       "      <td>0.417384</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.147774</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>51.503006</td>\n",
       "      <td>0.147774</td>\n",
       "      <td>0.002767</td>\n",
       "      <td>51.503006</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CatBoost_r177</td>\n",
       "      <td>0.158581</td>\n",
       "      <td>0.425714</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.030552</td>\n",
       "      <td>0.003253</td>\n",
       "      <td>49.464640</td>\n",
       "      <td>0.030552</td>\n",
       "      <td>0.003253</td>\n",
       "      <td>49.464640</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CatBoost_r69</td>\n",
       "      <td>0.156772</td>\n",
       "      <td>0.398781</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.351471</td>\n",
       "      <td>0.003724</td>\n",
       "      <td>102.562262</td>\n",
       "      <td>0.351471</td>\n",
       "      <td>0.003724</td>\n",
       "      <td>102.562262</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XGBoost_r194</td>\n",
       "      <td>0.151047</td>\n",
       "      <td>0.498849</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.062239</td>\n",
       "      <td>0.016122</td>\n",
       "      <td>24.531862</td>\n",
       "      <td>0.062239</td>\n",
       "      <td>0.016122</td>\n",
       "      <td>24.531862</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CatBoost_r13</td>\n",
       "      <td>0.149776</td>\n",
       "      <td>0.385663</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.054072</td>\n",
       "      <td>0.006468</td>\n",
       "      <td>210.230490</td>\n",
       "      <td>0.054072</td>\n",
       "      <td>0.006468</td>\n",
       "      <td>210.230490</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CatBoost_r50</td>\n",
       "      <td>0.146453</td>\n",
       "      <td>0.398781</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.211605</td>\n",
       "      <td>0.008396</td>\n",
       "      <td>58.211248</td>\n",
       "      <td>0.211605</td>\n",
       "      <td>0.008396</td>\n",
       "      <td>58.211248</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NeuralNetTorch_r79</td>\n",
       "      <td>0.143356</td>\n",
       "      <td>0.396250</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.728093</td>\n",
       "      <td>0.033396</td>\n",
       "      <td>150.633491</td>\n",
       "      <td>0.728093</td>\n",
       "      <td>0.033396</td>\n",
       "      <td>150.633491</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CatBoost_r70</td>\n",
       "      <td>0.142126</td>\n",
       "      <td>0.438583</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.219357</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>57.701282</td>\n",
       "      <td>0.219357</td>\n",
       "      <td>0.006903</td>\n",
       "      <td>57.701282</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NeuralNetTorch_r14</td>\n",
       "      <td>0.141889</td>\n",
       "      <td>0.263123</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.244909</td>\n",
       "      <td>0.031864</td>\n",
       "      <td>80.355489</td>\n",
       "      <td>0.244909</td>\n",
       "      <td>0.031864</td>\n",
       "      <td>80.355489</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CatBoost_r86</td>\n",
       "      <td>0.141749</td>\n",
       "      <td>0.431392</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.037711</td>\n",
       "      <td>0.003593</td>\n",
       "      <td>114.351807</td>\n",
       "      <td>0.037711</td>\n",
       "      <td>0.003593</td>\n",
       "      <td>114.351807</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CatBoost_r167</td>\n",
       "      <td>0.141231</td>\n",
       "      <td>0.387850</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.013611</td>\n",
       "      <td>0.003651</td>\n",
       "      <td>42.147018</td>\n",
       "      <td>0.013611</td>\n",
       "      <td>0.003651</td>\n",
       "      <td>42.147018</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CatBoost_r137</td>\n",
       "      <td>0.139254</td>\n",
       "      <td>0.389369</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.021138</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>108.383495</td>\n",
       "      <td>0.021138</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>108.383495</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NeuralNetTorch_r22</td>\n",
       "      <td>0.136628</td>\n",
       "      <td>0.374239</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.286723</td>\n",
       "      <td>0.033908</td>\n",
       "      <td>304.814172</td>\n",
       "      <td>0.286723</td>\n",
       "      <td>0.033908</td>\n",
       "      <td>304.814172</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CatBoost_r9</td>\n",
       "      <td>0.136167</td>\n",
       "      <td>0.385121</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.159969</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>53.819530</td>\n",
       "      <td>0.159969</td>\n",
       "      <td>0.008481</td>\n",
       "      <td>53.819530</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NeuralNetTorch</td>\n",
       "      <td>0.134427</td>\n",
       "      <td>0.333661</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.271192</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>291.131433</td>\n",
       "      <td>0.271192</td>\n",
       "      <td>0.031589</td>\n",
       "      <td>291.131433</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NeuralNetTorch_r41</td>\n",
       "      <td>0.134242</td>\n",
       "      <td>0.288288</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.735915</td>\n",
       "      <td>0.033186</td>\n",
       "      <td>213.634619</td>\n",
       "      <td>0.735915</td>\n",
       "      <td>0.033186</td>\n",
       "      <td>213.634619</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NeuralNetTorch_r30</td>\n",
       "      <td>0.132441</td>\n",
       "      <td>0.315244</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.297065</td>\n",
       "      <td>0.035576</td>\n",
       "      <td>303.393694</td>\n",
       "      <td>0.297065</td>\n",
       "      <td>0.035576</td>\n",
       "      <td>303.393694</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NeuralNetTorch_r86</td>\n",
       "      <td>0.125108</td>\n",
       "      <td>0.356226</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.723688</td>\n",
       "      <td>0.037021</td>\n",
       "      <td>195.352012</td>\n",
       "      <td>0.723688</td>\n",
       "      <td>0.037021</td>\n",
       "      <td>195.352012</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NeuralNetFastAI_r102</td>\n",
       "      <td>0.087097</td>\n",
       "      <td>0.115016</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.092321</td>\n",
       "      <td>0.012262</td>\n",
       "      <td>20.147879</td>\n",
       "      <td>0.092321</td>\n",
       "      <td>0.012262</td>\n",
       "      <td>20.147879</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NeuralNetFastAI</td>\n",
       "      <td>0.060760</td>\n",
       "      <td>0.055686</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.832277</td>\n",
       "      <td>0.014844</td>\n",
       "      <td>39.230214</td>\n",
       "      <td>0.832277</td>\n",
       "      <td>0.014844</td>\n",
       "      <td>39.230214</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NeuralNetFastAI_r156</td>\n",
       "      <td>0.060549</td>\n",
       "      <td>-0.011430</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.509034</td>\n",
       "      <td>0.008023</td>\n",
       "      <td>42.797305</td>\n",
       "      <td>0.509034</td>\n",
       "      <td>0.008023</td>\n",
       "      <td>42.797305</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ExtraTrees_r42</td>\n",
       "      <td>0.052854</td>\n",
       "      <td>0.293917</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.263247</td>\n",
       "      <td>0.024554</td>\n",
       "      <td>43.431954</td>\n",
       "      <td>0.263247</td>\n",
       "      <td>0.024554</td>\n",
       "      <td>43.431954</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>RandomForest_r39</td>\n",
       "      <td>0.052245</td>\n",
       "      <td>0.293917</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.238857</td>\n",
       "      <td>0.025666</td>\n",
       "      <td>137.630308</td>\n",
       "      <td>0.238857</td>\n",
       "      <td>0.025666</td>\n",
       "      <td>137.630308</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>RandomForest_r195</td>\n",
       "      <td>0.051760</td>\n",
       "      <td>0.342747</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.206467</td>\n",
       "      <td>0.024380</td>\n",
       "      <td>147.718517</td>\n",
       "      <td>0.206467</td>\n",
       "      <td>0.024380</td>\n",
       "      <td>147.718517</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>RandomForestEntr</td>\n",
       "      <td>0.048248</td>\n",
       "      <td>0.333539</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.485226</td>\n",
       "      <td>0.024837</td>\n",
       "      <td>6.805090</td>\n",
       "      <td>0.485226</td>\n",
       "      <td>0.024837</td>\n",
       "      <td>6.805090</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ExtraTrees_r172</td>\n",
       "      <td>0.043153</td>\n",
       "      <td>0.156790</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.242218</td>\n",
       "      <td>0.024833</td>\n",
       "      <td>51.432341</td>\n",
       "      <td>0.242218</td>\n",
       "      <td>0.024833</td>\n",
       "      <td>51.432341</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>RandomForestGini</td>\n",
       "      <td>0.043153</td>\n",
       "      <td>0.304294</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.279097</td>\n",
       "      <td>0.024663</td>\n",
       "      <td>7.536692</td>\n",
       "      <td>0.279097</td>\n",
       "      <td>0.024663</td>\n",
       "      <td>7.536692</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ExtraTreesEntr</td>\n",
       "      <td>0.043153</td>\n",
       "      <td>0.323532</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.310825</td>\n",
       "      <td>0.025065</td>\n",
       "      <td>3.518496</td>\n",
       "      <td>0.310825</td>\n",
       "      <td>0.025065</td>\n",
       "      <td>3.518496</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ExtraTreesGini</td>\n",
       "      <td>0.043153</td>\n",
       "      <td>0.351722</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.432763</td>\n",
       "      <td>0.024593</td>\n",
       "      <td>3.560127</td>\n",
       "      <td>0.432763</td>\n",
       "      <td>0.024593</td>\n",
       "      <td>3.560127</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>KNeighborsDist</td>\n",
       "      <td>0.039795</td>\n",
       "      <td>0.296257</td>\n",
       "      <td>mcc</td>\n",
       "      <td>2.117112</td>\n",
       "      <td>0.135079</td>\n",
       "      <td>0.214777</td>\n",
       "      <td>2.117112</td>\n",
       "      <td>0.135079</td>\n",
       "      <td>0.214777</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>KNeighborsUnif</td>\n",
       "      <td>0.036282</td>\n",
       "      <td>0.131379</td>\n",
       "      <td>mcc</td>\n",
       "      <td>2.031718</td>\n",
       "      <td>0.172940</td>\n",
       "      <td>0.206955</td>\n",
       "      <td>2.031718</td>\n",
       "      <td>0.172940</td>\n",
       "      <td>0.206955</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>LightGBM_r130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.010407</td>\n",
       "      <td>0.003875</td>\n",
       "      <td>1.331746</td>\n",
       "      <td>0.010407</td>\n",
       "      <td>0.003875</td>\n",
       "      <td>1.331746</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>LightGBMLarge</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.010906</td>\n",
       "      <td>0.001192</td>\n",
       "      <td>1.764692</td>\n",
       "      <td>0.010906</td>\n",
       "      <td>0.001192</td>\n",
       "      <td>1.764692</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>LightGBM_r15</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.011515</td>\n",
       "      <td>0.001263</td>\n",
       "      <td>1.283856</td>\n",
       "      <td>0.011515</td>\n",
       "      <td>0.001263</td>\n",
       "      <td>1.283856</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>LightGBM_r96</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.011652</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>0.653338</td>\n",
       "      <td>0.011652</td>\n",
       "      <td>0.001084</td>\n",
       "      <td>0.653338</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>LightGBM_r161</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.012635</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>1.484698</td>\n",
       "      <td>0.012635</td>\n",
       "      <td>0.001362</td>\n",
       "      <td>1.484698</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>LightGBM_r188</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.014626</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.732020</td>\n",
       "      <td>0.014626</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.732020</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>LightGBM</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.014687</td>\n",
       "      <td>0.001167</td>\n",
       "      <td>0.664778</td>\n",
       "      <td>0.014687</td>\n",
       "      <td>0.001167</td>\n",
       "      <td>0.664778</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>LightGBM_r196</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.016295</td>\n",
       "      <td>0.001274</td>\n",
       "      <td>1.419114</td>\n",
       "      <td>0.016295</td>\n",
       "      <td>0.001274</td>\n",
       "      <td>1.419114</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>LightGBM_r131</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mcc</td>\n",
       "      <td>0.016844</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>1.368974</td>\n",
       "      <td>0.016844</td>\n",
       "      <td>0.001209</td>\n",
       "      <td>1.368974</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   model  score_test  score_val eval_metric  pred_time_test  \\\n",
       "0               CatBoost    0.159878   0.417384         mcc        0.147774   \n",
       "1          CatBoost_r177    0.158581   0.425714         mcc        0.030552   \n",
       "2           CatBoost_r69    0.156772   0.398781         mcc        0.351471   \n",
       "3           XGBoost_r194    0.151047   0.498849         mcc        0.062239   \n",
       "4           CatBoost_r13    0.149776   0.385663         mcc        0.054072   \n",
       "5           CatBoost_r50    0.146453   0.398781         mcc        0.211605   \n",
       "6     NeuralNetTorch_r79    0.143356   0.396250         mcc        0.728093   \n",
       "7           CatBoost_r70    0.142126   0.438583         mcc        0.219357   \n",
       "8     NeuralNetTorch_r14    0.141889   0.263123         mcc        0.244909   \n",
       "9           CatBoost_r86    0.141749   0.431392         mcc        0.037711   \n",
       "10         CatBoost_r167    0.141231   0.387850         mcc        0.013611   \n",
       "11         CatBoost_r137    0.139254   0.389369         mcc        0.021138   \n",
       "12    NeuralNetTorch_r22    0.136628   0.374239         mcc        0.286723   \n",
       "13           CatBoost_r9    0.136167   0.385121         mcc        0.159969   \n",
       "14        NeuralNetTorch    0.134427   0.333661         mcc        0.271192   \n",
       "15    NeuralNetTorch_r41    0.134242   0.288288         mcc        0.735915   \n",
       "16    NeuralNetTorch_r30    0.132441   0.315244         mcc        0.297065   \n",
       "17    NeuralNetTorch_r86    0.125108   0.356226         mcc        0.723688   \n",
       "18  NeuralNetFastAI_r102    0.087097   0.115016         mcc        0.092321   \n",
       "19       NeuralNetFastAI    0.060760   0.055686         mcc        0.832277   \n",
       "20  NeuralNetFastAI_r156    0.060549  -0.011430         mcc        0.509034   \n",
       "21        ExtraTrees_r42    0.052854   0.293917         mcc        0.263247   \n",
       "22      RandomForest_r39    0.052245   0.293917         mcc        0.238857   \n",
       "23     RandomForest_r195    0.051760   0.342747         mcc        0.206467   \n",
       "24      RandomForestEntr    0.048248   0.333539         mcc        0.485226   \n",
       "25       ExtraTrees_r172    0.043153   0.156790         mcc        0.242218   \n",
       "26      RandomForestGini    0.043153   0.304294         mcc        0.279097   \n",
       "27        ExtraTreesEntr    0.043153   0.323532         mcc        0.310825   \n",
       "28        ExtraTreesGini    0.043153   0.351722         mcc        0.432763   \n",
       "29        KNeighborsDist    0.039795   0.296257         mcc        2.117112   \n",
       "30        KNeighborsUnif    0.036282   0.131379         mcc        2.031718   \n",
       "31         LightGBM_r130    0.000000   0.000000         mcc        0.010407   \n",
       "32         LightGBMLarge    0.000000   0.000000         mcc        0.010906   \n",
       "33          LightGBM_r15    0.000000   0.000000         mcc        0.011515   \n",
       "34          LightGBM_r96    0.000000   0.000000         mcc        0.011652   \n",
       "35         LightGBM_r161    0.000000   0.000000         mcc        0.012635   \n",
       "36         LightGBM_r188    0.000000   0.000000         mcc        0.014626   \n",
       "37              LightGBM    0.000000   0.000000         mcc        0.014687   \n",
       "38         LightGBM_r196    0.000000   0.000000         mcc        0.016295   \n",
       "39         LightGBM_r131    0.000000   0.000000         mcc        0.016844   \n",
       "\n",
       "    pred_time_val    fit_time  pred_time_test_marginal  \\\n",
       "0        0.002767   51.503006                 0.147774   \n",
       "1        0.003253   49.464640                 0.030552   \n",
       "2        0.003724  102.562262                 0.351471   \n",
       "3        0.016122   24.531862                 0.062239   \n",
       "4        0.006468  210.230490                 0.054072   \n",
       "5        0.008396   58.211248                 0.211605   \n",
       "6        0.033396  150.633491                 0.728093   \n",
       "7        0.006903   57.701282                 0.219357   \n",
       "8        0.031864   80.355489                 0.244909   \n",
       "9        0.003593  114.351807                 0.037711   \n",
       "10       0.003651   42.147018                 0.013611   \n",
       "11       0.003739  108.383495                 0.021138   \n",
       "12       0.033908  304.814172                 0.286723   \n",
       "13       0.008481   53.819530                 0.159969   \n",
       "14       0.031589  291.131433                 0.271192   \n",
       "15       0.033186  213.634619                 0.735915   \n",
       "16       0.035576  303.393694                 0.297065   \n",
       "17       0.037021  195.352012                 0.723688   \n",
       "18       0.012262   20.147879                 0.092321   \n",
       "19       0.014844   39.230214                 0.832277   \n",
       "20       0.008023   42.797305                 0.509034   \n",
       "21       0.024554   43.431954                 0.263247   \n",
       "22       0.025666  137.630308                 0.238857   \n",
       "23       0.024380  147.718517                 0.206467   \n",
       "24       0.024837    6.805090                 0.485226   \n",
       "25       0.024833   51.432341                 0.242218   \n",
       "26       0.024663    7.536692                 0.279097   \n",
       "27       0.025065    3.518496                 0.310825   \n",
       "28       0.024593    3.560127                 0.432763   \n",
       "29       0.135079    0.214777                 2.117112   \n",
       "30       0.172940    0.206955                 2.031718   \n",
       "31       0.003875    1.331746                 0.010407   \n",
       "32       0.001192    1.764692                 0.010906   \n",
       "33       0.001263    1.283856                 0.011515   \n",
       "34       0.001084    0.653338                 0.011652   \n",
       "35       0.001362    1.484698                 0.012635   \n",
       "36       0.001307    0.732020                 0.014626   \n",
       "37       0.001167    0.664778                 0.014687   \n",
       "38       0.001274    1.419114                 0.016295   \n",
       "39       0.001209    1.368974                 0.016844   \n",
       "\n",
       "    pred_time_val_marginal  fit_time_marginal  stack_level  can_infer  \\\n",
       "0                 0.002767          51.503006            1       True   \n",
       "1                 0.003253          49.464640            1       True   \n",
       "2                 0.003724         102.562262            1       True   \n",
       "3                 0.016122          24.531862            1       True   \n",
       "4                 0.006468         210.230490            1       True   \n",
       "5                 0.008396          58.211248            1       True   \n",
       "6                 0.033396         150.633491            1       True   \n",
       "7                 0.006903          57.701282            1       True   \n",
       "8                 0.031864          80.355489            1       True   \n",
       "9                 0.003593         114.351807            1       True   \n",
       "10                0.003651          42.147018            1       True   \n",
       "11                0.003739         108.383495            1       True   \n",
       "12                0.033908         304.814172            1       True   \n",
       "13                0.008481          53.819530            1       True   \n",
       "14                0.031589         291.131433            1       True   \n",
       "15                0.033186         213.634619            1       True   \n",
       "16                0.035576         303.393694            1       True   \n",
       "17                0.037021         195.352012            1       True   \n",
       "18                0.012262          20.147879            1       True   \n",
       "19                0.014844          39.230214            1       True   \n",
       "20                0.008023          42.797305            1       True   \n",
       "21                0.024554          43.431954            1       True   \n",
       "22                0.025666         137.630308            1       True   \n",
       "23                0.024380         147.718517            1       True   \n",
       "24                0.024837           6.805090            1       True   \n",
       "25                0.024833          51.432341            1       True   \n",
       "26                0.024663           7.536692            1       True   \n",
       "27                0.025065           3.518496            1       True   \n",
       "28                0.024593           3.560127            1       True   \n",
       "29                0.135079           0.214777            1       True   \n",
       "30                0.172940           0.206955            1       True   \n",
       "31                0.003875           1.331746            1       True   \n",
       "32                0.001192           1.764692            1       True   \n",
       "33                0.001263           1.283856            1       True   \n",
       "34                0.001084           0.653338            1       True   \n",
       "35                0.001362           1.484698            1       True   \n",
       "36                0.001307           0.732020            1       True   \n",
       "37                0.001167           0.664778            1       True   \n",
       "38                0.001274           1.419114            1       True   \n",
       "39                0.001209           1.368974            1       True   \n",
       "\n",
       "    fit_order  \n",
       "0           7  \n",
       "1          14  \n",
       "2          37  \n",
       "3          35  \n",
       "4          25  \n",
       "5          33  \n",
       "6          15  \n",
       "7          42  \n",
       "8          39  \n",
       "9          52  \n",
       "10         46  \n",
       "11         23  \n",
       "12         20  \n",
       "13         18  \n",
       "14         12  \n",
       "15         48  \n",
       "16         30  \n",
       "17         32  \n",
       "18         24  \n",
       "19         10  \n",
       "20         43  \n",
       "21         22  \n",
       "22         45  \n",
       "23         26  \n",
       "24          6  \n",
       "25         36  \n",
       "26          5  \n",
       "27          9  \n",
       "28          8  \n",
       "29          2  \n",
       "30          1  \n",
       "31         31  \n",
       "32         13  \n",
       "33         50  \n",
       "34         19  \n",
       "35         40  \n",
       "36         27  \n",
       "37          4  \n",
       "38         44  \n",
       "39         16  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leaderboard = fit.leaderboard(testdataset)\n",
    "leaderboard.iloc[:40]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "383f0fe6-5719-450a-a7af-9fb5e1cb8624",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'XGBoost_r194'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fit.model_best"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0162a794-b6b4-4dec-b53d-b27689e1dfeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.49884926592602014"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "leaderboard.query(f\"model == '{fit.model_best}'\")[\"score_val\"].item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "0a18beb4-990f-4a9a-90ed-47793a1e0283",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fit.set_model_best(\"NeuralNetFastAI_r37\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149acf9d-63bd-4f88-940a-58785eaa7a57",
   "metadata": {},
   "source": [
    "# Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "369160ad-55f9-4bd7-b52a-0496d47ef6b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mcc': 0.990807771391229,\n",
       " 'accuracy': 0.9989483976901647,\n",
       " 'balanced_accuracy': 0.9915411704694396,\n",
       " 'roc_auc': 0.9980941537495911,\n",
       " 'f1': 0.9913281606572342,\n",
       " 'precision': 0.9996931574102486,\n",
       " 'recall': 0.9831019915509958,\n",
       " 'confusion_matrix':         0     1\n",
       " 0  101776     2\n",
       " 1     112  6516,\n",
       " 'classification_report': {'0': {'precision': 0.9989007537688442,\n",
       "   'recall': 0.9999803493878834,\n",
       "   'f1-score': 0.9994402600335844,\n",
       "   'support': 101778.0},\n",
       "  '1': {'precision': 0.9996931574102486,\n",
       "   'recall': 0.9831019915509958,\n",
       "   'f1-score': 0.9913281606572342,\n",
       "   'support': 6628.0},\n",
       "  'accuracy': 0.9989483976901647,\n",
       "  'macro avg': {'precision': 0.9992969555895463,\n",
       "   'recall': 0.9915411704694396,\n",
       "   'f1-score': 0.9953842103454094,\n",
       "   'support': 108406.0},\n",
       "  'weighted avg': {'precision': 0.998949201745296,\n",
       "   'recall': 0.9989483976901647,\n",
       "   'f1-score': 0.9989442820003903,\n",
       "   'support': 108406.0}}}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trainresults = fit.evaluate(traindataset, detailed_report=True)\n",
    "trainresults"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f5dfca53-f48c-4c92-b310-6a80ba0568d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mcc': 0.15104724629743382,\n",
       " 'accuracy': 0.9378723404255319,\n",
       " 'balanced_accuracy': 0.5182977121286015,\n",
       " 'roc_auc': 0.76195746799451,\n",
       " 'f1': 0.07159679698539802,\n",
       " 'precision': 0.6846846846846847,\n",
       " 'recall': 0.03777335984095427,\n",
       " 'confusion_matrix':        0   1\n",
       " 0  29678  35\n",
       " 1   1936  76,\n",
       " 'classification_report': {'0': {'precision': 0.9387613082811412,\n",
       "   'recall': 0.9988220644162488,\n",
       "   'f1-score': 0.9678608117142531,\n",
       "   'support': 29713.0},\n",
       "  '1': {'precision': 0.6846846846846847,\n",
       "   'recall': 0.03777335984095427,\n",
       "   'f1-score': 0.07159679698539802,\n",
       "   'support': 2012.0},\n",
       "  'accuracy': 0.9378723404255319,\n",
       "  'macro avg': {'precision': 0.8117229964829129,\n",
       "   'recall': 0.5182977121286015,\n",
       "   'f1-score': 0.5197288043498256,\n",
       "   'support': 31725.0},\n",
       "  'weighted avg': {'precision': 0.9226477648082942,\n",
       "   'recall': 0.9378723404255319,\n",
       "   'f1-score': 0.9110197337746326,\n",
       "   'support': 31725.0}}}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = fit.evaluate(testdataset, detailed_report=True)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4bbc67ff-d21d-4315-99e2-11c14011ec1c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mcc 0.15104724629743382\n",
      "macro-f1 0.5197288043498256\n",
      "       0   1\n",
      "0  29678  35\n",
      "1   1936  76\n"
     ]
    }
   ],
   "source": [
    "print(\"mcc\", results[\"mcc\"])\n",
    "print(\"macro-f1\", results[\"classification_report\"][\"macro avg\"][\"f1-score\"])\n",
    "print(results[\"confusion_matrix\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3afb0a42-4220-4cde-9a00-04912bf7404c",
   "metadata": {},
   "source": [
    "## Extra set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "821038bb-6cac-4d4e-b34a-c9b4d3085b30",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label_label</th>\n",
       "      <th>Amino acids_label_comp_id_A</th>\n",
       "      <th>Amino acids_label_comp_id_C</th>\n",
       "      <th>Amino acids_label_comp_id_D</th>\n",
       "      <th>Amino acids_label_comp_id_E</th>\n",
       "      <th>Amino acids_label_comp_id_F</th>\n",
       "      <th>Amino acids_label_comp_id_G</th>\n",
       "      <th>Amino acids_label_comp_id_H</th>\n",
       "      <th>Amino acids_label_comp_id_I</th>\n",
       "      <th>Amino acids_label_comp_id_K</th>\n",
       "      <th>...</th>\n",
       "      <th>HHBlits_M-&gt;M</th>\n",
       "      <th>HHBlits_M-&gt;I</th>\n",
       "      <th>HHBlits_M-&gt;D</th>\n",
       "      <th>HHBlits_I-&gt;M</th>\n",
       "      <th>HHBlits_I-&gt;I</th>\n",
       "      <th>HHBlits_D-&gt;M</th>\n",
       "      <th>HHBlits_D-&gt;D</th>\n",
       "      <th>HHBlits_Neff</th>\n",
       "      <th>HHBlits_Neff_I</th>\n",
       "      <th>HHBlits_Neff_D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7gqu_A_12</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.992404</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007872</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.507</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7gqu_A_13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.970634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029688</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>11.647</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7gqu_A_14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.984184</td>\n",
       "      <td>0.016020</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.317098</td>\n",
       "      <td>0.683020</td>\n",
       "      <td>0.677362</td>\n",
       "      <td>0.322641</td>\n",
       "      <td>11.184</td>\n",
       "      <td>1.054</td>\n",
       "      <td>1.083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7gqu_A_15</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>11.324</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7gqu_A_16</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.983502</td>\n",
       "      <td>0.008717</td>\n",
       "      <td>0.007775</td>\n",
       "      <td>0.564091</td>\n",
       "      <td>0.435879</td>\n",
       "      <td>0.515842</td>\n",
       "      <td>0.483974</td>\n",
       "      <td>11.412</td>\n",
       "      <td>1.024</td>\n",
       "      <td>1.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dnm_C_488</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.971981</td>\n",
       "      <td>0.007883</td>\n",
       "      <td>0.020362</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.550571</td>\n",
       "      <td>0.449689</td>\n",
       "      <td>2.860</td>\n",
       "      <td>1.013</td>\n",
       "      <td>1.050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dnm_C_489</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.495171</td>\n",
       "      <td>0.504876</td>\n",
       "      <td>2.862</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dnm_C_490</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.374750</td>\n",
       "      <td>0.625031</td>\n",
       "      <td>2.849</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dnm_C_491</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.453445</td>\n",
       "      <td>0.546389</td>\n",
       "      <td>2.833</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9dnm_C_492</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.765</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5443 rows × 167 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Label_label  Amino acids_label_comp_id_A  \\\n",
       "7gqu_A_12             0                            0   \n",
       "7gqu_A_13             0                            0   \n",
       "7gqu_A_14             0                            0   \n",
       "7gqu_A_15             0                            0   \n",
       "7gqu_A_16             0                            1   \n",
       "...                 ...                          ...   \n",
       "9dnm_C_488            0                            0   \n",
       "9dnm_C_489            0                            0   \n",
       "9dnm_C_490            0                            0   \n",
       "9dnm_C_491            0                            0   \n",
       "9dnm_C_492            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_C  Amino acids_label_comp_id_D  \\\n",
       "7gqu_A_12                             0                            0   \n",
       "7gqu_A_13                             0                            0   \n",
       "7gqu_A_14                             0                            0   \n",
       "7gqu_A_15                             0                            0   \n",
       "7gqu_A_16                             0                            0   \n",
       "...                                 ...                          ...   \n",
       "9dnm_C_488                            0                            0   \n",
       "9dnm_C_489                            0                            0   \n",
       "9dnm_C_490                            0                            0   \n",
       "9dnm_C_491                            0                            1   \n",
       "9dnm_C_492                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_E  Amino acids_label_comp_id_F  \\\n",
       "7gqu_A_12                             0                            1   \n",
       "7gqu_A_13                             0                            0   \n",
       "7gqu_A_14                             0                            0   \n",
       "7gqu_A_15                             0                            0   \n",
       "7gqu_A_16                             0                            0   \n",
       "...                                 ...                          ...   \n",
       "9dnm_C_488                            0                            0   \n",
       "9dnm_C_489                            0                            1   \n",
       "9dnm_C_490                            1                            0   \n",
       "9dnm_C_491                            0                            0   \n",
       "9dnm_C_492                            0                            1   \n",
       "\n",
       "            Amino acids_label_comp_id_G  Amino acids_label_comp_id_H  \\\n",
       "7gqu_A_12                             0                            0   \n",
       "7gqu_A_13                             0                            0   \n",
       "7gqu_A_14                             0                            0   \n",
       "7gqu_A_15                             0                            0   \n",
       "7gqu_A_16                             0                            0   \n",
       "...                                 ...                          ...   \n",
       "9dnm_C_488                            0                            0   \n",
       "9dnm_C_489                            0                            0   \n",
       "9dnm_C_490                            0                            0   \n",
       "9dnm_C_491                            0                            0   \n",
       "9dnm_C_492                            0                            0   \n",
       "\n",
       "            Amino acids_label_comp_id_I  Amino acids_label_comp_id_K  ...  \\\n",
       "7gqu_A_12                             0                            0  ...   \n",
       "7gqu_A_13                             0                            0  ...   \n",
       "7gqu_A_14                             0                            0  ...   \n",
       "7gqu_A_15                             0                            0  ...   \n",
       "7gqu_A_16                             0                            0  ...   \n",
       "...                                 ...                          ...  ...   \n",
       "9dnm_C_488                            0                            0  ...   \n",
       "9dnm_C_489                            0                            0  ...   \n",
       "9dnm_C_490                            0                            0  ...   \n",
       "9dnm_C_491                            0                            0  ...   \n",
       "9dnm_C_492                            0                            0  ...   \n",
       "\n",
       "            HHBlits_M->M  HHBlits_M->I  HHBlits_M->D  HHBlits_I->M  \\\n",
       "7gqu_A_12       0.992404      0.000000      0.007872      0.000000   \n",
       "7gqu_A_13       0.970634      0.000000      0.029688      0.000000   \n",
       "7gqu_A_14       0.984184      0.016020      0.000000      0.317098   \n",
       "7gqu_A_15       1.000000      0.000000      0.000000      0.000000   \n",
       "7gqu_A_16       0.983502      0.008717      0.007775      0.564091   \n",
       "...                  ...           ...           ...           ...   \n",
       "9dnm_C_488      0.971981      0.007883      0.020362      1.000000   \n",
       "9dnm_C_489      1.000000      0.000000      0.000000      0.000000   \n",
       "9dnm_C_490      1.000000      0.000000      0.000000      0.000000   \n",
       "9dnm_C_491      1.000000      0.000000      0.000000      0.000000   \n",
       "9dnm_C_492      1.000000      0.000000      0.000000      0.000000   \n",
       "\n",
       "            HHBlits_I->I  HHBlits_D->M  HHBlits_D->D  HHBlits_Neff  \\\n",
       "7gqu_A_12       0.000000      0.000000      0.000000        11.507   \n",
       "7gqu_A_13       0.000000      0.000000      1.000000        11.647   \n",
       "7gqu_A_14       0.683020      0.677362      0.322641        11.184   \n",
       "7gqu_A_15       0.000000      0.000000      1.000000        11.324   \n",
       "7gqu_A_16       0.435879      0.515842      0.483974        11.412   \n",
       "...                  ...           ...           ...           ...   \n",
       "9dnm_C_488      0.000000      0.550571      0.449689         2.860   \n",
       "9dnm_C_489      0.000000      0.495171      0.504876         2.862   \n",
       "9dnm_C_490      0.000000      0.374750      0.625031         2.849   \n",
       "9dnm_C_491      0.000000      0.453445      0.546389         2.833   \n",
       "9dnm_C_492      0.000000      1.000000      0.000000         2.765   \n",
       "\n",
       "            HHBlits_Neff_I  HHBlits_Neff_D  \n",
       "7gqu_A_12            0.000           0.000  \n",
       "7gqu_A_13            0.000           1.004  \n",
       "7gqu_A_14            1.054           1.083  \n",
       "7gqu_A_15            0.000           1.019  \n",
       "7gqu_A_16            1.024           1.019  \n",
       "...                    ...             ...  \n",
       "9dnm_C_488           1.013           1.050  \n",
       "9dnm_C_489           0.000           1.059  \n",
       "9dnm_C_490           0.000           1.028  \n",
       "9dnm_C_491           0.000           1.016  \n",
       "9dnm_C_492           0.000           1.006  \n",
       "\n",
       "[5443 rows x 167 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"../training_data/7.Extra_set/features.pkl\", \"rb\") as f:\n",
    "    extra = process_dataframe(pd.concat((\n",
    "        pickle.load(f).values()\n",
    "    )))\n",
    "\n",
    "extra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "531dc670-ab8c-48b2-b433-3cecc19cdab4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mcc -0.004819616749499412\n",
      "macro-f1 0.4895432805026728\n",
      "      0  1\n",
      "0  5220  3\n",
      "1   220  0\n"
     ]
    }
   ],
   "source": [
    "extra_results = fit.evaluate(extra, detailed_report=True)\n",
    "print(\"mcc\", extra_results[\"mcc\"])\n",
    "print(\"macro-f1\", extra_results[\"classification_report\"][\"macro avg\"][\"f1-score\"])\n",
    "print(extra_results[\"confusion_matrix\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:pybiomed]",
   "language": "python",
   "name": "conda-env-pybiomed-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
